Index: android-2.6.35/arch/arm/plat-omap/include/plat/uncompress.h
===================================================================
--- android-2.6.35.orig/arch/arm/plat-omap/include/plat/uncompress.h	2010-12-10 15:11:19.000000000 +0530
+++ android-2.6.35/arch/arm/plat-omap/include/plat/uncompress.h	2010-12-22 14:37:00.112612633 +0530
@@ -146,7 +146,8 @@
 		DEBUG_LL_OMAP3(3, touchbook);
 
 		/* omap4 based boards using UART3 */
-		DEBUG_LL_OMAP4(3, omap_4430sdp);
+		/* DEBUG_LL_OMAP4(3, omap_4430sdp); */
+		DEBUG_LL_OMAP4(4, omap_4430sdp);
 		DEBUG_LL_OMAP4(3, omap4_panda);
 
 		/* zoom2/3 external uart */
