Index: android-2.6.35/arch/arm/mach-omap2/board-omap4tab10-1.c
===================================================================
--- android-2.6.35.orig/arch/arm/mach-omap2/board-omap4tab10-1.c	2010-12-24 21:15:21.831661805 +0530
+++ android-2.6.35/arch/arm/mach-omap2/board-omap4tab10-1.c	2010-12-24 21:15:23.652294304 +0530
@@ -71,77 +71,17 @@
 };
 
 static int omap4tab_keymap[] = {
-	KEY(0, 0, KEY_E),
-	KEY(0, 1, KEY_R),
-	KEY(0, 2, KEY_T),
-	KEY(0, 3, KEY_HOME),
-	KEY(0, 4, KEY_F5),
-	KEY(0, 5, KEY_UNKNOWN),
-	KEY(0, 6, KEY_I),
-	KEY(0, 7, KEY_LEFTSHIFT),
-
-	KEY(1, 0, KEY_D),
-	KEY(1, 1, KEY_F),
-	KEY(1, 2, KEY_G),
-	KEY(1, 3, KEY_SEND),
-	KEY(1, 4, KEY_F6),
-	KEY(1, 5, KEY_UNKNOWN),
-	KEY(1, 6, KEY_K),
-	KEY(1, 7, KEY_ENTER),
-
-	KEY(2, 0, KEY_X),
-	KEY(2, 1, KEY_C),
-	KEY(2, 2, KEY_V),
-	KEY(2, 3, KEY_END),
-	KEY(2, 4, KEY_F7),
-	KEY(2, 5, KEY_UNKNOWN),
-	KEY(2, 6, KEY_DOT),
-	KEY(2, 7, KEY_CAPSLOCK),
-
-	KEY(3, 0, KEY_Z),
-	KEY(3, 1, KEY_KPPLUS),
-	KEY(3, 2, KEY_B),
-	KEY(3, 3, KEY_F1),
-	KEY(3, 4, KEY_F8),
-	KEY(3, 5, KEY_UNKNOWN),
-	KEY(3, 6, KEY_O),
-	KEY(3, 7, KEY_SPACE),
-
-	KEY(4, 0, KEY_W),
-	KEY(4, 1, KEY_Y),
-	KEY(4, 2, KEY_U),
-	KEY(4, 3, KEY_F2),
-	KEY(4, 4, KEY_VOLUMEUP),
-	KEY(4, 5, KEY_UNKNOWN),
-	KEY(4, 6, KEY_L),
-	KEY(4, 7, KEY_LEFT),
-
-	KEY(5, 0, KEY_S),
-	KEY(5, 1, KEY_H),
-	KEY(5, 2, KEY_J),
-	KEY(5, 3, KEY_F3),
-	KEY(5, 4, KEY_F9),
-	KEY(5, 5, KEY_VOLUMEDOWN),
-	KEY(5, 6, KEY_M),
-	KEY(5, 7, KEY_RIGHT),
-
-	KEY(6, 0, KEY_Q),
-	KEY(6, 1, KEY_A),
-	KEY(6, 2, KEY_N),
-	KEY(6, 3, KEY_BACK),
-	KEY(6, 4, KEY_BACKSPACE),
-	KEY(6, 5, KEY_UNKNOWN),
-	KEY(6, 6, KEY_P),
-	KEY(6, 7, KEY_UP),
-
-	KEY(7, 0, KEY_PROG1),
-	KEY(7, 1, KEY_PROG2),
-	KEY(7, 2, KEY_PROG3),
-	KEY(7, 3, KEY_PROG4),
-	KEY(7, 4, KEY_F4),
-	KEY(7, 5, KEY_UNKNOWN),
-	KEY(7, 6, KEY_OK),
-	KEY(7, 7, KEY_DOWN),
+	/* TODO: The row and column coded here are based on what is detected by
+	 * OMAP4. It is different from the one shown in schematic. Need the
+	 * identify the root cause of the problem */
+	/* row, col, key */
+	KEY(3, 3, KEY_HOME),		//HOME key
+	KEY(3, 4, KEY_VOLUMEUP),	//Volume up
+	KEY(4, 3, KEY_F1),		//Spare key
+	KEY(4, 4, KEY_VOLUMEDOWN),	//volume down
+	KEY(5, 3, KEY_ENTER),		//Search Key
+	KEY(5, 4, KEY_BACK),
+	0,
 };
 
 static struct matrix_keymap_data omap4tab_keymap_data = {
Index: android-2.6.35/arch/arm/mach-omap2/Kconfig
===================================================================
--- android-2.6.35.orig/arch/arm/mach-omap2/Kconfig	2010-12-24 21:14:58.092294213 +0530
+++ android-2.6.35/arch/arm/mach-omap2/Kconfig	2010-12-24 21:15:23.652294304 +0530
@@ -247,19 +247,10 @@
 	select LEDS_OMAP4430_SDP_KEYPAD
 	select LEDS_TRIGGERS
 	select LEDS_TRIGGER_TIMER
-	select INPUT_CMA3000_I2C
-#	select SENSORS_BH1780
-#	select INPUT_BH1780
 	select MISC_DEVICES
-#	select BMP085
-	select INPUT_BMP085
 	select STAGING
 	select IIO
-#	select SENSORS_HMC5843
-	select INPUT_HMC5843
-	select INPUT_SFH7741
 	select HWMON
-#	select SENSORS_LM75
 	select INPUT_MISC
 	select INPUT_KEYBOARD
 	select INPUT_TOUCHSCREEN
Index: android-2.6.35/arch/arm/configs/android_4430_samsung_defconfig
===================================================================
--- android-2.6.35.orig/arch/arm/configs/android_4430_samsung_defconfig	2010-12-24 21:15:21.841662047 +0530
+++ android-2.6.35/arch/arm/configs/android_4430_samsung_defconfig	2010-12-24 21:15:23.662288272 +0530
@@ -1,7 +1,7 @@
 #
 # Automatically generated make config: don't edit
 # Linux kernel version: 2.6.35
-# Fri Oct  8 14:41:16 2010
+# Fri Oct  8 18:25:01 2010
 #
 CONFIG_ARM=y
 CONFIG_HAVE_PWM=y
@@ -950,7 +950,7 @@
 # CONFIG_INPUT_BH1780 is not set
 # CONFIG_INPUT_ATI_REMOTE is not set
 # CONFIG_INPUT_ATI_REMOTE2 is not set
-CONFIG_INPUT_KEYCHORD=y
+# CONFIG_INPUT_KEYCHORD is not set
 # CONFIG_INPUT_KEYSPAN_REMOTE is not set
 # CONFIG_INPUT_POWERMATE is not set
 # CONFIG_INPUT_YEALINK is not set
@@ -962,10 +962,10 @@
 CONFIG_INPUT_GPIO=y
 # CONFIG_INPUT_PCF8574 is not set
 # CONFIG_INPUT_GPIO_ROTARY_ENCODER is not set
-CONFIG_INPUT_SFH7741=y
-CONFIG_INPUT_CMA3000_I2C=y
-CONFIG_INPUT_HMC5843=y
-CONFIG_INPUT_BMP085=y
+# CONFIG_INPUT_SFH7741 is not set
+# CONFIG_INPUT_CMA3000_I2C is not set
+# CONFIG_INPUT_HMC5843 is not set
+# CONFIG_INPUT_BMP085 is not set
 
 #
 # Hardware I/O ports
@@ -993,15 +993,7 @@
 #
 # Serial drivers
 #
-CONFIG_SERIAL_8250=y
-CONFIG_SERIAL_8250_CONSOLE=y
-CONFIG_SERIAL_8250_NR_UARTS=32
-CONFIG_SERIAL_8250_RUNTIME_UARTS=4
-CONFIG_SERIAL_8250_EXTENDED=y
-CONFIG_SERIAL_8250_MANY_PORTS=y
-CONFIG_SERIAL_8250_SHARE_IRQ=y
-CONFIG_SERIAL_8250_DETECT_IRQ=y
-CONFIG_SERIAL_8250_RSA=y
+# CONFIG_SERIAL_8250 is not set
 
 #
 # Non-8250 serial port support
@@ -1562,7 +1554,8 @@
 # CONFIG_USB_DEVICEFS is not set
 CONFIG_USB_DEVICE_CLASS=y
 # CONFIG_USB_DYNAMIC_MINORS is not set
-# CONFIG_USB_SUSPEND is not set
+CONFIG_USB_SUSPEND=y
+# CONFIG_USB_OTG is not set
 # CONFIG_USB_OTG_WHITELIST is not set
 # CONFIG_USB_OTG_BLACKLIST_HUB is not set
 # CONFIG_USB_MON is not set
@@ -1573,16 +1566,13 @@
 # USB Host Controller Drivers
 #
 # CONFIG_USB_C67X00_HCD is not set
-CONFIG_USB_EHCI_HCD=y
-# CONFIG_USB_EHCI_ROOT_HUB_TT is not set
-# CONFIG_USB_EHCI_TT_NEWSCHED is not set
+# CONFIG_USB_EHCI_HCD is not set
 # CONFIG_USB_OXU210HP_HCD is not set
 # CONFIG_USB_ISP116X_HCD is not set
 # CONFIG_USB_ISP1760_HCD is not set
 # CONFIG_USB_ISP1362_HCD is not set
 CONFIG_USB_OHCI_HCD=y
 CONFIG_USB_OHCI_HCD_OMAP3=y
-CONFIG_USB_EHCI_HCD_OMAP3=y
 # CONFIG_USB_OHCI_BIG_ENDIAN_DESC is not set
 # CONFIG_USB_OHCI_BIG_ENDIAN_MMIO is not set
 CONFIG_USB_OHCI_LITTLE_ENDIAN=y
