Index: android-2.6.35_10.1/arch/arm/mach-omap2/Makefile
===================================================================
--- android-2.6.35_10.1.orig/arch/arm/mach-omap2/Makefile	2010-12-30 11:28:33.078706697 +0530
+++ android-2.6.35_10.1/arch/arm/mach-omap2/Makefile	2010-12-30 11:48:14.147463061 +0530
@@ -188,6 +188,7 @@
 obj-$(CONFIG_MACH_OMAP_4430SDP)		+= board-4430sdp.o \
 					   hsmmc.o cdc_tcxo.o
 obj-$(CONFIG_MACH_OMAP4_TAB_10_1)	+= board-omap4tab10-1.o \
+					   board-4430univ-sensors.o \
 					   hsmmc.o
 
 obj-$(CONFIG_MACH_OMAP4_PANDA)		+= board-omap4panda.o \
Index: android-2.6.35_10.1/arch/arm/mach-omap2/board-omap4tab10-1.c
===================================================================
--- android-2.6.35_10.1.orig/arch/arm/mach-omap2/board-omap4tab10-1.c	2010-12-30 11:48:11.388707044 +0530
+++ android-2.6.35_10.1/arch/arm/mach-omap2/board-omap4tab10-1.c	2010-12-30 11:48:14.147463061 +0530
@@ -55,7 +55,14 @@
 #define SDP4430_LCD_PANEL_BACKLIGHT_ENABLE_GPIO		16
 #define SDP4430_MLCD1_PANEL_ENABLE_GPIO			2
 #define SDP4430_MLCD2_PANEL_ENABLE_GPIO			40
-#define SDP4430_LVDS_nRST_ENABLE_GPIO                  17
+#define SDP4430_LVDS_nRST_ENABLE_GPIO			17
+#define OMAP4SDP_MDM_PWR_EN_GPIO			157
+#define OMAP4430UNIV_ISL29023_INT_GPIO			52
+#define OMAP4430UNIV_KXSD9_INT_GPIO			94
+#define OMAP4430UNIV_AKM8973_INT_GPIO			157
+
+#define CONTROL_CORE_PAD0_I2C3_SDA_PAD1_I2C4_SCL	0x4A10012C
+#define CONTROL_CORE_PAD0_I2C4_SDA_PAD1_MCSPI1_CLK	0x4A100130
 
 
 static unsigned backlight_gpio;
@@ -64,6 +71,7 @@
 static unsigned lvds_nrst_gpio;
 static int lcd_enabled;
 
+void omap4430univ_sensors_init(void)
 
 static struct platform_device omap4tab_hdmi_audio_device = {
 	.name		= "hdmi-dai",
@@ -517,8 +525,11 @@
 
 static struct twl4030_codec_audio_data twl6040_audio = {
 	.audio_mclk	= 38400000,
-	.audpwron_gpio  = 127,
-	.naudint_irq    = OMAP44XX_IRQ_SYS_2N,
+	/* provide GPIO number above the valid value
+	 * to mean there is no GPIO connected,
+	 * likewise do not provide any valid IRQ number */
+	.audpwron_gpio	= 1024,
+	.naudint_irq	= 0,
 };
 
 static struct twl4030_codec_vibra_data twl6040_vibra = {
@@ -563,7 +574,10 @@
 };
 
 static struct i2c_board_info __initdata omap4tab_i2c_2_boardinfo[] = {
-
+	{
+		I2C_BOARD_INFO("isl29023", 0x44),
+		.irq	= OMAP_GPIO_IRQ(OMAP4430UNIV_ISL29023_INT_GPIO),
+	},
 };
 
 static struct i2c_board_info __initdata omap4tab_i2c_3_boardinfo[] = {
@@ -571,7 +585,14 @@
 };
 
 static struct i2c_board_info __initdata omap4tab_i2c_4_boardinfo[] = {
-
+	{
+		I2C_BOARD_INFO("akm8973", 0x1C),
+		.irq	= OMAP_GPIO_IRQ(OMAP4430UNIV_AKM8973_INT_GPIO),
+	},
+	{
+		I2C_BOARD_INFO("kxsd9", 0x18),
+		.irq	= OMAP_GPIO_IRQ(OMAP4430UNIV_KXSD9_INT_GPIO),
+	},
 };
 
 static const struct usbhs_omap_platform_data usbhs_pdata __initconst = {
@@ -764,11 +785,46 @@
 	omap_writew(padconf, CONTROL_CORE_PAD1_GPMC_AD11);
 }
 
+void omap4430_internal_i2c4_pull_up(void)
+{
+	int val = 0;
+
+	void __iomem *ctrl_core_pad0_i2c4_scl, *ctrl_core_pad0_i2c4_sda;
+
+	/* I2C4_SCL pull-up */
+	ctrl_core_pad0_i2c4_scl = (u32 *) ioremap(CONTROL_CORE_PAD0_I2C3_SDA_PAD1_I2C4_SCL, 4);
+	if (!ctrl_core_pad0_i2c4_scl) {
+		printk(KERN_ERR"OMAP_pad_config: ioremap failed with addr %x\n",
+				CONTROL_CORE_PAD0_I2C3_SDA_PAD1_I2C4_SCL);
+		return;
+	}
+
+	val =  __raw_readl(ctrl_core_pad0_i2c4_scl);
+	val |= 0x180000;  /* Set 19,20 bit high for I2C4 SCL pull up */
+	__raw_writel(val, ctrl_core_pad0_i2c4_scl);
+
+	/* I2C4_SDA pull-up */
+	ctrl_core_pad0_i2c4_sda = (u32 *) ioremap(CONTROL_CORE_PAD0_I2C4_SDA_PAD1_MCSPI1_CLK, 4);
+	if (!ctrl_core_pad0_i2c4_sda) {
+		printk(KERN_ERR"OMAP_pad_config: ioremap failed with addr %x\n",
+				CONTROL_CORE_PAD0_I2C4_SDA_PAD1_MCSPI1_CLK);
+		return;
+	}
+
+	val =  __raw_readl(ctrl_core_pad0_i2c4_sda);
+	val |= 0x18;  /* Set 3, 4 bit high for I2C4 SCL pull up */
+	__raw_writel(val, ctrl_core_pad0_i2c4_sda);
+
+	iounmap(ctrl_core_pad0_i2c4_scl);
+	iounmap(ctrl_core_pad0_i2c4_sda);
+}
+
 static void __init omap4_tab_10_1_init(void)
 {
 	int status;
 
 	omap4tab_gpiowk_setup();
+	omap4430_internal_i2c4_pull_up();
 	omap4_i2c_init();
 	omap4tab_display_init();
 	platform_add_devices(omap4tab_devices, ARRAY_SIZE(omap4tab_devices));
@@ -798,6 +854,8 @@
 	msleep(10);
 
 	samsung_omap4_pwr_key_irq_init();
+
+	omap4430univ_sensors_init();
 }
 
 static void __init omap4_tab_10_1_map_io(void)
Index: android-2.6.35_10.1/arch/arm/mach-omap2/board-4430univ-sensors.c
===================================================================
--- /dev/null	1970-01-01 00:00:00.000000000 +0000
+++ android-2.6.35_10.1/arch/arm/mach-omap2/board-4430univ-sensors.c	2010-12-30 11:48:14.147463061 +0530
@@ -0,0 +1,75 @@
+/*
+ * Board support file for containing sensors specific details for OMAP4430 SDP.
+ *
+ * Copyright (C) 2009 Google, Inc.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 2 as
+ * published by the Free Software Foundation.
+ */
+
+#include <linux/kernel.h>
+#include <linux/platform_device.h>
+#include <linux/init.h>
+#include <linux/input.h>
+#include <linux/delay.h>
+#include <linux/regulator/consumer.h>
+
+#include <plat/mux.h>
+#include <plat/gpio.h>
+#include <plat/keypad.h>
+
+#define OMAP4430SDP_AKM8973_INT_GPIO		157
+#define OMAP4430SDP_AKM8973_RESET_GPIO		46
+#define OMAP4430SDP_ISL29023_INT_GPIO		52
+#define OMAP4430SDP_KXSD9_INT_GPIO		94
+
+#define CONTROL_CORE_PAD0_GPMC_A22_PAD1_GPMC_A23 0x4A10006c
+
+
+static void __init omap4430univ_akm8973_init(void)
+{
+	int val =0;
+	u32 * ctrl_core_pad0_gpmc_a22;
+
+	/* MUX-ing GPMC_A22 into GPIIO_46 */
+	ctrl_core_pad0_gpmc_a22 = (u32 *) ioremap(CONTROL_CORE_PAD0_GPMC_A22_PAD1_GPMC_A23, 4);
+	if (!ctrl_core_pad0_gpmc_a22) {
+		printk(KERN_ERR"OMAP_pad_config: ioremap failed with addr %lx\n",
+				CONTROL_CORE_PAD0_GPMC_A22_PAD1_GPMC_A23);
+		return;
+	}
+
+	val =  __raw_readl(ctrl_core_pad0_gpmc_a22);
+	val |= 0x3;  /* Set 0,1 bits high for GPMC_A22 to be configured as GPIO_46 */
+	__raw_writel(val, ctrl_core_pad0_gpmc_a22);
+
+	iounmap(ctrl_core_pad0_gpmc_a22);
+
+	gpio_request(OMAP4430SDP_AKM8973_RESET_GPIO, "akm8973 reset");
+	gpio_direction_output(OMAP4430SDP_AKM8973_RESET_GPIO, 0);
+	gpio_direction_output(OMAP4430SDP_AKM8973_RESET_GPIO, 1);
+
+	gpio_request(OMAP4430SDP_AKM8973_INT_GPIO, "akm8973 irq");
+	gpio_direction_input(OMAP4430SDP_AKM8973_INT_GPIO);
+}
+
+static void __init omap4430univ_isl29023_init(void)
+{
+	gpio_request(OMAP4430SDP_ISL29023_INT_GPIO, "isl29023 irq");
+	gpio_direction_input(OMAP4430SDP_ISL29023_INT_GPIO);
+}
+
+static void __init omap4430univ_kxsd9_init(void)
+{
+	gpio_request(OMAP4430SDP_KXSD9_INT_GPIO, "kxsd9 irq");
+	gpio_direction_input(OMAP4430SDP_KXSD9_INT_GPIO);
+}
+
+void __init omap4430univ_sensors_init(void)
+{
+	omap4430univ_akm8973_init();
+	omap4430univ_isl29023_init();
+	omap4430univ_kxsd9_init();
+}
+
Index: android-2.6.35_10.1/arch/arm/plat-omap/include/plat/board.h
===================================================================
--- android-2.6.35_10.1.orig/arch/arm/plat-omap/include/plat/board.h	2010-12-30 11:48:11.378712464 +0530
+++ android-2.6.35_10.1/arch/arm/plat-omap/include/plat/board.h	2010-12-30 11:48:14.147463061 +0530
@@ -161,6 +161,8 @@
 /* for TI reference platforms sharing the same debug card */
 extern int debug_card_init(u32 addr, unsigned gpio);
 
+extern void omap4430univ_sensors_init(void);
+
 /* OMAP3EVM revision */
 #if defined(CONFIG_MACH_OMAP3EVM)
 u8 get_omap3_evm_rev(void);
Index: android-2.6.35_10.1/drivers/Kconfig
===================================================================
--- android-2.6.35_10.1.orig/drivers/Kconfig	2010-12-10 15:11:20.000000000 +0530
+++ android-2.6.35_10.1/drivers/Kconfig	2010-12-30 11:48:14.157457157 +0530
@@ -14,6 +14,8 @@
 
 source "drivers/block/Kconfig"
 
+source "drivers/sensors/Kconfig"
+
 # misc before ide - BLK_DEV_SGIIOC4 depends on SGI_IOC4
 
 source "drivers/misc/Kconfig"
Index: android-2.6.35_10.1/drivers/sensors/Kconfig
===================================================================
--- /dev/null	1970-01-01 00:00:00.000000000 +0000
+++ android-2.6.35_10.1/drivers/sensors/Kconfig	2010-12-30 11:48:14.157457157 +0530
@@ -0,0 +1,7 @@
+
+menuconfig NEW_SENSORS
+	bool "Sensors Support"
+	help
+	  Say Y to enable Sensors support.  This allows control of supported
+	  Sensors.
+
Index: android-2.6.35_10.1/drivers/sensors/Makefile
===================================================================
--- /dev/null	1970-01-01 00:00:00.000000000 +0000
+++ android-2.6.35_10.1/drivers/sensors/Makefile	2010-12-30 11:48:14.157457157 +0530
@@ -0,0 +1,9 @@
+#
+#  Makefile for the kernel Sensor device drivers.
+#
+
+# Object files in subdirectories
+obj-$(CONFIG_ACCEL)		+= accel/
+obj-$(CONFIG_OPTICAL)		+= optical/
+obj-$(CONFIG_COMPASS)		+= compass/
+
