Index: android-2.6.35/drivers/video/omap2/displays/panel-samsung-ltn121at.c
===================================================================
--- android-2.6.35.orig/drivers/video/omap2/displays/panel-samsung-ltn121at.c	2010-12-22 14:36:02.491988531 +0530
+++ android-2.6.35/drivers/video/omap2/displays/panel-samsung-ltn121at.c	2010-12-22 14:36:05.353239112 +0530
@@ -92,7 +92,6 @@
 	struct samsung_data *sd = dev_get_drvdata(&dssdev->dev);
 	int r = 0;
 
-	printk("[SHANKAR] %s [%d] entry \n", __func__, __LINE__);
 	/* wait couple of vsyncs until enabling the LCD */
 	msleep(1);
 	regulator_enable(sd->vusim_reg);
@@ -108,7 +107,6 @@
 		return r;
 	}
 	dssdev->state = OMAP_DSS_DISPLAY_ACTIVE;
-	printk("[SHANKAR] %s [%d] exit \n", __func__, __LINE__);
 	return r;
 }
 
@@ -116,7 +114,6 @@
 {
 	struct samsung_data *sd = dev_get_drvdata(&dssdev->dev);
 
-	printk("[SHANKAR] %s [%d] entry \n", __func__, __LINE__);
 	if (dssdev->state != OMAP_DSS_DISPLAY_ACTIVE)
 		return;
 
@@ -130,7 +127,6 @@
 
 	/* wait at least 5 vsyncs after disabling the LCD */
 	msleep(1);
-	printk("[SHANKAR] %s [%d] exit \n", __func__, __LINE__);
 }
 
 static int samsung_ltn121at_panel_suspend(struct omap_dss_device *dssdev)
@@ -143,6 +139,23 @@
 {
 	return samsung_ltn121at_panel_enable(dssdev);
 }
+static void samsung_ltn121at_panel_set_timings(struct omap_dss_device *dssdev,
+                struct omap_video_timings *timings)
+{
+        dpi_set_timings(dssdev, timings);
+}
+
+static void samsung_ltn121at_panel_get_timings(struct omap_dss_device *dssdev,
+                struct omap_video_timings *timings)
+{
+        *timings = dssdev->panel.timings;
+}
+
+static int samsung_ltn121at_panel_check_timings(struct omap_dss_device *dssdev,
+                struct omap_video_timings *timings)
+{
+        return dpi_check_timings(dssdev, timings);
+}
 
 static struct omap_dss_driver samsung_ltn121at_driver = {
 	.probe		= samsung_ltn121at_panel_probe,
@@ -153,6 +166,9 @@
 	.get_resolution = samsung_ltn121at_panel_get_resolution,
 	.suspend	= samsung_ltn121at_panel_suspend,
 	.resume		= samsung_ltn121at_panel_resume,
+	.set_timings	= samsung_ltn121at_panel_set_timings,
+	.get_timings	= samsung_ltn121at_panel_get_timings,
+	.check_timings	= samsung_ltn121at_panel_check_timings,
 
 	.driver         = {
 		.name   = "samsung_ltn121at_panel",
