Index: android-2.6.35/arch/arm/mach-omap2/board-omap4tab10-1.c
===================================================================
--- android-2.6.35.orig/arch/arm/mach-omap2/board-omap4tab10-1.c	2010-12-22 14:36:05.151988409 +0530
+++ android-2.6.35/arch/arm/mach-omap2/board-omap4tab10-1.c	2010-12-22 14:36:50.553862523 +0530
@@ -79,7 +79,7 @@
 static unsigned lvds_nrst_gpio;
 static int lcd_enabled;
 
-void omap4430univ_sensors_init(void)
+void omap4430univ_sensors_init(void);
 
 static struct platform_device omap4tab_hdmi_audio_device = {
 	.name		= "hdmi-dai",
Index: android-2.6.35/drivers/sensors/accel/kxsd9.c
===================================================================
--- android-2.6.35.orig/drivers/sensors/accel/kxsd9.c	2010-12-22 14:36:04.141988371 +0530
+++ android-2.6.35/drivers/sensors/accel/kxsd9.c	2010-12-22 14:36:40.301988465 +0530
@@ -329,7 +329,7 @@
 	kxsd9->client = client;
 	i2c_set_clientdata(client, kxsd9);
 
-	result = register_chrdev( ACC_DEV_MAJOR, ACC_DEV_NAME, &acc_fops);
+	result = register_chrdev( 0, ACC_DEV_NAME, &acc_fops);
 	if (result < 0) {
 		return result;
 	}
@@ -337,12 +337,12 @@
         acc_class = class_create (THIS_MODULE, ACC_DEV_NAME);
 
         if (IS_ERR(acc_class))	{
-		unregister_chrdev( ACC_DEV_MAJOR, ACC_DEV_NAME);
+		unregister_chrdev( result, ACC_DEV_NAME);
 		return PTR_ERR( acc_class );
 	}
 
 
-	dev_t = device_create( acc_class, NULL, MKDEV(ACC_DEV_MAJOR, 0), "%s", ACC_DEV_NAME);
+	dev_t = device_create( acc_class, NULL, MKDEV(result, 0), "%s", ACC_DEV_NAME);
         if (IS_ERR(dev_t))
         {
                 return PTR_ERR(dev_t);
