Index: Kernel-p1/arch/arm/mach-omap2/board-4430sdp-wifi.c
===================================================================
--- Kernel-p1.orig/arch/arm/mach-omap2/board-4430sdp-wifi.c	2011-01-24 11:55:54.000000000 +0530
+++ Kernel-p1/arch/arm/mach-omap2/board-4430sdp-wifi.c	2011-01-24 15:48:03.307780992 +0530
@@ -15,6 +15,8 @@
 /* linux/arch/arm/mach-omap2/board-4430sdp-wifi.c
 */
 
+
+
 #include <linux/kernel.h>
 #include <linux/init.h>
 #include <linux/platform_device.h>
