Index: android-2.6.35_10.1/arch/arm/mach-omap2/board-omap4tab10-1.c
===================================================================
--- android-2.6.35_10.1.orig/arch/arm/mach-omap2/board-omap4tab10-1.c	2010-12-30 11:49:40.037462456 +0530
+++ android-2.6.35_10.1/arch/arm/mach-omap2/board-omap4tab10-1.c	2010-12-30 11:49:42.648086297 +0530
@@ -580,7 +580,7 @@
 static struct omap2_hsmmc_info mmc[] = {
 	{
 		.mmc		= 2,
-		.wires		= 8,
+		.caps		= MMC_CAP_8_BIT_DATA,
 		.gpio_cd	= -EINVAL,
 		.gpio_wp	= -EINVAL,
 		.ocr_mask	= MMC_VDD_165_195,
@@ -591,7 +591,7 @@
 	},
 	{
 		.mmc		= 1,
-		.wires		= 8,
+		.caps		= MMC_CAP_8_BIT_DATA,
 		.gpio_cd	= -EINVAL,
 		.gpio_wp	= -EINVAL,
 #ifdef CONFIG_PM_RUNTIME
@@ -600,7 +600,7 @@
 	},
 	{
 		.mmc		= 5,
-		.wires		= 8,
+		.caps		= MMC_CAP_8_BIT_DATA,
 		.gpio_cd	= -EINVAL,
 		.gpio_wp        = 4,
 		.ocr_mask	= MMC_VDD_165_195,
Index: android-2.6.35_10.1/arch/arm/plat-omap/include/plat/control.h
===================================================================
--- android-2.6.35_10.1.orig/arch/arm/plat-omap/include/plat/control.h	2010-12-30 11:20:06.898086027 +0530
+++ android-2.6.35_10.1/arch/arm/plat-omap/include/plat/control.h	2010-12-30 11:49:42.648086297 +0530
@@ -230,6 +230,9 @@
 #define OMAP3_PADCONF_SAD2D_MSTANDBY   0x250
 #define OMAP3_PADCONF_SAD2D_IDLEACK    0x254
 
+/* 44xx-only CONTROL_GENERAL register offsets */
+#define OMAP44XX_CONTROL_MMC1			0x628
+#define OMAP44XX_CONTROL_PBIAS_LITE		0x600
 
 /* 44xx control efuse offsets */
 #define OMAP44XX_CONTROL_FUSE_IVA_OPP50		0x22C
