Index: android-2.6.35/arch/arm/mach-omap2/board-omap4tab10-1.c
===================================================================
--- android-2.6.35.orig/arch/arm/mach-omap2/board-omap4tab10-1.c	2010-12-22 14:36:08.431988271 +0530
+++ android-2.6.35/arch/arm/mach-omap2/board-omap4tab10-1.c	2010-12-22 14:36:41.372612299 +0530
@@ -73,6 +73,8 @@
 #elif (CONFIG_SAMSUNG_OMAP4_TAB_REV == 4)
 #define OMAP4TAB_LCD_EN_GPIO				40
 #define OMAP4TAB_MLCD_RST_GPIO				37
+#define OMAP4TAB_MICBIAS_EN_GPIO			48
+#define OMAP4TAB_MICBIAS_EN2_GPIO			49
 #endif
 
 #define OMAP4SDP_MDM_PWR_EN_GPIO			157
@@ -1297,6 +1299,13 @@
 	printk("[SHANKAR] %s[%d] exit \n", __func__, __LINE__);
 	
 }
+static void omap4tab_audio_micbias_init(void)
+{
+	gpio_request(OMAP4TAB_MICBIAS_EN_GPIO, "MICBIAS ENABLE GPIO for MAIN_MIC");
+	gpio_direction_output(OMAP4TAB_MICBIAS_EN_GPIO, 1);
+	gpio_request(OMAP4TAB_MICBIAS_EN2_GPIO, "MICBIAS ENABLE2 GPIO for EAR_MIC");
+	gpio_direction_output(OMAP4TAB_MICBIAS_EN2_GPIO, 1);
+}
 static void omap4tab_vib_motor_init(void)
 {
 }
@@ -1354,6 +1363,7 @@
 	omap4tab_gpiowk_setup();
 	omap4tab_touch_init();
 	omap4tab_vib_motor_init();
+	omap4tab_audio_micbias_init();
 	omap4430_internal_i2c4_pull_up();
 	omap4_i2c_init();
 	omap4tab_display_init();
