Index: android-2.6.35_10.1/drivers/video/omap2/dss/overlay.c
===================================================================
--- android-2.6.35_10.1.orig/drivers/video/omap2/dss/overlay.c	2010-12-30 11:19:50.087455562 +0530
+++ android-2.6.35_10.1/drivers/video/omap2/dss/overlay.c	2010-12-30 11:50:58.117462603 +0530
@@ -553,6 +553,7 @@
 			outh = info->out_height;
 	}
 
+#if 0
 	if ((dw < info->pos_x + outw) && !info->out_wb) {
 		DSSDBG("check_overlay failed 1: %d < %d + %d\n",
 				dw, info->pos_x, outw);
@@ -564,6 +565,7 @@
 				dh, info->pos_y, outh);
 		return -EINVAL;
 	}
+#endif
 
 	if ((ovl->supported_modes & info->color_mode) == 0) {
 		DSSERR("overlay doesn't support mode %d\n", info->color_mode);
