Index: android-2.6.35_10.1/arch/arm/mach-omap2/board-omap4tab10-1.c
===================================================================
--- android-2.6.35_10.1.orig/arch/arm/mach-omap2/board-omap4tab10-1.c	2010-12-30 11:50:56.098712222 +0530
+++ android-2.6.35_10.1/arch/arm/mach-omap2/board-omap4tab10-1.c	2010-12-30 11:54:06.758712853 +0530
@@ -931,11 +931,8 @@
 
 static struct twl4030_codec_audio_data twl6040_audio = {
 	.audio_mclk	= 38400000,
-	/* provide GPIO number above the valid value
-	 * to mean there is no GPIO connected,
-	 * likewise do not provide any valid IRQ number */
-	.audpwron_gpio	= 1024,
-	.naudint_irq	= 0,
+	.audpwron_gpio	= 127,
+	.naudint_irq	= OMAP44XX_IRQ_SYS_2N,
 };
 
 static struct twl4030_codec_vibra_data twl6040_vibra = {
