Index: android-2.6.35/arch/arm/mach-omap2/board-omap4tab10-1.c
===================================================================
--- android-2.6.35.orig/arch/arm/mach-omap2/board-omap4tab10-1.c	2010-12-22 14:36:09.263238425 +0530
+++ android-2.6.35/arch/arm/mach-omap2/board-omap4tab10-1.c	2010-12-22 14:36:40.211988384 +0530
@@ -54,6 +54,7 @@
 #include "hsmmc.h"
 #include "smartreflex-class3.h"
 #include <linux/haptic.h>
+#include <plat/opp_twl_tps.h>
 
 #include <linux/atmel_mxt1386.h> /* For mXT1386 Touch */
 
@@ -1319,6 +1320,83 @@
 	omap_writew(padconf, CONTROL_CORE_PAD1_GPMC_AD11);
 }
 
+
+static struct omap_volt_pmic_info omap_pmic_core = {
+	.name = "twl",
+	.slew_rate = 4000,
+	.step_size = 12500,
+	.i2c_addr = 0x12,
+	.i2c_vreg = 0x61,
+	.vsel_to_uv = omap_twl_vsel_to_uv,
+	.uv_to_vsel = omap_twl_uv_to_vsel,
+	.onforce_cmd = omap_twl_onforce_cmd,
+	.on_cmd = omap_twl_on_cmd,
+	.sleepforce_cmd = omap_twl_sleepforce_cmd,
+	.sleep_cmd = omap_twl_sleep_cmd,
+	.vp_config_erroroffset = 0,
+	.vp_vstepmin_vstepmin = 0x01,
+	.vp_vstepmax_vstepmax = 0x04,
+	.vp_vlimitto_timeout_us = 0x200,
+	.vp_vlimitto_vddmin = 0x18,
+	.vp_vlimitto_vddmax = 0x30,
+};
+
+static struct omap_volt_pmic_info omap_pmic_mpu = {
+	.name = "twl",
+	.slew_rate = 4000,
+	.step_size = 12500,
+	.i2c_addr = 0x12,
+	.i2c_vreg = 0x55,
+	.vsel_to_uv = omap_twl_vsel_to_uv,
+	.uv_to_vsel = omap_twl_uv_to_vsel,
+	.onforce_cmd = omap_twl_onforce_cmd,
+	.on_cmd = omap_twl_on_cmd,
+	.sleepforce_cmd = omap_twl_sleepforce_cmd,
+	.sleep_cmd = omap_twl_sleep_cmd,
+	.vp_config_erroroffset = 0,
+	.vp_vstepmin_vstepmin = 0x01,
+	.vp_vstepmax_vstepmax = 0x04,
+	.vp_vlimitto_timeout_us = 0x200,
+	.vp_vlimitto_vddmin = 0x18,
+	.vp_vlimitto_vddmax = 0x3c,
+};
+
+static struct omap_volt_pmic_info omap_pmic_iva = {
+	.name = "twl",
+	.slew_rate = 4000,
+	.step_size = 12500,
+	.i2c_addr = 0x12,
+	.i2c_vreg = 0x5b,
+	.vsel_to_uv = omap_twl_vsel_to_uv,
+	.uv_to_vsel = omap_twl_uv_to_vsel,
+	.onforce_cmd = omap_twl_onforce_cmd,
+	.on_cmd = omap_twl_on_cmd,
+	.sleepforce_cmd = omap_twl_sleepforce_cmd,
+	.sleep_cmd = omap_twl_sleep_cmd,
+	.vp_config_erroroffset = 0,
+	.vp_vstepmin_vstepmin = 0x01,
+	.vp_vstepmax_vstepmax = 0x04,
+	.vp_vlimitto_timeout_us = 0x200,
+	.vp_vlimitto_vddmin = 0x18,
+	.vp_vlimitto_vddmax = 0x3c,
+};
+
+static struct omap_volt_vc_data vc_config = {
+	.vdd0_on = 1350000,        /* 1.35v */
+	.vdd0_onlp = 1350000,      /* 1.35v */
+	.vdd0_ret = 837500,       /* 0.8375v */
+	.vdd0_off = 600000,       /* 0.6v */
+	.vdd1_on = 1100000,        /* 1.1v */
+	.vdd1_onlp = 1100000,      /* 1.1v */
+	.vdd1_ret = 837500,       /* 0.8375v */
+	.vdd1_off = 600000,       /* 0.6v */
+	.vdd2_on = 1100000,        /* 1.1v */
+	.vdd2_onlp = 1100000,      /* 1.1v */
+	.vdd2_ret = 837500,       /* .8375v */
+	.vdd2_off = 600000,       /* 0.6v */
+};
+
+
 void omap4430_internal_i2c4_pull_up(void)
 {
 	int val = 0;
@@ -1388,6 +1466,10 @@
 	omap_display_init(&omap4tab_dss_data);
 	enable_board_wakeup_source();
 
+	omap_voltage_register_pmic(&omap_pmic_core, "core");
+	omap_voltage_register_pmic(&omap_pmic_mpu, "mpu");
+	omap_voltage_register_pmic(&omap_pmic_iva, "iva");
+	omap_voltage_init_vc(&vc_config);
 
 	samsung_omap4_pwr_key_irq_init();
 #if (CONFIG_SAMSUNG_OMAP4_TAB_REV == 4)
