Index: android-2.6.35_10.1/arch/arm/mach-omap2/board-omap4tab10-1.c
===================================================================
--- android-2.6.35_10.1.orig/arch/arm/mach-omap2/board-omap4tab10-1.c	2010-12-30 11:54:20.097462793 +0530
+++ android-2.6.35_10.1/arch/arm/mach-omap2/board-omap4tab10-1.c	2010-12-30 11:54:22.667456661 +0530
@@ -941,12 +941,25 @@
 	.irq_line	= 1,
 };
 
+static int omap4tab_batt_table[] = {
+	/* adc code for temperature in degree C */
+	929, 925, /* -2 ,-1 */
+	920, 917, 912, 908, 904, 899, 895, 890, 885, 880, /* 00 - 09 */
+	875, 869, 864, 858, 853, 847, 841, 835, 829, 823, /* 10 - 19 */
+	816, 810, 804, 797, 790, 783, 776, 769, 762, 755, /* 20 - 29 */
+	748, 740, 732, 725, 718, 710, 703, 695, 687, 679, /* 30 - 39 */
+	671, 663, 655, 647, 639, 631, 623, 615, 607, 599, /* 40 - 49 */
+	591, 583, 575, 567, 559, 551, 543, 535, 527, 519, /* 50 - 59 */
+	511, 504, 496 /* 60 - 62 */
+};
 static struct twl4030_bci_platform_data omap4tab_bci_data = {
 	.monitoring_interval		= 10,
 	.max_charger_currentmA		= 1500,
 	.max_charger_voltagemV		= 4560,
 	.max_bat_voltagemV		= 4200,
 	.low_bat_voltagemV		= 3300,
+	.battery_tmp_tbl		= omap4tab_batt_table,
+	.tblsize			= ARRAY_SIZE(omap4tab_batt_table),
 };
 
 static struct twl4030_codec_audio_data twl6040_audio = {
Index: android-2.6.35_10.1/drivers/power/twl6030_bci_battery.c
===================================================================
--- android-2.6.35_10.1.orig/drivers/power/twl6030_bci_battery.c	2010-12-30 11:19:50.117455642 +0530
+++ android-2.6.35_10.1/drivers/power/twl6030_bci_battery.c	2010-12-30 11:54:22.667456661 +0530
@@ -218,6 +218,7 @@
 	int			vac_priority;
 	int			bat_health;
 	int			charger_source;
+	int			bat_technology;
 
 	int			fuelgauge_mode;
 	int			timer_n2;
@@ -757,6 +758,7 @@
 	POWER_SUPPLY_PROP_CURRENT_AVG,
 	POWER_SUPPLY_PROP_CAPACITY,
 	POWER_SUPPLY_PROP_TEMP,
+	POWER_SUPPLY_PROP_TECHNOLOGY,
 };
 
 static enum power_supply_property twl6030_usb_props[] = {
@@ -1079,6 +1081,9 @@
 	case POWER_SUPPLY_PROP_HEALTH:
 		val->intval = di->bat_health;
 		break;
+	case POWER_SUPPLY_PROP_TECHNOLOGY:
+		val->intval = di->bat_technology;
+		break;
 	case POWER_SUPPLY_PROP_CAPACITY:
 		val->intval = di->capacity;
 		break;
@@ -1620,6 +1625,8 @@
 	di->low_bat_voltagemV = pdata->low_bat_voltagemV;
 	di->battery_tmp_tbl = pdata->battery_tmp_tbl;
 	di->tblsize = pdata->tblsize;
+	di->bat_health = POWER_SUPPLY_HEALTH_GOOD;
+	di->bat_technology = POWER_SUPPLY_TECHNOLOGY_LION;
 
 	di->dev = &pdev->dev;
 	di->bat.name = "twl6030_battery";
