Index: android-2.6.35_10.1/drivers/video/omap2/dss/dispc.c
===================================================================
--- android-2.6.35_10.1.orig/drivers/video/omap2/dss/dispc.c	2010-12-30 11:28:57.938081182 +0530
+++ android-2.6.35_10.1/drivers/video/omap2/dss/dispc.c	2010-12-30 11:58:39.777456992 +0530
@@ -4668,7 +4668,8 @@
 			}
 		}
 	}
-
+#if 0 
+	[SHANKAR] workaround added for continus flickering while Video playback
 	if (errors & DISPC_IRQ_VID1_FIFO_UNDERFLOW) {
 		DSSERR("VID1_FIFO_UNDERFLOW, disabling VID1\n");
 		for (i = 0; i < omap_dss_get_num_overlays(); ++i) {
@@ -4686,7 +4687,7 @@
 			}
 		}
 	}
-
+#endif
 	if (errors & DISPC_IRQ_VID2_FIFO_UNDERFLOW) {
 		DSSERR("VID2_FIFO_UNDERFLOW, disabling VID2\n");
 		for (i = 0; i < omap_dss_get_num_overlays(); ++i) {
