Index: android-2.6.35/arch/arm/mach-omap2/board-omap4tab10-1.c
===================================================================
--- android-2.6.35.orig/arch/arm/mach-omap2/board-omap4tab10-1.c	2010-12-22 14:36:10.393239023 +0530
+++ android-2.6.35/arch/arm/mach-omap2/board-omap4tab10-1.c	2010-12-22 14:36:37.013862680 +0530
@@ -34,6 +34,8 @@
 
 #include <mach/hardware.h>
 #include <mach/omap4-common.h>
+#include <mach/emif.h>
+#include <mach/lpddr2-elpida.h>
 #include <asm/mach-types.h>
 #include <asm/mach/arch.h>
 #include <asm/mach/map.h>
@@ -51,6 +53,7 @@
 #include <plat/omap4-keypad.h>
 #include <plat/hwspinlock.h>
 #include <plat/nokia-dsi-panel.h>
+//#include "mux.h"
 #include "hsmmc.h"
 #include "smartreflex-class3.h"
 #include <linux/haptic.h>
@@ -1100,6 +1103,23 @@
 static struct omap_i2c_bus_board_data __initdata omap4tab_i2c_3_bus_pdata;
 static struct omap_i2c_bus_board_data __initdata omap4tab_i2c_4_bus_pdata;
 
+/*
+ * LPDDR2 Configuration Data:
+ * The memory organisation is as below :
+ *	EMIF1 - CS0 -   2 Gb
+ *	CS1 -   NULL
+ *	EMIF2 - CS0 -   2 Gb
+ *	CS1 -    NULL
+ *	--------------------
+ *	TOTAL -         4 Gb
+ *
+ * Same devices installed on EMIF1 and EMIF2
+ */
+static __initdata struct emif_device_details emif_devices = {
+	.cs0_device = &elpida_2G_S4,
+	.cs1_device = NULL
+};
+
 static void __init omap_i2c_hwspinlock_init(int bus_id, unsigned int
 			spinlock_id, struct omap_i2c_bus_board_data *pdata)
 {
@@ -1519,6 +1539,8 @@
                 package = OMAP_PACKAGE_CBL;
         omap4_mux_init(board_mux, package);
 
+	omap_emif_setup_device_details(&emif_devices, &emif_devices);
+	omap_init_emif_timings();
 	omap4tab_gpiowk_setup();
 	omap4tab_touch_init();
 	omap4tab_vib_motor_init();
