Index: android-2.6.35/arch/arm/mach-omap2/board-omap4tab10-1.c
===================================================================
--- android-2.6.35.orig/arch/arm/mach-omap2/board-omap4tab10-1.c	2010-12-22 14:36:11.501988741 +0530
+++ android-2.6.35/arch/arm/mach-omap2/board-omap4tab10-1.c	2010-12-22 14:36:33.243862668 +0530
@@ -120,7 +120,7 @@
 
 
 /* For uUSB Switch */
-#define OMAP4430_GPIO_JACK_NINT                140
+#define OMAP4430_GPIO_JACK_NINT                121
 
 static void omap4_usb_cb(u8 attached);
 static void omap4_charger_cb(u8 attached);
@@ -480,7 +480,7 @@
 };
 
 /* wl128x BT, FM, GPS connectivity chip */
-static int gpios[] = {55, -1, -1};
+static int gpios[] = {103, -1, -1};
 static struct platform_device wl128x_device = {
 	.name		= "kim",
 	.id		= -1,
@@ -1071,11 +1071,6 @@
 	{
 		I2C_BOARD_INFO("secFuelgaugeDev", 0x34),
 	},
-       {
-               I2C_BOARD_INFO("fsa9480", (0x4A >> 1)),
-               .platform_data = &omap4_fsa9480_pdata,
-               .irq = OMAP_GPIO_IRQ(OMAP4430_GPIO_JACK_NINT),
-       },
 
 };
 
@@ -1138,6 +1133,11 @@
 		I2C_BOARD_INFO("gp2a",0x44),
 		
 	},
+        {
+               I2C_BOARD_INFO("fsa9480", (0x4A >> 1)),
+               .platform_data = &omap4_fsa9480_pdata,
+               .irq = OMAP_GPIO_IRQ(OMAP4430_GPIO_JACK_NINT),
+        },
 #endif
 
 };
