Index: Kernel-p1/drivers/video/omap2/dss/dispc.c
===================================================================
--- Kernel-p1.orig/drivers/video/omap2/dss/dispc.c	2011-01-24 16:03:15.000000000 +0530
+++ Kernel-p1/drivers/video/omap2/dss/dispc.c	2011-01-24 16:04:18.757775758 +0530
@@ -280,9 +280,6 @@
 	DISPC_VID_V3_WB_ATTRIBUTES(0),	/* VID 3 pipeline */
 	DISPC_VID_V3_WB_ATTRIBUTES(1),	/* WB pipeline */
 	};
-#ifdef CONFIG_ARCH_OMAP4
-        static u32 pcd_highres_flag=0;
-#endif
 
 
 struct dispc_irq_stats {
@@ -5003,11 +5000,6 @@
 	DSSDBG("dispc_enable_plane %d, %d\n", plane, enable);
 
 	enable_clocks(1);
-        if( (!enable) && (pcd_highres_flag)){
-                dispc_set_lcd_divisor(OMAP_DSS_CHANNEL_LCD2, 1, 4);
-                pcd_highres_flag = 0;
-        }
-
 	_dispc_enable_plane(plane, enable);
 	enable_clocks(0);
 }
@@ -5040,16 +5032,6 @@
 
 	enable_clocks(1);
 
-        if( (plane != OMAP_DSS_GFX) && (channel != OMAP_DSS_CHANNEL_DIGIT)
-                && (pcd_highres_flag != 1) && (width >= 1280)){
-                DSSDBG(KERN_ERR"\n [DISPC] In %s , Changing disp clock\n",__func__);
-                //Jags_31_08 es2.0 dss clock timing corrections ++
-                pcd_highres_flag = 1;
-                dispc_set_lcd_divisor(OMAP_DSS_CHANNEL_LCD2, 1, 7);
-                //Jags_31_08 es2.0 dss clock timing corrections --
-        }
-
-
 	r = _dispc_setup_plane(plane,
 			   paddr, screen_width,
 			   pos_x, pos_y,
Index: Kernel-p1/drivers/video/omap2/dss/dss.c
===================================================================
--- Kernel-p1.orig/drivers/video/omap2/dss/dss.c	2011-01-24 11:55:54.000000000 +0530
+++ Kernel-p1/drivers/video/omap2/dss/dss.c	2011-01-24 16:04:18.757775758 +0530
@@ -500,9 +500,19 @@
 	int match = 0;
 	int min_fck_per_pck;
 
-	prate = dss_get_dpll4_rate();
-
 	fck = dss_clk_get_rate(DSS_CLK_FCK1);
+	if (!cpu_is_omap44xx()) {
+		prate = dss_get_dpll4_rate();
+	} else {
+		/* Linux ARM clk framework issue currently all DPLL lock frequency
+		 * multipliers are not configured properly. Hence multiplying by
+		 * 2 to get the right pixel clock dividers.
+		 *
+		 * TODO: Need to change this once appropriate patches are added
+		 * to clock framework.
+		 */
+		fck *= 2;
+	}
 	if (req_pck == dss.cache_req_pck &&
 			((cpu_is_omap34xx() && prate == dss.cache_prate) ||
 			 dss.cache_dss_cinfo.fck == fck)) {
@@ -574,6 +584,21 @@
 					goto found;
 			}
 		}
+	} else if (cpu_is_omap44xx()) {
+		struct dispc_clock_info cur_dispc;
+
+		dispc_find_clk_divs(is_tft, req_pck, fck, &cur_dispc);
+		match = 1;
+		
+		if (abs(cur_dispc.pck - req_pck) < abs(best_dispc.pck - req_pck)) {
+			best_dss.fck = fck;
+			best_dss.fck_div = 1;
+
+			best_dispc = cur_dispc;
+
+			if (cur_dispc.pck == req_pck)
+				goto found;
+		}
 	} else {
 		BUG();
 	}
Index: Kernel-p1/drivers/video/omap2/displays/panel-ams353.c
===================================================================
--- Kernel-p1.orig/drivers/video/omap2/displays/panel-ams353.c	2011-01-24 15:57:06.000000000 +0530
+++ Kernel-p1/drivers/video/omap2/displays/panel-ams353.c	2011-01-24 16:04:18.757775758 +0530
@@ -49,7 +49,7 @@
 #define LCD_VFP		    28
 #define LCD_VSW		    3   // 2	
 
-#define LCD_PIXCLOCK_MAX	        24685       // Archer_LSJ DB07 25530 //12000  
+#define LCD_PIXCLOCK_MAX	        24380       
 #define LCD_DEFAULT_BRIGHTNESS	    6
 #define LCD_DEFAULT_GAMMA           GAMMA_2_2   // 2.2 gamma table.
 #define HAND_FLASH_DEFAULT          flash_off	// Archer_LSJ DB10
