Index: android-2.6.35/drivers/video/omap2/displays/panel-ams353.c
===================================================================
--- android-2.6.35.orig/drivers/video/omap2/displays/panel-ams353.c	2010-12-22 14:36:12.061988855 +0530
+++ android-2.6.35/drivers/video/omap2/displays/panel-ams353.c	2010-12-22 14:36:12.371988666 +0530
@@ -384,7 +384,7 @@
 		refresh_count = 10;
 	}
     #endif
-    
+	dssdev->state = OMAP_DSS_DISPLAY_SUSPENDED;
 	return 0;
 }
 
@@ -395,7 +395,7 @@
 	spi_setup(ams353lcd_spi);
     
 	ams353_panel_enable(dssdev);
-
+	dssdev->state = OMAP_DSS_DISPLAY_ACTIVE;
 	return 0;
 }
 
