Index: android-2.6.35/drivers/input/keyboard/cypress/cypress-touchkey.c
===================================================================
--- android-2.6.35.orig/drivers/input/keyboard/cypress/cypress-touchkey.c	2010-12-22 14:36:08.193238472 +0530
+++ android-2.6.35/drivers/input/keyboard/cypress/cypress-touchkey.c	2010-12-22 14:36:12.903238733 +0530
@@ -291,6 +291,7 @@
 	}
 
 	disable_irq(IRQ_TOUCH_INT);
+	disable_irq(home_key_irq);
 	gpio_direction_output(_3_GPIO_TOUCH_EN, 0);
 #ifdef I2C_BIT_BANG
 	gpio_direction_output(_3_TOUCH_SDA_28V, 0);
@@ -317,8 +318,8 @@
 #endif
 
 	enable_irq(IRQ_TOUCH_INT);
+	enable_irq(home_key_irq);
 	touchkey_enable = 1;
-
 }
 #endif				// End of CONFIG_HAS_EARLYSUSPEND
 
