Index: android-2.6.35_10.1/arch/arm/mach-omap2/pm44xx.c
===================================================================
--- android-2.6.35_10.1.orig/arch/arm/mach-omap2/pm44xx.c	2010-12-30 11:21:02.937457001 +0530
+++ android-2.6.35_10.1/arch/arm/mach-omap2/pm44xx.c	2010-12-30 12:01:15.378712696 +0530
@@ -386,6 +386,7 @@
 	omap4_wakeupgen_set_interrupt(cpu_id, OMAP44XX_IRQ_GPT1);
 	omap4_wakeupgen_set_interrupt(cpu_id, OMAP44XX_IRQ_PRCM);
 	omap4_wakeupgen_set_interrupt(cpu_id, OMAP44XX_IRQ_SYS_1N);
+	omap4_wakeupgen_set_interrupt(cpu_id, OMAP44XX_IRQ_GPIO1);
 
 	/* Read current next_pwrsts */
 	list_for_each_entry(pwrst, &pwrst_list, node) {
@@ -647,6 +648,9 @@
 	prm_rmw_mod_reg_bits(OMAP4430_IO_ST_MASK, OMAP4430_IO_ST_MASK,
 		OMAP4430_PRM_OCP_SOCKET_MOD, OMAP4_PRM_IRQENABLE_MPU_OFFSET);
 
+	/* Enable WKUP Dpendency for GPIO1*/
+	prm_write_mod_reg(0x1, OMAP4430_PRM_WKUP_MOD, OMAP4_PM_WKUP_GPIO1_WKDEP_OFFSET);
+
 	/* Enable GLOBAL_WUEN */
 	prm_rmw_mod_reg_bits(OMAP4430_GLOBAL_WUEN_MASK, OMAP4430_GLOBAL_WUEN_MASK,
 		OMAP4430_PRM_DEVICE_MOD, OMAP4_PRM_IO_PMCTRL_OFFSET);
Index: android-2.6.35_10.1/arch/arm/plat-omap/gpio.c
===================================================================
--- android-2.6.35_10.1.orig/arch/arm/plat-omap/gpio.c	2010-12-10 15:11:19.000000000 +0530
+++ android-2.6.35_10.1/arch/arm/plat-omap/gpio.c	2010-12-30 12:01:15.378712696 +0530
@@ -1575,6 +1575,10 @@
 					OMAP4_GPIO_IRQSTATUSCLR0);
 			__raw_writel(0x00000000, bank->base +
 					 OMAP4_GPIO_DEBOUNCENABLE);
+			if(bank->id == 0) {
+				__raw_writel(0x0000001D, bank->base +
+					OMAP1610_GPIO_SYSCONFIG);
+			}
 			/* Initialize interface clk ungated, module enabled */
 			__raw_writel(0, bank->base + OMAP4_GPIO_CTRL);
 		} else if (cpu_is_omap34xx()) {
@@ -1778,8 +1782,14 @@
 
 	spin_lock_irqsave(&bank->lock, flags);
 	bank->saved_wakeup = __raw_readl(wake_status);
-	__raw_writel(0xffffffff, wake_clear);
-	__raw_writel(bank->suspend_wakeup, wake_set);
+	if((bank->id == 1) && ((bank->saved_wakeup & 0x8) == 0x8)) {
+		__raw_writel((0xffffffff & ~0x8), wake_clear);
+		__raw_writel((bank->suspend_wakeup & ~0x8), wake_set);
+		__raw_writel(0x1D, bank->base + OMAP1610_GPIO_SYSCONFIG);
+	} else {
+		__raw_writel(0xffffffff, wake_clear);
+		__raw_writel(bank->suspend_wakeup, wake_set);
+	}
 	spin_unlock_irqrestore(&bank->lock, flags);
 
 	return 0;
