Index: android-2.6.35_10.1/arch/arm/mach-omap2/pm44xx.c
===================================================================
--- android-2.6.35_10.1.orig/arch/arm/mach-omap2/pm44xx.c	2010-12-30 12:01:15.000000000 +0530
+++ android-2.6.35_10.1/arch/arm/mach-omap2/pm44xx.c	2010-12-30 12:09:26.047461054 +0530
@@ -335,9 +335,9 @@
 
 	/* Check if a IO_ST interrupt */
 	if (irqstatus_mpu & OMAP4430_IO_ST_MASK) {
-		/* Re-enable UART3 */
-		omap_writel(0x2, 0x4A009550);
-		omap_writel(0xD, 0x48020054);
+		/* Re-enable UART4 */
+		omap_writel(0x2, 0x4A009558);
+		omap_writel(0xD, 0x4806E054);
 
 		/* usbhs remote wakeup */
 		usbhs_wakeup();
@@ -381,7 +381,7 @@
 	 * as a wakeup event from MPU/Device OFF
 	 */
 	omap4_wakeupgen_clear_all(cpu_id);
-	omap4_wakeupgen_set_interrupt(cpu_id, OMAP44XX_IRQ_UART3);
+	omap4_wakeupgen_set_interrupt(cpu_id, OMAP44XX_IRQ_UART4);
 	omap4_wakeupgen_set_interrupt(cpu_id, OMAP44XX_IRQ_KBD_CTL);
 	omap4_wakeupgen_set_interrupt(cpu_id, OMAP44XX_IRQ_GPT1);
 	omap4_wakeupgen_set_interrupt(cpu_id, OMAP44XX_IRQ_PRCM);
