Index: android-2.6.35_10.1/arch/arm/mach-omap2/pm44xx.c
===================================================================
--- android-2.6.35_10.1.orig/arch/arm/mach-omap2/pm44xx.c	2010-12-30 12:10:09.068712793 +0530
+++ android-2.6.35_10.1/arch/arm/mach-omap2/pm44xx.c	2010-12-30 12:10:54.718712627 +0530
@@ -382,7 +382,6 @@
 	 */
 	omap4_wakeupgen_clear_all(cpu_id);
 	omap4_wakeupgen_set_interrupt(cpu_id, OMAP44XX_IRQ_UART4);
-	omap4_wakeupgen_set_interrupt(cpu_id, OMAP44XX_IRQ_KBD_CTL);
 	omap4_wakeupgen_set_interrupt(cpu_id, OMAP44XX_IRQ_GPT1);
 	omap4_wakeupgen_set_interrupt(cpu_id, OMAP44XX_IRQ_PRCM);
 	omap4_wakeupgen_set_interrupt(cpu_id, OMAP44XX_IRQ_SYS_1N);
Index: android-2.6.35_10.1/drivers/input/keyboard/omap4-keypad.c
===================================================================
--- android-2.6.35_10.1.orig/drivers/input/keyboard/omap4-keypad.c	2010-12-10 15:11:21.000000000 +0530
+++ android-2.6.35_10.1/drivers/input/keyboard/omap4-keypad.c	2010-12-30 12:10:54.718712627 +0530
@@ -87,8 +87,6 @@
 			keypad_data->base + OMAP4_KBD_IRQSTATUS);
 	__raw_writel(OMAP4_DEF_IRQENABLE_EVENTEN | OMAP4_DEF_IRQENABLE_LONGKEY,
 			keypad_data->base + OMAP4_KBD_IRQENABLE);
-	__raw_writel(OMAP4_DEF_WUP_EVENT_ENA | OMAP4_DEF_WUP_LONG_KEY_ENA,
-			keypad_data->base + OMAP4_KBD_WAKEUPENABLE);
 }
 
 /* Interrupt handler */
