Index: Kernel-p1/drivers/video/omap2/dss/dispc.c
===================================================================
--- Kernel-p1.orig/drivers/video/omap2/dss/dispc.c	2011-01-24 16:04:18.000000000 +0530
+++ Kernel-p1/drivers/video/omap2/dss/dispc.c	2011-01-24 16:07:31.569022860 +0530
@@ -1924,7 +1924,7 @@
 			REG_GET(ftrs_reg[plane], 11, 0),
 			REG_GET(ftrs_reg[plane], 27, 16),
 			low, high);
-
+#if 0
 	if (cpu_is_omap24xx())
 		dispc_write_reg(ftrs_reg[plane],
 				FLD_VAL(high, 24, 16) | FLD_VAL(low, 8, 0));
@@ -1934,6 +1934,7 @@
 	else	/* cpu is omap44xx */
 		dispc_write_reg(ftrs_reg[plane],
 				FLD_VAL(high, 31, 16) | FLD_VAL(low, 15, 0));
+#endif
 
 	enable_clocks(0);
 }
@@ -4939,6 +4940,9 @@
 
 	dispc_set_loadmode(OMAP_DSS_LOAD_FRAME_ONLY);
 
+	REG_FLD_MOD(DISPC_DIVISOR1, 1, 0, 0);
+	REG_FLD_MOD(DISPC_DIVISOR1, 1, 23, 16);
+
 	dispc_read_plane_fifo_sizes();
 
 	/* Enabling the DISPC_DIVISOR and setting the LCD to 1 */
