Index: android-2.6.35_10.1/arch/arm/mach-omap2/board-omap4tab10-1.c
===================================================================
--- android-2.6.35_10.1.orig/arch/arm/mach-omap2/board-omap4tab10-1.c	2010-12-30 12:04:01.157456831 +0530
+++ android-2.6.35_10.1/arch/arm/mach-omap2/board-omap4tab10-1.c	2010-12-30 12:16:07.718712725 +0530
@@ -1466,6 +1466,14 @@
 	padconf = omap_readw(CONTROL_CORE_PAD1_GPMC_AD11);
 	padconf |= OMAP44XX_PADCONF_WAKEUPENABLE0;
 	omap_writew(padconf, CONTROL_CORE_PAD1_GPMC_AD11);
+	/* Enabling WKUP capability for TouchKey IRQ GPI0_32 */
+	padconf = omap_readw(CONTROL_CORE_PAD0_GPMC_AD08);
+	padconf |= OMAP44XX_PADCONF_WAKEUPENABLE0;
+	omap_writew(padconf, CONTROL_CORE_PAD0_GPMC_AD08);
+	/* Enabling WKUP capability for TouchScreen IRQ GPI0_43 */
+	padconf = omap_readw(CONTROL_CORE_PAD1_GPMC_AD19);
+	padconf |= OMAP44XX_PADCONF_WAKEUPENABLE1;
+	omap_writew(padconf, CONTROL_CORE_PAD1_GPMC_AD19);
 }
 
 
Index: android-2.6.35_10.1/arch/arm/mach-omap2/pm44xx.c
===================================================================
--- android-2.6.35_10.1.orig/arch/arm/mach-omap2/pm44xx.c	2010-12-30 12:10:54.718712627 +0530
+++ android-2.6.35_10.1/arch/arm/mach-omap2/pm44xx.c	2010-12-30 12:16:07.718712725 +0530
@@ -650,6 +650,9 @@
 	/* Enable WKUP Dpendency for GPIO1*/
 	prm_write_mod_reg(0x1, OMAP4430_PRM_WKUP_MOD, OMAP4_PM_WKUP_GPIO1_WKDEP_OFFSET);
 
+	/* Enable WKUP Dependency for GPIO2*/
+	prm_write_mod_reg(0x1, OMAP4430_PRM_L4PER_MOD, OMAP4_PM_L4PER_GPIO2_WKDEP_OFFSET);
+
 	/* Enable GLOBAL_WUEN */
 	prm_rmw_mod_reg_bits(OMAP4430_GLOBAL_WUEN_MASK, OMAP4430_GLOBAL_WUEN_MASK,
 		OMAP4430_PRM_DEVICE_MOD, OMAP4_PRM_IO_PMCTRL_OFFSET);
Index: android-2.6.35_10.1/arch/arm/plat-omap/include/plat/control.h
===================================================================
--- android-2.6.35_10.1.orig/arch/arm/plat-omap/include/plat/control.h	2010-12-30 11:49:42.648086297 +0530
+++ android-2.6.35_10.1/arch/arm/plat-omap/include/plat/control.h	2010-12-30 12:16:07.718712725 +0530
@@ -256,6 +256,8 @@
 
 #define CONTROL_CORE_PAD0_GPMC_AD10	0x4A100054
 #define CONTROL_CORE_PAD1_GPMC_AD11	0x4A100056
+#define CONTROL_CORE_PAD0_GPMC_AD08	0x4A100050
+#define CONTROL_CORE_PAD1_GPMC_AD19	0x4A100066
 
 /*
  * REVISIT: This list of registers is not comprehensive - there are more
Index: android-2.6.35_10.1/drivers/input/keyboard/cypress/cypress-touchkey.c
===================================================================
--- android-2.6.35_10.1.orig/drivers/input/keyboard/cypress/cypress-touchkey.c	2010-12-30 12:01:14.397462577 +0530
+++ android-2.6.35_10.1/drivers/input/keyboard/cypress/cypress-touchkey.c	2010-12-30 12:16:07.718712725 +0530
@@ -291,6 +291,7 @@
 	}
 
 	disable_irq(IRQ_TOUCH_INT);
+	disable_irq_wake(IRQ_TOUCH_INT);
 	disable_irq(home_key_irq);
 	gpio_direction_output(_3_GPIO_TOUCH_EN, 0);
 #ifdef I2C_BIT_BANG
@@ -317,6 +318,7 @@
 		       gpio_pend_mask_mem);
 #endif
 
+	enable_irq_wake(IRQ_TOUCH_INT);
 	enable_irq(IRQ_TOUCH_INT);
 	enable_irq(home_key_irq);
 	touchkey_enable = 1;
@@ -397,6 +399,8 @@
 		return -EBUSY;
 	}
 
+	/* Let the touch key gpio int has wakeup capable enabled */
+	enable_irq_wake(IRQ_TOUCH_INT);
 	home_key_irq = gpio_to_irq(HOME_KEY_GPIO);
 	if (request_irq(home_key_irq, home_key_interrupt, IRQF_DISABLED, DEVICE_NAME, NULL)) {
 		printk("[CYPRESS] Can't allocate irq for HOME key\n");
Index: android-2.6.35_10.1/drivers/input/touchscreen/atmel_touchscreen.c
===================================================================
--- android-2.6.35_10.1.orig/drivers/input/touchscreen/atmel_touchscreen.c	2010-12-30 11:50:50.937457167 +0530
+++ android-2.6.35_10.1/drivers/input/touchscreen/atmel_touchscreen.c	2010-12-30 12:16:07.718712725 +0530
@@ -1212,6 +1212,7 @@
 static int touchscreen_suspend(struct platform_device *pdev, pm_message_t state) {
 	dbg("touchscreen_suspend : touch power off\n");
 	disable_irq_nosync(tsp.irq);
+	disable_irq_wake(tsp.irq);
 	atmel_suspend();
 	initialize_multi_touch();
 	return 0;
@@ -1232,6 +1233,7 @@
 #endif
 	atmel_resume();
 	initialize_multi_touch();
+	enable_irq_wake(tsp.irq);
 	enable_irq(tsp.irq);
 	dbg("touchscreen_resume end\n");
 	return 0;
