Index: android-2.6.35/arch/arm/mach-omap2/clock44xx_data.c
===================================================================
--- android-2.6.35.orig/arch/arm/mach-omap2/clock44xx_data.c	2010-12-21 18:24:18.000000000 +0530
+++ android-2.6.35/arch/arm/mach-omap2/clock44xx_data.c	2010-12-30 19:16:17.747463078 +0530
@@ -885,8 +885,8 @@
 
 static struct clk dpll_per_m5x2_ck = {
 	.name		= "dpll_per_m5x2_ck",
-	.parent		= &dpll_per_x2_ck,
-	.clksel		= dpll_per_m2x2_div,
+	.parent		= &dpll_per_ck,
+	.clksel		= dpll_per_m2_div,
 	.clksel_reg	= OMAP4430_CM_DIV_M5_DPLL_PER,
 	.clksel_mask	= OMAP4430_HSDIVIDER_CLKOUT2_DIV_MASK,
 	.ops		= &clkops_null,
