Index: android-2.6.35/arch/arm/mach-omap2/board-omap4tab10-1.c
===================================================================
--- android-2.6.35.orig/arch/arm/mach-omap2/board-omap4tab10-1.c	2010-12-31 18:09:26.480221051 +0530
+++ android-2.6.35/arch/arm/mach-omap2/board-omap4tab10-1.c	2010-12-31 18:20:03.418965017 +0530
@@ -209,8 +209,8 @@
         KEY(3, 3, KEY_VOLUMEDOWN),      //volume down
 #endif
 #if (CONFIG_SAMSUNG_OMAP4_TAB_REV == 4 || CONFIG_SAMSUNG_OMAP4_TAB_REV == 5)
-        KEY(5, 4, KEY_VOLUMEUP),        //Volume up
-        KEY(4, 4, KEY_VOLUMEDOWN),      //volume down
+        KEY(2, 1, KEY_VOLUMEUP),        //Volume up
+        KEY(1, 1, KEY_VOLUMEDOWN),      //volume down
 #endif
 	0,
 };
@@ -285,6 +285,22 @@
 };
 
 
+void keyboard_mux_init(void)
+{
+#if (CONFIG_SAMSUNG_OMAP4_TAB_REV == 4 || CONFIG_SAMSUNG_OMAP4_TAB_REV == 5)
+        omap_mux_init_signal("kpd_col1.kpd_col1",
+                                OMAP_WAKEUP_EN | OMAP_MUX_MODE0);
+        omap_mux_init_signal("kpd_row1.kpd_row1",
+                                OMAP_PULL_ENA | OMAP_PULL_UP |
+                                OMAP_WAKEUP_EN | OMAP_MUX_MODE0 |
+                                OMAP_INPUT_EN);
+        omap_mux_init_signal("kpd_row2.kpd_row2",
+                                OMAP_PULL_ENA | OMAP_PULL_UP |
+                                OMAP_WAKEUP_EN | OMAP_MUX_MODE0 |
+                                OMAP_INPUT_EN);
+#endif
+}
+
 
 static int omap4tab_panel_enable_hdmi(struct omap_dss_device *dssdev)
 {
