Index: Kernel/arch/arm/mach-omap2/board-omap4tab10-1.c
===================================================================
--- Kernel.orig/arch/arm/mach-omap2/board-omap4tab10-1.c	2011-01-04 21:18:19.056592080 +0530
+++ Kernel/arch/arm/mach-omap2/board-omap4tab10-1.c	2011-01-04 21:22:59.076394749 +0530
@@ -1788,6 +1788,16 @@
 	padconf = omap_readw(CONTROL_CORE_PAD1_GPMC_AD19);
 	padconf |= OMAP44XX_PADCONF_WAKEUPENABLE1;
 	omap_writew(padconf, CONTROL_CORE_PAD1_GPMC_AD19);
+
+#if (CONFIG_SAMSUNG_OMAP4_TAB_REV == 5)
+	/*
+	 * Enable IO daisy for sys_nirq1/2, to be able to
+	 * wakeup from interrupts from PMIC/Audio IC.
+	 * Needed only in Device OFF mode.
+	 */
+	omap_mux_enable_wakeup("sys_nirq1");
+	omap_mux_enable_wakeup("sys_nirq2");
+#endif
 }
 
 
@@ -1797,6 +1807,7 @@
 	.step_size = 12500,
 	.i2c_addr = 0x12,
 	.i2c_vreg = 0x61,
+	.i2c_cmdreg = 0x62,
 	.vsel_to_uv = omap_twl_vsel_to_uv,
 	.uv_to_vsel = omap_twl_uv_to_vsel,
 	.onforce_cmd = omap_twl_onforce_cmd,
@@ -1817,6 +1828,7 @@
 	.step_size = 12500,
 	.i2c_addr = 0x12,
 	.i2c_vreg = 0x55,
+	.i2c_cmdreg = 0x56,
 	.vsel_to_uv = omap_twl_vsel_to_uv,
 	.uv_to_vsel = omap_twl_uv_to_vsel,
 	.onforce_cmd = omap_twl_onforce_cmd,
@@ -1837,6 +1849,7 @@
 	.step_size = 12500,
 	.i2c_addr = 0x12,
 	.i2c_vreg = 0x5b,
+	.i2c_cmdreg = 0x5c,
 	.vsel_to_uv = omap_twl_vsel_to_uv,
 	.uv_to_vsel = omap_twl_uv_to_vsel,
 	.onforce_cmd = omap_twl_onforce_cmd,
@@ -1855,15 +1868,27 @@
 	.vdd0_on = 1350000,        /* 1.35v */
 	.vdd0_onlp = 1350000,      /* 1.35v */
 	.vdd0_ret = 837500,       /* 0.8375v */
+#if (CONFIG_SAMSUNG_OMAP4_TAB_REV == 5)
+	.vdd0_off = 0,          /* 0 v */
+#else
 	.vdd0_off = 600000,       /* 0.6v */
+#endif
 	.vdd1_on = 1100000,        /* 1.1v */
 	.vdd1_onlp = 1100000,      /* 1.1v */
 	.vdd1_ret = 837500,       /* 0.8375v */
+#if (CONFIG_SAMSUNG_OMAP4_TAB_REV == 5)
+	.vdd1_off = 0,		/* 0 v */
+#else
 	.vdd1_off = 600000,       /* 0.6v */
+#endif
 	.vdd2_on = 1100000,        /* 1.1v */
 	.vdd2_onlp = 1100000,      /* 1.1v */
 	.vdd2_ret = 837500,       /* .8375v */
+#if (CONFIG_SAMSUNG_OMAP4_TAB_REV == 5)
+	.vdd2_off = 0,		/* 0 v */
+#else
 	.vdd2_off = 600000,       /* 0.6v */
+#endif
 };
 
 
