Index: Kernel/arch/arm/configs/android_omap4tab_rev05_defconfig
===================================================================
--- Kernel.orig/arch/arm/configs/android_omap4tab_rev05_defconfig	2011-01-04 21:59:42.252481048 +0530
+++ Kernel/arch/arm/configs/android_omap4tab_rev05_defconfig	2011-01-04 22:02:03.303737457 +0530
@@ -813,7 +813,7 @@
 # CONFIG_WL127X_RFKILL is not set
 # CONFIG_APANIC is not set
 # CONFIG_LP530AL_GYRO_SENSOR is not set
-CONFIG_PN554_NFC=y
+CONFIG_PN544_NFC=y
 # CONFIG_C2PORT is not set
 
 #
Index: Kernel/arch/arm/mach-omap2/board-omap4tab10-1.c
===================================================================
--- Kernel.orig/arch/arm/mach-omap2/board-omap4tab10-1.c	2011-01-04 22:00:14.742484881 +0530
+++ Kernel/arch/arm/mach-omap2/board-omap4tab10-1.c	2011-01-05 20:31:51.000000000 +0530
@@ -64,6 +64,10 @@
 #include <plat/opp_twl_tps.h>
 #include <linux/fsa9480.h>
 
+#ifdef CONFIG_PN544_NFC
+#include<linux/pn544.h>
+#endif /* For NFC Support */
+
 #include <linux/atmel_mxt1386.h> /* For mXT1386 Touch */
 #include "mux.h"
 
@@ -155,6 +159,13 @@
 #define OMAP4430_GPIO_MIPI_HSI_PHONE_ACTIVE			120
 #endif  //CONFIG_SAMSUNG_PHONE_SVNET
 
+/*PN544 GPIOs*/
+#if (CONFIG_SAMSUNG_OMAP4_TAB_REV == 5)
+#define OMAP4430_GPIO_NFC_IRQ		59
+#define OMAP4430_GPIO_NFC_FIRM		135
+#define OMAP4430_GPIO_NFC_VEN		136
+#endif
+
 static void omap4_usb_cb(u8 attached);
 static void omap4_charger_cb(u8 attached);
 static void omap4_jig_cb(u8 attached);
@@ -1391,7 +1402,13 @@
 	},
 
 };
-
+#if (CONFIG_SAMSUNG_OMAP4_TAB_REV == 5)
+static struct pn544_i2c_platform_data pn544_data = {
+         .irq_gpio = OMAP4430_GPIO_NFC_IRQ,
+         .ven_gpio = OMAP4430_GPIO_NFC_VEN,
+         .firm_gpio = OMAP4430_GPIO_NFC_FIRM,
+};
+#endif
 static struct i2c_board_info __initdata omap4tab_i2c_3_boardinfo[] = {
 #if (CONFIG_SAMSUNG_OMAP4_TAB_REV == 4 || CONFIG_SAMSUNG_OMAP4_TAB_REV == 5)
 	{
@@ -1456,6 +1473,11 @@
                .platform_data = &omap4_fsa9480_pdata,
                .irq = OMAP_GPIO_IRQ(OMAP4430_GPIO_JACK_NINT),
         },
+	{
+               I2C_BOARD_INFO("pn544",0x2b),
+              .platform_data = &pn544_data,
+               .irq = OMAP_GPIO_IRQ(OMAP4430_GPIO_NFC_IRQ),
+        },
 #endif
 
 };
Index: Kernel/drivers/misc/Kconfig
===================================================================
--- Kernel.orig/drivers/misc/Kconfig	2011-01-04 21:58:12.582484759 +0530
+++ Kernel/drivers/misc/Kconfig	2011-01-04 22:00:47.226638815 +0530
@@ -437,7 +437,7 @@
 	default n
 	---help---
 	  Provides support for reading LP530AL GYRO SENSOR values through TWL6030 GPADC.
-config PN554_NFC
+config PN544_NFC
         bool "NXP PN554 NFC Controller Driver"
         default n
         help
Index: Kernel/drivers/misc/Makefile
===================================================================
--- Kernel.orig/drivers/misc/Makefile	2011-01-04 21:58:22.312484464 +0530
+++ Kernel/drivers/misc/Makefile	2011-01-04 22:01:04.814359560 +0530
@@ -42,5 +42,5 @@
 obj-$(CONFIG_TWL6030_POWER_OFF) += twl6030-pwroff.o
 obj-$(CONFIG_LP530AL_GYRO_SENSOR)	+= gyro_lp530al.o
 obj-$(CONFIG_TWL6040_VIB)	+= twl6040-vib.o
-obj-$(CONFIG_PN554_NFC)         += pn544.o
+obj-$(CONFIG_PN544_NFC)         += pn544.o
 
Index: Kernel/drivers/misc/pn544.c
===================================================================
--- Kernel.orig/drivers/misc/pn544.c	2011-01-05 20:24:16.573734484 +0530
+++ Kernel/drivers/misc/pn544.c	2011-01-05 20:24:46.000000000 +0530
@@ -180,18 +180,18 @@
 	case PN544_SET_PWR:
 		if (arg > 1) {  /* enable with firmware download */
 			pr_info("%s power on with firmware\n", __func__);
-			gpio_set_value(pn544_dev->firm_gpio, 1);
-			gpio_set_value(pn544_dev->ven_gpio, 1);
+			gpio_direction_output(pn544_dev->firm_gpio, 1);
+			gpio_direction_output(pn544_dev->ven_gpio, 1);
 			msleep(3);
 		} else if (arg == 1) {  /* enable */
 			pr_info("%s power on\n", __func__);
-			gpio_set_value(pn544_dev->firm_gpio, 0);
-			gpio_set_value(pn544_dev->ven_gpio, 1);
+			gpio_direction_output(pn544_dev->firm_gpio, 0);
+			gpio_direction_output(pn544_dev->ven_gpio, 1);
 			msleep(3);
 		} else {
 			pr_info("%s power off\n", __func__);
-			gpio_set_value(pn544_dev->firm_gpio, 0);
-			gpio_set_value(pn544_dev->ven_gpio, 0);
+			gpio_direction_output(pn544_dev->firm_gpio, 0);
+			gpio_direction_output(pn544_dev->ven_gpio, 0);
 			msleep(7);
 		}
 		break;
