Index: Kernel/arch/arm/mach-omap2/board-omap4tab10-1.c
===================================================================
--- Kernel.orig/arch/arm/mach-omap2/board-omap4tab10-1.c	2011-01-06 03:56:50.824352184 +0530
+++ Kernel/arch/arm/mach-omap2/board-omap4tab10-1.c	2011-01-06 04:06:42.242484575 +0530
@@ -1414,7 +1414,12 @@
 #if (CONFIG_SAMSUNG_OMAP4_TAB_REV == 4 || CONFIG_SAMSUNG_OMAP4_TAB_REV == 5)
 	{
 		I2C_BOARD_INFO("cypress_touchkey", 0x20),
-	}
+	},
+#endif
+#if (CONFIG_SAMSUNG_OMAP4_TAB_REV == 5)
+	{
+		I2C_BOARD_INFO("max17040", 0x36),
+ 	}
 #endif
 #if (CONFIG_SAMSUNG_OMAP4_TAB_REV == 3)
 	{
Index: Kernel/drivers/power/twl6030_bci_battery.c
===================================================================
--- Kernel.orig/drivers/power/twl6030_bci_battery.c	2011-01-06 03:56:50.824352184 +0530
+++ Kernel/drivers/power/twl6030_bci_battery.c	2011-01-06 14:37:10.862476951 +0530
@@ -257,6 +257,9 @@
 };
 static struct blocking_notifier_head notifier_list;
 extern u32 wakeup_timer_seconds;
+#if (CONFIG_SAMSUNG_OMAP4_TAB_REV == 5)
+	extern int read_max17040_vcell(void);
+#endif
 
 static void twl6030_config_min_vbus_reg(struct twl6030_bci_device_info *di,
 						unsigned int value)
@@ -852,7 +855,11 @@
 		twl6030_stop_charger(di);
 		/*voltage setteling time*/
 		msleep(200);
-		di->voltage_uV = twl6030_get_gpadc_conversion(7);
+		#if (CONFIG_SAMSUNG_OMAP4_TAB_REV == 5)
+			di->voltage_uV = read_max17040_vcell();
+		#else
+			di->voltage_uV = twl6030_get_gpadc_conversion(7);
+		#endif
 
 	}
 
@@ -919,8 +926,13 @@
 		return;
 	}
 
-	if (req.rbuf[7] > 0)
-		di->voltage_uV = req.rbuf[7];
+	#if (CONFIG_SAMSUNG_OMAP4_TAB_REV == 5)
+		di->voltage_uV = read_max17040_vcell();
+	#else
+ 		if (req.rbuf[7] > 0)
+ 			di->voltage_uV = req.rbuf[7];
+	#endif
+
 	if (req.rbuf[8] > 0)
 		di->bk_voltage_uV = req.rbuf[8];
 
@@ -1057,7 +1069,11 @@
 		val->intval = di->charge_status;
 		break;
 	case POWER_SUPPLY_PROP_VOLTAGE_NOW:
-		di->voltage_uV = twl6030_get_gpadc_conversion(7);
+		#if (CONFIG_SAMSUNG_OMAP4_TAB_REV == 5)
+			di->voltage_uV = read_max17040_vcell();
+		#else
+			di->voltage_uV = twl6030_get_gpadc_conversion(7);
+		#endif
 		val->intval = di->voltage_uV * 1000;
 		break;
 	case POWER_SUPPLY_PROP_CURRENT_NOW:
@@ -1745,7 +1761,11 @@
 	di->charger_incurrentmA = di->max_charger_currentmA;
 	di->charger_outcurrentmA = di->max_charger_currentmA;
 	di->watchdog_duration = 32;
-	di->voltage_uV = twl6030_get_gpadc_conversion(7);
+	#if (CONFIG_SAMSUNG_OMAP4_TAB_REV == 5)
+		di->voltage_uV = read_max17040_vcell();
+	#else
+		di->voltage_uV = twl6030_get_gpadc_conversion(7);
+	#endif
 	dev_info(&pdev->dev, "Battery Voltage at Bootup is %d mV\n",
 							di->voltage_uV);
 
Index: Kernel/arch/arm/configs/android_omap4tab_rev05_defconfig
===================================================================
--- Kernel.orig/arch/arm/configs/android_omap4tab_rev05_defconfig	2011-01-06 03:56:50.824352184 +0530
+++ Kernel/arch/arm/configs/android_omap4tab_rev05_defconfig	2011-01-06 04:06:42.242484575 +0530
@@ -1131,7 +1131,7 @@
 CONFIG_TWL6030_BCI_BATTERY=y
 # CONFIG_CHARGER_BQ2415x is not set
 # CONFIG_BATTERY_BQ27x00 is not set
-# CONFIG_BATTERY_MAX17040 is not set
+CONFIG_BATTERY_MAX17040=y
 # CONFIG_BATTERY_MONITOR is not set
 CONFIG_HWMON=y
 # CONFIG_HWMON_VID is not set
Index: Kernel/drivers/power/max17040_battery.c
===================================================================
--- Kernel.orig/drivers/power/max17040_battery.c	2011-01-08 16:35:00.000000000 +0530
+++ Kernel/drivers/power/max17040_battery.c	2011-01-08 16:38:45.237490143 +0530
@@ -53,6 +53,8 @@
 	int status;
 };
 
+struct i2c_client *max17040_client;
+
 static int max17040_get_property(struct power_supply *psy,
 			    enum power_supply_property psp,
 			    union power_supply_propval *val)
@@ -121,6 +123,22 @@
 	chip->vcell = (msb << 4) + (lsb >> 4);
 }
 
+int read_max17040_vcell(void)
+{
+	struct max17040_chip *chip = i2c_get_clientdata(max17040_client);
+	int result = 0;
+	u8 msb;
+	u8 lsb;
+
+	msb = max17040_read_reg(max17040_client, MAX17040_VCELL_MSB);
+	lsb = max17040_read_reg(max17040_client, MAX17040_VCELL_LSB);
+
+	chip->vcell = (msb << 4) + (lsb >> 4);
+
+	result = chip->vcell;
+	return result;
+}
+
 static void max17040_get_soc(struct i2c_client *client)
 {
 	struct max17040_chip *chip = i2c_get_clientdata(client);
@@ -212,6 +230,7 @@
 		return -ENOMEM;
 
 	chip->client = client;
+	max17040_client = client;
 	chip->pdata = client->dev.platform_data;
 
 	i2c_set_clientdata(client, chip);
@@ -232,8 +251,10 @@
 	max17040_reset(client);
 	max17040_get_version(client);
 
-	INIT_DELAYED_WORK_DEFERRABLE(&chip->work, max17040_work);
-	schedule_delayed_work(&chip->work, MAX17040_DELAY);
+	/* Disabled this work queue because we are reading the updated
+	   voltage value from TWL6030 battery driver */
+	//INIT_DELAYED_WORK_DEFERRABLE(&chip->work, max17040_work);
+	//schedule_delayed_work(&chip->work, MAX17040_DELAY);
 
 	return 0;
 }
@@ -255,7 +276,7 @@
 {
 	struct max17040_chip *chip = i2c_get_clientdata(client);
 
-	cancel_delayed_work(&chip->work);
+	//cancel_delayed_work(&chip->work);
 	return 0;
 }
 
@@ -263,7 +284,7 @@
 {
 	struct max17040_chip *chip = i2c_get_clientdata(client);
 
-	schedule_delayed_work(&chip->work, MAX17040_DELAY);
+	//schedule_delayed_work(&chip->work, MAX17040_DELAY);
 	return 0;
 }
 
