Index: Kernel/drivers/gpu/pvr/omaplfb/omaplfb_displayclass.c
===================================================================
--- Kernel.orig/drivers/gpu/pvr/omaplfb/omaplfb_displayclass.c	2011-01-06 10:53:10.078157495 +0530
+++ Kernel/drivers/gpu/pvr/omaplfb/omaplfb_displayclass.c	2011-01-06 14:38:05.706908332 +0530
@@ -1443,16 +1443,6 @@
 	psPVRFBInfo->ulFBSize = FBSize;
 	psPVRFBInfo->ulBufferSize =
 		psPVRFBInfo->ulHeight * psPVRFBInfo->ulByteStride;
-	/* Get physical display size for DPI calculation */
-	if (psLINFBInfo->var.width < 0 || psLINFBInfo->var.height < 0) {
-		psDevInfo->sDisplayInfo.ui32PhysicalWidthmm = 0;
-		psDevInfo->sDisplayInfo.ui32PhysicalHeightmm = 0;
-	} else {
-		psDevInfo->sDisplayInfo.ui32PhysicalWidthmm =
-			psLINFBInfo->var.width;
-		psDevInfo->sDisplayInfo.ui32PhysicalHeightmm =
-			psLINFBInfo->var.height;
-	}
 
 	/* XXX: Page aligning with 16bpp causes the
 	 * position of framebuffer address to look in the wrong place.
