Index: Kernel-p1/arch/arm/mach-omap2/id.c
===================================================================
--- Kernel-p1.orig/arch/arm/mach-omap2/id.c	2011-01-24 11:55:54.000000000 +0530
+++ Kernel-p1/arch/arm/mach-omap2/id.c	2011-01-24 16:16:52.167780662 +0530
@@ -14,6 +14,8 @@
  * published by the Free Software Foundation.
  */
 
+
+
 #include <linux/module.h>
 #include <linux/kernel.h>
 #include <linux/init.h>
