Index: Kernel/arch/arm/mach-omap2/board-omap4tab10-1.c
===================================================================
--- Kernel.orig/arch/arm/mach-omap2/board-omap4tab10-1.c	2011-01-17 15:28:58.000000000 +0530
+++ Kernel/arch/arm/mach-omap2/board-omap4tab10-1.c	2011-01-17 16:59:09.802682850 +0530
@@ -120,6 +120,8 @@
 #define OMAP4TAB_KXSD9_INT_GPIO				122
 #define OMAP4TAB_AKM_INT_GPIO				157
 #define OMAP4TAB_MPU3050_INT_GPIO			45
+#define OMAP4TAB_GP2A_PS_ON				35
+#define OMAP4TAB_GP2A_PS_VOUT				33
 #elif (CONFIG_SAMSUNG_OMAP4_TAB_REV == 5)
 #define OMAP4TAB_KXSD9_INT_GPIO				122
 #define OMAP4TAB_AKM_INT_GPIO				157
