Index: Kernel/arch/arm/mach-omap2/board-omap4tab10-1.c
===================================================================
--- Kernel.orig/arch/arm/mach-omap2/board-omap4tab10-1.c	2011-01-17 17:06:03.000000000 +0530
+++ Kernel/arch/arm/mach-omap2/board-omap4tab10-1.c	2011-01-17 17:51:26.632057551 +0530
@@ -67,7 +67,7 @@
 
 #include <linux/i2c/twl6030-gpadc.h>
 
-#ifdef CONFIG_PN544_NFC
+#if (CONFIG_SAMSUNG_OMAP4_TAB_REV == 5)
 #include<linux/pn544.h>
 #endif /* For NFC Support */
 
