Index: Kernel/drivers/input/touchscreen/atmel_touch.h
===================================================================
--- Kernel.orig/drivers/input/touchscreen/atmel_touch.h	2011-01-20 19:04:23.712243681 +0530
+++ Kernel/drivers/input/touchscreen/atmel_touch.h	2011-01-20 19:04:42.149755012 +0530
@@ -52,7 +52,7 @@
 
 /* This feature is about making qt chip enter sleep mode by gating power of Driver IC,
  * not sending I2C sleep command. */
-//#define FEATURE_SUSPEND_BY_DISABLING_POWER
+#define FEATURE_SUSPEND_BY_DISABLING_POWER
 
 //#define FEATURE_LOGGING_TOUCH_EVENT
 
Index: Kernel/drivers/input/touchscreen/atmel_touch_602240.c
===================================================================
--- Kernel.orig/drivers/input/touchscreen/atmel_touch_602240.c	2011-01-20 19:07:09.289748433 +0530
+++ Kernel/drivers/input/touchscreen/atmel_touch_602240.c	2011-01-20 19:08:34.219746242 +0530
@@ -877,7 +877,7 @@
 #endif
 		dbg("suspend by power\n");
 		disable_irq(touchscreen_get_tsp_int_num());
-		change_pins_of_touch_at_suspend();
+		//change_pins_of_touch_at_suspend();
 		gpio_direction_output(OMAP_GPIO_TSP_EN, 0);
 		gpio_set_value(OMAP_GPIO_TSP_EN, 0);  // power down Driver ICy
 		config_normal.gripfacesuppression_config.xhigrip = 5;
@@ -968,7 +968,7 @@
 	} else {
 #endif
 		dbg("resume by power\n");
-		change_pins_of_touch_at_resume();
+		//change_pins_of_touch_at_resume();
 		gpio_set_value(OMAP_GPIO_TSP_EN, 1);
 
 		mdelay(80); // recommended value
@@ -983,10 +983,10 @@
 		calibrate_chip();
 		while (resume_success != 1 && repeat_count++ <RETRY_COUNT) {
 			if ( do_reset ) {
-				change_pins_of_touch_at_suspend();
+				//change_pins_of_touch_at_suspend();
 				gpio_set_value(OMAP_GPIO_TSP_EN, 0);  // TOUCH EN
 				mdelay(200);
-				change_pins_of_touch_at_resume();
+				//change_pins_of_touch_at_resume();
 				gpio_set_value(OMAP_GPIO_TSP_EN, 1);  // TOUCH EN
 				mdelay(80); // recommended value
 				calibrate_chip();
