Index: Kernel/drivers/video/omap2/dss/dss.c
===================================================================
--- Kernel.orig/drivers/video/omap2/dss/dss.c	2011-01-21 02:51:13.722696181 +0530
+++ Kernel/drivers/video/omap2/dss/dss.c	2011-01-21 18:08:13.992695972 +0530
@@ -506,16 +506,8 @@
 	fck = dss_clk_get_rate(DSS_CLK_FCK1);
 	if (!cpu_is_omap44xx()) {
 		prate = dss_get_dpll4_rate();
-	} else {
-		/* Linux ARM clk framework issue currently all DPLL lock frequency
-		 * multipliers are not configured properly. Hence multiplying by
-		 * 2 to get the right pixel clock dividers.
-		 *
-		 * TODO: Need to change this once appropriate patches are added
-		 * to clock framework.
-		 */
-		fck *= 2;
 	}
+
 	if (req_pck == dss.cache_req_pck &&
 			((cpu_is_omap34xx() && prate == dss.cache_prate) ||
 			 dss.cache_dss_cinfo.fck == fck)) {
Index: Kernel/arch/arm/mach-omap2/clock44xx_data.c
===================================================================
--- Kernel.orig/arch/arm/mach-omap2/clock44xx_data.c	2011-01-21 02:48:34.822696785 +0530
+++ Kernel/arch/arm/mach-omap2/clock44xx_data.c	2011-01-21 18:08:14.073321350 +0530
@@ -885,8 +885,8 @@
 
 static struct clk dpll_per_m5x2_ck = {
 	.name		= "dpll_per_m5x2_ck",
-	.parent		= &dpll_per_ck,
-	.clksel		= dpll_per_m2_div,
+	.parent		= &dpll_per_x2_ck,
+	.clksel		= dpll_per_m2x2_div,
 	.clksel_reg	= OMAP4430_CM_DIV_M5_DPLL_PER,
 	.clksel_mask	= OMAP4430_HSDIVIDER_CLKOUT2_DIV_MASK,
 	.ops		= &clkops_null,
