Index: Kernel/arch/arm/mach-omap2/board-omap4tab10-1.c
===================================================================
--- Kernel.orig/arch/arm/mach-omap2/board-omap4tab10-1.c	2011-01-21 12:32:33.199742054 +0530
+++ Kernel/arch/arm/mach-omap2/board-omap4tab10-1.c	2011-01-21 12:32:53.212250146 +0530
@@ -1488,7 +1488,7 @@
         },
         {
                 I2C_BOARD_INFO("gp2a",0x44),
-
+		.platform_data = &gp2a_pdata,
         },
         {
                I2C_BOARD_INFO("fsa9480", (0x4A >> 1)),
Index: Kernel/drivers/sensors/optical/gp2a.c
===================================================================
--- Kernel.orig/drivers/sensors/optical/gp2a.c	2011-01-21 12:32:33.259749053 +0530
+++ Kernel/drivers/sensors/optical/gp2a.c	2011-01-21 12:32:53.219741880 +0530
@@ -34,7 +34,11 @@
 #include <linux/uaccess.h>
 #include <linux/gp2a.h>
 
+#if (CONFIG_SAMSUNG_OMAP4_TAB_REV == 4 )
+#define OMAP4TAB_GP2A_PS_ON				35
+#elif (CONFIG_SAMSUNG_OMAP4_TAB_REV == 5 )
 #define OMAP4TAB_GP2A_PS_ON				37
+#endif
 
 /* Note about power vs enable/disable:
  *  The chip has two functions, proximity and ambient light sensing.
