Index: Kernel/arch/arm/mach-omap2/board-omap4tab10-1.c
===================================================================
--- Kernel.orig/arch/arm/mach-omap2/board-omap4tab10-1.c	2011-01-21 19:18:04.000000000 +0530
+++ Kernel/arch/arm/mach-omap2/board-omap4tab10-1.c	2011-01-21 19:35:35.179738478 +0530
@@ -81,7 +81,11 @@
 #endif // CONFIG_SAMSUNG_PHONE_SVNET
 
 #define _OMAP_MUX_SETTING
-
+#if (CONFIG_SAMSUNG_OMAP4_TAB_REV == 5)
+#define DISABLE_I2C2_HW_SPINLOCK_FOR_REV5
+#else
+#undef DISABLE_I2C2_HW_SPINLOCK_FOR_REV5
+#endif
 #if (CONFIG_SAMSUNG_OMAP4_TAB_REV <= 2)
 #define SDP4430_LCD_PANEL_BACKLIGHT_ENABLE_GPIO		16
 #define SDP4430_MLCD1_PANEL_ENABLE_GPIO			2
@@ -1396,7 +1400,7 @@
 		.platform_data = &omap4tab_twldata,
 	},
 };
-
+#ifndef DISABLE_I2C2_HW_SPINLOCK_FOR_REV5
 static struct i2c_board_info __initdata omap4tab_i2c_2_boardinfo[] = {
 #if (CONFIG_SAMSUNG_OMAP4_TAB_REV <= 2)
 	{
@@ -1414,6 +1418,7 @@
 		I2C_BOARD_INFO("M5MO", 0x3E >> 1),
 	}
 };
+#endif
 #if (CONFIG_SAMSUNG_OMAP4_TAB_REV == 5)
 static struct pn544_i2c_platform_data pn544_data = {
          .irq_gpio = OMAP4430_GPIO_NFC_IRQ,
@@ -1575,7 +1580,9 @@
 static int __init omap4_i2c_init(void)
 {
 	omap_i2c_hwspinlock_init(1, 0, &omap4tab_i2c_bus_pdata);
+#ifndef DISABLE_I2C2_HW_SPINLOCK_FOR_REV5
 	omap_i2c_hwspinlock_init(2, 1, &omap4tab_i2c_2_bus_pdata);
+#endif
 	omap_i2c_hwspinlock_init(3, 2, &omap4tab_i2c_3_bus_pdata);
 	omap_i2c_hwspinlock_init(4, 3, &omap4tab_i2c_4_bus_pdata);
 	/*
@@ -1584,8 +1591,10 @@
 	 */
 	omap_register_i2c_bus(1, 400, &omap4tab_i2c_bus_pdata,
 		omap4tab_i2c_boardinfo, ARRAY_SIZE(omap4tab_i2c_boardinfo));
+#ifndef DISABLE_I2C2_HW_SPINLOCK_FOR_REV5
 	omap_register_i2c_bus(2, 400, &omap4tab_i2c_2_bus_pdata,
 		omap4tab_i2c_2_boardinfo, ARRAY_SIZE(omap4tab_i2c_2_boardinfo));
+#endif
 	omap_register_i2c_bus(3, 400, &omap4tab_i2c_3_bus_pdata,
 		omap4tab_i2c_3_boardinfo, ARRAY_SIZE(omap4tab_i2c_3_boardinfo));
 	omap_register_i2c_bus(4, 400, &omap4tab_i2c_4_bus_pdata,
@@ -2052,7 +2061,32 @@
 #else
 #define board_mux       NULL
 #endif
+#ifdef DISABLE_I2C2_HW_SPINLOCK_FOR_REV5
+void m5mo_fref_start(void)
+{
+    volatile u32 val=0,val2=0;
+    void __iomem *frefclk1_base=NULL, *frefclk2_base=NULL;
 
+    val = 0;
+
+    frefclk1_base = ioremap(0x4A30A314, 4);
+    val = __raw_readl(frefclk1_base);
+    val = val | (1<<8) | (0x01<<2) | (0xF<<16);
+    __raw_writel(val, (frefclk1_base));
+    val = __raw_readl(frefclk1_base);
+    printk(KERN_ERR"\n\t [MBG043 MCLK] AUXCLK1 is %x \n\n",val);
+    iounmap(frefclk1_base);
+      
+	val2 = 0;
+
+    frefclk2_base = ioremap(0x4A30A318, 4);
+    val2 = __raw_readl(frefclk2_base);
+    val2 = val2 | (1<<8) | (0x01<<2) | (0xF<<16);
+    __raw_writel(val2, (frefclk2_base));
+    val2 = __raw_readl(frefclk2_base);
+    printk(KERN_ERR"\n\t [S5KAAFX MCLK] AUXCLK2 is %x \n\n",val2);
+}
+#endif
 
 static void __init omap4_tab_10_1_init(void)
 {
@@ -2109,7 +2143,9 @@
 #if (CONFIG_SAMSUNG_OMAP4_TAB_REV == 3)
 	samsung_omap4_battery_init();
 #endif
-
+#ifdef DISABLE_I2C2_HW_SPINLOCK_FOR_REV5
+	m5mo_fref_start();
+#endif
 }
 
 static void __init omap4_tab_10_1_map_io(void)
