Index: Kernel_907838/drivers/input/keyboard/cypress/cypress-touchkey.c
===================================================================
--- Kernel_907838.orig/drivers/input/keyboard/cypress/cypress-touchkey.c	2011-02-14 16:20:11.000000000 +0530
+++ Kernel_907838/drivers/input/keyboard/cypress/cypress-touchkey.c	2011-02-14 16:27:49.202510881 +0530
@@ -31,10 +31,12 @@
 
 #if (CONFIG_SAMSUNG_OMAP4_TAB_REV <= 4)
 #define _3_GPIO_TOUCH_EN	51
+#define _3_GPIO_TOUCH_LED_EN	51
 #define _3_GPIO_TOUCH_INT	32
 #define HOME_KEY_GPIO		100
 #elif (CONFIG_SAMSUNG_OMAP4_TAB_REV == 5)
 #define _3_GPIO_TOUCH_EN        101
+#define _3_GPIO_TOUCH_LED_EN	102
 #define _3_GPIO_TOUCH_INT       32
 #define HOME_KEY_GPIO           176
 #endif
@@ -195,6 +197,7 @@
 			retry = 10;
 			while (retry--) {
 				gpio_direction_output(_3_GPIO_TOUCH_EN, 0);
+				gpio_direction_output(_3_GPIO_TOUCH_LED_EN, 0);
 				mdelay(300);
 				init_hw();
 				if (i2c_touchkey_read(KEYCODE_REG, data, 3) >= 0) {
@@ -208,6 +211,7 @@
 			//enable_irq(IRQ_TOUCH_INT);
 			touchkey_enable = -1;
 			gpio_direction_output(_3_GPIO_TOUCH_EN, 0);
+			gpio_direction_output(_3_GPIO_TOUCH_LED_EN, 0);
 #ifdef I2C_BIT_BANG
 			gpio_direction_output(_3_TOUCH_SDA_28V, 0);
 			gpio_direction_output(_3_TOUCH_SCL_28V, 0);
@@ -300,6 +304,7 @@
 	disable_irq_wake(IRQ_TOUCH_INT);
 	disable_irq(home_key_irq);
 	gpio_direction_output(_3_GPIO_TOUCH_EN, 0);
+	gpio_direction_output(_3_GPIO_TOUCH_LED_EN, 0);
 #ifdef I2C_BIT_BANG
 	gpio_direction_output(_3_TOUCH_SDA_28V, 0);
 	gpio_direction_output(_3_TOUCH_SCL_28V, 0);
@@ -315,6 +320,7 @@
 		return;
 	}
 	gpio_direction_output(_3_GPIO_TOUCH_EN, 1);
+	gpio_direction_output(_3_GPIO_TOUCH_LED_EN, 1);
 	msleep(50);
 
 #if 0
@@ -425,6 +431,7 @@
 {
 	dbg("Enter %s\n", __FUNCTION__);
 	gpio_direction_output(_3_GPIO_TOUCH_EN, 1);
+	gpio_direction_output(_3_GPIO_TOUCH_LED_EN, 1);
 	msleep(200);
 	set_irq_type(IRQ_TOUCH_INT, IRQ_TYPE_LEVEL_LOW);
 	dbg("Exit %s\n", __FUNCTION__);
@@ -601,10 +608,12 @@
 	if (*buf == '0') {
 		disable_irq(IRQ_TOUCH_INT);
 		gpio_direction_output(_3_GPIO_TOUCH_EN, 0);
+		gpio_direction_output(_3_GPIO_TOUCH_LED_EN, 0);
 		touchkey_enable = -2;
 	} else if (*buf == '1') {
 		if (touchkey_enable == -2) {
 			gpio_direction_output(_3_GPIO_TOUCH_EN, 1);
+			gpio_direction_output(_3_GPIO_TOUCH_LED_EN, 1);
 			touchkey_enable = 1;
 			enable_irq(IRQ_TOUCH_INT);
 		}
@@ -650,6 +659,9 @@
 	if ((ret = gpio_request(_3_GPIO_TOUCH_EN, "_3_GPIO_TOUCH_EN")))
 		printk(KERN_ERR "Failed to request gpio %s:%d\n", __func__, __LINE__);
 
+	if ((ret = gpio_request(_3_GPIO_TOUCH_LED_EN, "_3_GPIO_TOUCH_LED_EN")))
+		printk(KERN_ERR "Failed to request gpio %s:%d\n", __func__, __LINE__);
+
 	ret = misc_register(&touchkey_update_device);
 	if (ret) {
 		printk("%s misc_register fail\n", __FUNCTION__);
@@ -707,6 +719,7 @@
 	if (touchkey_wq)
 		destroy_workqueue(touchkey_wq);
 	gpio_free(_3_GPIO_TOUCH_EN);
+	gpio_free(_3_GPIO_TOUCH_LED_EN);
 	dbg("Exit %s\n", __FUNCTION__);
 }
 
Index: Kernel_907838/drivers/input/keyboard/cypress/issp_defs.h
===================================================================
--- Kernel_907838.orig/drivers/input/keyboard/cypress/issp_defs.h	2011-02-14 16:20:11.000000000 +0530
+++ Kernel_907838/drivers/input/keyboard/cypress/issp_defs.h	2011-02-14 16:22:07.812511141 +0530
@@ -1,8 +1,15 @@
 // filename: ISSP_Defs.h
 #include "issp_revision.h"
 
+#if (CONFIG_SAMSUNG_OMAP4_TAB_REV <= 4)
 #define _3_GPIO_TOUCH_EN	51
+#define _3_GPIO_TOUCH_LED_EN	51
 #define _3_GPIO_TOUCH_INT	32
+#elif (CONFIG_SAMSUNG_OMAP4_TAB_REV == 5)
+#define _3_GPIO_TOUCH_EN	101
+#define _3_GPIO_TOUCH_LED_EN	102
+#define _3_GPIO_TOUCH_INT	32
+#endif
 
 #ifdef PROJECT_REV_304
 // Copyright 2006-2007, Cypress Semiconductor Corporation.
Index: Kernel_907838/drivers/input/keyboard/cypress/issp_driver_routines.c
===================================================================
--- Kernel_907838.orig/drivers/input/keyboard/cypress/issp_driver_routines.c	2011-02-14 16:20:11.000000000 +0530
+++ Kernel_907838/drivers/input/keyboard/cypress/issp_driver_routines.c	2011-02-14 16:28:53.322510681 +0530
@@ -460,6 +460,7 @@
 void ApplyTargetVDD(void)
 {
     gpio_direction_output(_3_GPIO_TOUCH_EN, 1);
+    gpio_direction_output(_3_GPIO_TOUCH_LED_EN, 1);
     mdelay(1);
 }
 
@@ -475,6 +476,7 @@
 void RemoveTargetVDD(void)
 {
     gpio_direction_output(_3_GPIO_TOUCH_EN, 0);
+    gpio_direction_output(_3_GPIO_TOUCH_LED_EN, 0);
 }
 #endif
 
Index: Kernel_907838/drivers/input/keyboard/cypress/issp_main.c
===================================================================
--- Kernel_907838.orig/drivers/input/keyboard/cypress/issp_main.c	2011-02-14 16:20:11.000000000 +0530
+++ Kernel_907838/drivers/input/keyboard/cypress/issp_main.c	2011-02-14 16:29:55.972510498 +0530
@@ -921,6 +921,7 @@
     //s3c_gpio_setpull(_3_TOUCH_SCL_28V, S3C_GPIO_PULL_NONE); 
     //s3c_gpio_setpull(_3_TOUCH_SDA_28V, S3C_GPIO_PULL_NONE); 
     gpio_direction_output(_3_GPIO_TOUCH_EN, 0);
+    gpio_direction_output(_3_GPIO_TOUCH_LED_EN, 0);
     msleep(1);
 #ifdef RESET_MODE
     gpio_tlmm_config(LED_26V_EN);
