Index: Kernel/drivers/video/omap2/dss/dss.c
===================================================================
--- Kernel.orig/drivers/video/omap2/dss/dss.c	2011-01-31 18:43:02.679756451 +0530
+++ Kernel/drivers/video/omap2/dss/dss.c	2011-02-02 18:27:58.569756782 +0530
@@ -724,16 +724,18 @@
 
 #ifdef CONFIG_FB_OMAP_BOOTLOADER_INIT
 	/* DISPC_CONTROL */
-	if (omap_readl(0x48050440) & 1)	/* LCD enabled? */
-		skip_init = true;
+		/* unable to read l3 address of DISPC_CONTROL2(0x58001238),
+		   So directly using the L4 addredd 0x48041238*/
+		if (omap_readl(0x48041238) & 1)	/* LCD enabled? */
+			skip_init = true;
+
 #endif
 
-	if (!skip_init) {
+	if (skip_init) {
 		/* disable LCD and DIGIT output. This seems to fix the synclost
 		 * problem that we get, if the bootloader starts the DSS and
 		 * the kernel resets it */
-		omap_writel(omap_readl(0x48050440) & ~0x3, 0x48050440);
-
+		omap_writel(omap_readl(0x48041238) & ~0x1, 0x48041238);
 		/* We need to wait here a bit, otherwise we sometimes start to
 		 * get synclost errors, and after that only power cycle will
 		 * restore DSS functionality. I have no idea why this happens.
@@ -746,7 +748,6 @@
 		if (cpu_is_omap44xx())
 			_omap_dss_reset();
 	}
-
 	/* autoidle */
 	REG_FLD_MOD(DSS_SYSCONFIG, 1, 0, 0);
 
