Index: Kernel/arch/arm/mach-omap2/board-omap4tab10-1.c
===================================================================
--- Kernel.orig/arch/arm/mach-omap2/board-omap4tab10-1.c	2011-02-01 15:52:02.000000000 +0530
+++ Kernel/arch/arm/mach-omap2/board-omap4tab10-1.c	2011-02-03 10:21:55.417780096 +0530
@@ -66,6 +66,7 @@
 #include <linux/gp2a.h>
 
 #include <linux/i2c/twl6030-gpadc.h>
+#include "board-4430sdp-wifi.h"
 
 #if (CONFIG_SAMSUNG_OMAP4_TAB_REV == 5)
 #include<linux/pn544.h>
@@ -1603,38 +1604,6 @@
 }
 
 
-#ifdef CONFIG_TIWLAN_SDIO
-static void pad_config(unsigned long pad_addr, u32 andmask, u32 ormask)
-{
-	int val;
-	 u32 *addr;
-
-	addr = (u32 *) ioremap(pad_addr, 4);
-	if (!addr) {
-		printk(KERN_ERR"OMAP_pad_config: ioremap failed with addr %lx\n",
-		pad_addr);
-	return;
-	}
-
-	val =  __raw_readl(addr);
-	val &= andmask;
-	val |= ormask;
-	__raw_writel(val, addr);
-
-	iounmap(addr);
-}
-
-void wlan_1283_config()
-{
-	pad_config(0x4A100078, 0xFFECFFFF, 0x00030000);
-	pad_config(0x4A10007C, 0xFFFFFFEF, 0x0000000B);
-	if (gpio_request(104, NULL) != 0)
-		printk(KERN_ERR "GPIO 104 request failed\n");
-	gpio_direction_output(104, 0);
-	return ;
-}
-#endif
-
 static u16 control_pbias_offset;
 
 void omap4tab_gpiowk_setup()
@@ -2110,7 +2079,7 @@
 	omap4_twl6030_hsmmc_init(mmc);
 
 #ifdef CONFIG_TIWLAN_SDIO
-	wlan_1283_config();
+	config_wlan_mux();
 #endif
 # if 0
 	usb_uhhtll_init(&usbhs_pdata);
