Index: android-2.6.35/arch/arm/mach-omap2/board-omap4tab10-1.c
===================================================================
--- android-2.6.35.orig/arch/arm/mach-omap2/board-omap4tab10-1.c	2011-02-07 16:37:13.142254817 +0530
+++ android-2.6.35/arch/arm/mach-omap2/board-omap4tab10-1.c	2011-02-07 18:05:56.669756141 +0530
@@ -52,6 +52,7 @@
 #include <plat/omap-serial.h>
 #include <plat/serial.h>
 #endif
+#include <linux/wakelock.h>
 #include <plat/mmc.h>
 #include <plat/syntm12xx.h>
 #include <plat/omap4-keypad.h>
@@ -180,7 +181,8 @@
 static void omap4_jig_cb(u8 attached);
 static void omap4_fsa9480_reset_cb(void);
 
-
+static struct wake_lock uart_lock;
+#define TWL6030_RTC_GPIO 6
 
 #if (CONFIG_SAMSUNG_OMAP4_TAB_REV <= 2)
 static unsigned backlight_gpio;
@@ -330,6 +332,7 @@
 	.keymap_data		= &omap4tab_keymap_data,
 	.rows			= 8,
 	.cols			= 8,
+	.rep                    = 0,
 };
 
 static struct twl6040_vib_platform_data omap4tab_vib_data = {
@@ -355,7 +358,6 @@
 	},
 };
 
-
 void keyboard_mux_init(void)
 {
 #if (CONFIG_SAMSUNG_OMAP4_TAB_REV == 4 || CONFIG_SAMSUNG_OMAP4_TAB_REV == 5)
@@ -1965,58 +1967,48 @@
 	iounmap(ctrl_core_pad0_i2c4_sda);
 }
 
+void plat_hold_wakelock(void)
+{
+        wake_lock_timeout(&uart_lock, 2*HZ);
+        return;
+}
+
 static struct omap_uart_port_info omap_serial_platform_data[] = {
 	{
-#if defined(CONFIG_SERIAL_OMAP_UART1_DMA)
-		.use_dma        = CONFIG_SERIAL_OMAP_UART1_DMA,
-		.dma_rx_buf_size = CONFIG_SERIAL_OMAP_UART1_RXDMA_BUFSIZE,
-		.dma_rx_timeout = CONFIG_SERIAL_OMAP_UART1_RXDMA_TIMEOUT,
-#else
 		.use_dma        = 0,
-		.dma_rx_buf_size = 0,
-		.dma_rx_timeout = 0,
-#endif /* CONFIG_SERIAL_OMAP_UART1_DMA */
-		.idle_timeout   = CONFIG_SERIAL_OMAP_IDLE_TIMEOUT,
+		.dma_rx_buf_size = DEFAULT_RXDMA_BUFSIZE,
+		.dma_rx_poll_rate = DEFAULT_RXDMA_POLLRATE,
+		.dma_rx_timeout = DEFAULT_RXDMA_TIMEOUT,
+		.idle_timeout   = DEFAULT_IDLE_TIMEOUT,
 		.flags          = 1,
+		.plat_hold_wakelock = NULL,
 	},
 	{
-#if defined(CONFIG_SERIAL_OMAP_UART2_DMA)
-		.use_dma        = CONFIG_SERIAL_OMAP_UART2_DMA,
-		.dma_rx_buf_size = CONFIG_SERIAL_OMAP_UART2_RXDMA_BUFSIZE,
-		.dma_rx_timeout = CONFIG_SERIAL_OMAP_UART2_RXDMA_TIMEOUT,
-#else
 		.use_dma        = 0,
-		.dma_rx_buf_size = 0,
-		.dma_rx_timeout = 0,
-#endif /* CONFIG_SERIAL_OMAP_UART2_DMA */
-		.idle_timeout   = CONFIG_SERIAL_OMAP_IDLE_TIMEOUT,
+		.dma_rx_buf_size = DEFAULT_RXDMA_BUFSIZE,
+		.dma_rx_poll_rate = DEFAULT_RXDMA_POLLRATE,
+		.dma_rx_timeout = DEFAULT_RXDMA_TIMEOUT,
+		.idle_timeout   = DEFAULT_IDLE_TIMEOUT,
 		.flags          = 1,
+		.plat_hold_wakelock = plat_hold_wakelock,
 	},
 	{
-#if defined(CONFIG_SERIAL_OMAP_UART3_DMA)
-		.use_dma        = CONFIG_SERIAL_OMAP_UART3_DMA,
-		.dma_rx_buf_size = CONFIG_SERIAL_OMAP_UART3_RXDMA_BUFSIZE,
-		.dma_rx_timeout = CONFIG_SERIAL_OMAP_UART3_RXDMA_TIMEOUT,
-#else
 		.use_dma        = 0,
-		.dma_rx_buf_size = 0,
-		.dma_rx_timeout = 0,
-#endif /* CONFIG_SERIAL_OMAP_UART3_DMA */
-		.idle_timeout   = CONFIG_SERIAL_OMAP_IDLE_TIMEOUT,
+		.dma_rx_buf_size = DEFAULT_RXDMA_BUFSIZE,
+		.dma_rx_poll_rate = DEFAULT_RXDMA_POLLRATE,
+		.dma_rx_timeout = DEFAULT_RXDMA_TIMEOUT,
+		.idle_timeout   = DEFAULT_IDLE_TIMEOUT,
 		.flags          = 1,
+		.plat_hold_wakelock = plat_hold_wakelock,
 	},
 	{
-#if defined(CONFIG_SERIAL_OMAP_UART4_DMA)
-		.use_dma        = CONFIG_SERIAL_OMAP_UART4_DMA,
-		.dma_rx_buf_size = CONFIG_SERIAL_OMAP_UART4_RXDMA_BUFSIZE,
-		.dma_rx_timeout = CONFIG_SERIAL_OMAP_UART4_RXDMA_TIMEOUT,
-#else
 		.use_dma        = 0,
-		.dma_rx_buf_size = 0,
-		.dma_rx_timeout = 0,
-#endif /* CONFIG_SERIAL_OMAP_UART3_DMA */
-		.idle_timeout   = CONFIG_SERIAL_OMAP_IDLE_TIMEOUT,
+		.dma_rx_buf_size = DEFAULT_RXDMA_BUFSIZE,
+		.dma_rx_poll_rate = DEFAULT_RXDMA_POLLRATE,
+		.dma_rx_timeout = DEFAULT_RXDMA_TIMEOUT,
+		.idle_timeout   = DEFAULT_IDLE_TIMEOUT,
 		.flags          = 1,
+		.plat_hold_wakelock = NULL,
 	},
 	{
 		.flags          = 0
@@ -2057,6 +2049,35 @@
 }
 #endif
 
+
+static void enable_rtc_gpio(void){
+        /* To access twl registers we enable gpio6
+         * we need this so the RTC driver can work.
+         */
+        gpio_request(TWL6030_RTC_GPIO, "h_SYS_DRM_MSEC");
+        gpio_direction_output(TWL6030_RTC_GPIO, 1);
+
+        omap_mux_init_signal("fref_clk0_out.gpio_wk6", \
+                OMAP_PIN_OUTPUT | OMAP_PIN_OFF_NONE);
+        return;
+}
+
+/* Disable default configuration of VREF_EN to minimize DDR leakage */
+static void omap4_lpddr2_config(void)
+{
+	int control_io1_3;
+	int control_io2_3;
+
+        control_io1_3 =
+                omap4_ctrl_pad_readl(OMAP4_CTRL_MODULE_PAD_CORE_CONTROL_LPDDR2IO1_3);
+        control_io2_3 =
+                omap4_ctrl_pad_readl(OMAP4_CTRL_MODULE_PAD_CORE_CONTROL_LPDDR2IO2_3);
+        omap4_ctrl_pad_writel(control_io1_3 & 0xFFFFFFF0,
+                OMAP4_CTRL_MODULE_PAD_CORE_CONTROL_LPDDR2IO1_3);
+        omap4_ctrl_pad_writel(control_io2_3 & 0xFFFFFFF0,
+                OMAP4_CTRL_MODULE_PAD_CORE_CONTROL_LPDDR2IO2_3);
+}
+
 static void __init omap4_tab_10_1_init(void)
 {
 	int status;
@@ -2067,7 +2088,9 @@
         omap4_mux_init(board_mux, package);
 
 	omap_emif_setup_device_details(&emif_devices, &emif_devices);
+	omap4_lpddr2_config();
 	omap_init_emif_timings();
+	enable_rtc_gpio();
 	omap4tab_gpiowk_setup();
 	omap4tab_touch_init();
 	omap4tab_vib_motor_init();
@@ -2075,9 +2098,10 @@
 	omap4_i2c_init();
 	omap4tab_display_init();
 	platform_add_devices(omap4tab_devices, ARRAY_SIZE(omap4tab_devices));
+	wake_lock_init(&uart_lock, WAKE_LOCK_SUSPEND, "uart_wake_lock");
 	omap_serial_init(omap_serial_platform_data);
 	omap4_twl6030_hsmmc_init(mmc);
-
+	
 #ifdef CONFIG_TIWLAN_SDIO
 	config_wlan_mux();
 #endif
Index: android-2.6.35/arch/arm/configs/android_omap4tab_rev05_defconfig
===================================================================
--- android-2.6.35.orig/arch/arm/configs/android_omap4tab_rev05_defconfig	2011-02-07 17:17:45.542241693 +0530
+++ android-2.6.35/arch/arm/configs/android_omap4tab_rev05_defconfig	2011-02-07 17:29:46.499754236 +0530
@@ -292,8 +292,11 @@
 CONFIG_SAMSUNG_OMAP4_TAB_REV=5
 # CONFIG_MACH_OMAP4_PANDA is not set
 # CONFIG_ERRATA_OMAP4_AXI2OCP is not set
+CONFIG_INTERCONNECT_IO_POSTING=y
 # CONFIG_ENABLE_L3_ERRORS is not set
 CONFIG_OMAP_ALLOW_OSWR=y
+CONFIG_OMAP4_PMD_CLKS_ENABLE=y
+CONFIG_OMAP4_KEEP_STATIC_DEPENDENCIES=y
 
 #
 # Processor Type
@@ -1020,7 +1023,6 @@
 CONFIG_SERIAL_CORE_CONSOLE=y
 CONFIG_SERIAL_OMAP=y
 CONFIG_SERIAL_OMAP_CONSOLE=y
-CONFIG_SERIAL_OMAP_IDLE_TIMEOUT=5
 # CONFIG_SERIAL_OMAP_UART1_DMA is not set
 # CONFIG_SERIAL_OMAP_UART2_DMA is not set
 # CONFIG_SERIAL_OMAP_UART3_DMA is not set
