Index: Kernel/drivers/video/omap2/displays/panel-ams353.c
===================================================================
--- Kernel.orig/drivers/video/omap2/displays/panel-ams353.c	2011-02-08 16:55:18.331004432 +0530
+++ Kernel/drivers/video/omap2/displays/panel-ams353.c	2011-02-08 16:58:08.259756338 +0530
@@ -42,14 +42,14 @@
 
 #define LCD_XRES		480
 #define LCD_YRES		800
-#define LCD_HBP	        16 //8   // Archer_LSJ DB07 : 16 => 8 
-#define LCD_HFP         16 //8   // Archer_LSJ DB07 : 16 => 8 
-#define LCD_HSW		    64  //3   // 2	
-#define LCD_VBP		    1   // 3 recomanded value in smd
-#define LCD_VFP		    28
-#define LCD_VSW		    3   // 2	
+#define LCD_HBP	        8 //16 //8   // Archer_LSJ DB07 : 16 => 8 
+#define LCD_HFP         8 //16 //8   // Archer_LSJ DB07 : 16 => 8 
+#define LCD_HSW		2   // 64  //3   // 2	
+#define LCD_VBP		3   // 1   // 3 recomanded value in smd
+#define LCD_VFP		28    //28
+#define LCD_VSW		3    //3   // 2	
 
-#define LCD_PIXCLOCK_MAX	        24380       
+#define LCD_PIXCLOCK_MAX	      25690  //24380       
 #define LCD_DEFAULT_BRIGHTNESS	    6
 #define LCD_DEFAULT_GAMMA           GAMMA_2_2   // 2.2 gamma table.
 #define HAND_FLASH_DEFAULT          flash_off	// Archer_LSJ DB10
