Index: Kernel/drivers/video/omap2/dss/overlay.c
===================================================================
--- Kernel.orig/drivers/video/omap2/dss/overlay.c	2011-02-08 19:06:29.989749742 +0530
+++ Kernel/drivers/video/omap2/dss/overlay.c	2011-02-08 19:10:12.839754773 +0530
@@ -61,6 +61,7 @@
 	struct omap_overlay_manager *old_mgr;
 	struct omap_overlay_info info;
 	int len = size;
+	int tmp;
 
 	if (buf[size-1] == '\n')
 		--len;
@@ -87,6 +88,12 @@
 
 	if (mgr && sysfs_streq(mgr->name, "tv")) {
 		ovl->get_overlay_info(ovl, &info);
+		if(info.rotation == 1 || info.rotation == 3) {
+         		tmp = info.width;
+         		info.width = info.height;
+         		info.height = tmp;
+        	}
+
 		if (mgr->device->panel.timings.x_res < info.width ||
 			mgr->device->panel.timings.y_res < info.height) {
 			printk(KERN_ERR"TV does not support downscaling"
