Index: Kernel/drivers/sensors/accel/kxsd9.c
===================================================================
--- Kernel.orig/drivers/sensors/accel/kxsd9.c	2011-01-31 16:25:52.000000000 +0530
+++ Kernel/drivers/sensors/accel/kxsd9.c	2011-02-01 03:09:57.492623991 +0530
@@ -420,6 +420,7 @@
 
 static int kxsd9_suspend(struct i2c_client * client, pm_message_t mesg)
 {
+	kxsd9_set_operation_mode(0);
 
 	if (kxsd9_timeout > -1) {
 
@@ -432,7 +433,6 @@
 		// at the same poll_delay rate after resuming.
 		timeout_at_suspend = kxsd9_timeout;
 
-		kxsd9_set_operation_mode(0);
 		shutdown_at_suspend = 1;
 	}
 	else {
Index: Kernel/drivers/sensors/gyro-mpu3050/mpu3050.c
===================================================================
--- Kernel.orig/drivers/sensors/gyro-mpu3050/mpu3050.c	2011-01-31 16:25:52.000000000 +0530
+++ Kernel/drivers/sensors/gyro-mpu3050/mpu3050.c	2011-02-01 03:12:10.712624306 +0530
@@ -462,17 +462,17 @@
         printk(KERN_INFO MODULE_NAME ": suspending device...\n");
         #endif
 
+       ret = mpu3050_set_operation_mode(GYRO_POWER_OFF);
+       if (ret < 0) {
+               printk("GYRO POWER_OFF failed!!\n");
+               return -EINVAL;
+       }
+
 	if(mpu3050_timeout > -1) {
 
 		shutdown_at_suspend = 1;
 		timeout_at_suspend = mpu3050_timeout;
 
-                ret = mpu3050_set_operation_mode(GYRO_POWER_OFF);
-                if (ret < 0) {
-                        printk("GYRO POWER_OFF failed!!\n");
-                        ret = -EINVAL;
-                }
-
 	}
 	else {
 		shutdown_at_suspend = 0;
@@ -485,6 +485,12 @@
 {
 	int ret = 0;
 
+       ret = mpu3050_set_operation_mode(GYRO_POWER_ON);
+       if (ret < 0) {
+               printk("GYRO POWER_ON failed!!\n");
+               ret = -EINVAL;
+       }
+
 	if (shutdown_at_suspend) {
 
 		#if MPU3050_DEBUG
@@ -494,11 +500,6 @@
 		shutdown_at_suspend = 0;
 
 		mpu3050_timeout = timeout_at_suspend;
-                ret = mpu3050_set_operation_mode(GYRO_POWER_ON);
-                if (ret < 0) {
-                        printk("GYRO POWER_ON failed!!\n");
-                        ret = -EINVAL;
-                }
 	}
 
 	return ret;
