Index: Kernel/sound/soc/a1026/a1026_i2c_drv.c
===================================================================
--- Kernel.orig/sound/soc/a1026/a1026_i2c_drv.c	2011-02-11 22:23:24.413322646 +0530
+++ Kernel/sound/soc/a1026/a1026_i2c_drv.c	2011-02-16 20:31:26.482696148 +0530
@@ -5,6 +5,7 @@
 #include <linux/timer.h>
 #include <linux/delay.h>
 #include <linux/slab.h>
+#include <linux/clk.h>
 
 #include <asm/gpio.h>
 #include <mach/gpio.h>
@@ -15,7 +16,7 @@
 
 
 #define A1026_NAME "A1026"
-#define OMAP4430_FREF_CLK3_OUT_BASE 0x4A30A31C
+#define OMAP4_SCRM_AUXCLK3 0x4A30A31C
 
 /*static functions*/
 static int a1026_probe (struct i2c_client *);
@@ -231,15 +232,48 @@
 
 void frefclk3_out_set(void) {
 	void __iomem *frefclk3_base=NULL;
-	int val = 0;
-
-	/* In omap4 TRM, look for AUXCLK3 to get FREF_CLK3_OUT_BASE address */
-	frefclk3_base = ioremap(OMAP4430_FREF_CLK3_OUT_BASE, 4);
+	struct clk *auxclk3_clk;
+	unsigned long rate;
+	int ret, val = 0;
+
+#if 1
+	/* In omap4 TRM, look for AUXCLK3 to get OMAP4_SCRM_AUXCLK3 base address */
+	frefclk3_base = ioremap(OMAP4_SCRM_AUXCLK3, 4);
 	val = __raw_readl(frefclk3_base);
-	val = val | (1<<8) | (0x1<<2) | (0x1<<16);
+
+	/* CLK_IN 19.2MHz */
+	//val = val | (1<<8) | (0x1<<16);
+	//val &= ~(0x3 << 1);
+
+	/* CLK_IN 96MHz */
+	val = val | (0x1<<8) | (0x1<<2) | (0x1<<16);
+
 	__raw_writel(val, (frefclk3_base));
 	val = __raw_readl(frefclk3_base);
 	debug(KERN_ERR"\n\n\t [2MIC CLK_IN] AUXCLK3 is %x \n\n",val);
+#else
+
+	auxclk3_clk = clk_get(NULL, "auxclk3_ck");
+	if (!auxclk3_clk)
+		debug("Could not get 2MIC clock - auxclk3_ck\n");
+
+	rate = clk_get_rate(auxclk3_clk);
+	debug("auxclk3_clk clk_get_rate = %ld:\n",rate);
+
+	/* CLK_IN 19.2MHz */
+	rate = rate >> 1;
+	debug("auxclk3_clk setting rate = %ld:\n",rate);
+
+	ret = clk_set_rate(auxclk3_clk, rate);
+	if (ret) 
+		debug("Unable to set 2MIC rate to %ld:\n", rate);
+	
+	clk_enable(auxclk3_clk);
+
+	frefclk3_base = ioremap(OMAP4_SCRM_AUXCLK3, 4);
+	val = __raw_readl(frefclk3_base);
+	debug(KERN_ERR"\n\n\t [2MIC CLK_IN] AUXCLK3 is %x \n\n",val);
+#endif
 
 }
 
@@ -306,6 +340,11 @@
 		debug("configure error!! \n");
 	}
 
+	/* 
+	 * In boot-up, 2MIC_POWERDOWN(WAKE_UP_) : Low -> High and continued high 
+	 * in sleep mode as well
+	 */
+	gpio_set_value(GPIO_2MIC_POWERDOWN, 1);
 	/* HWreset and sleep for 50ms!! */
 	gpio_set_value(GPIO_2MIC_RST, 0);
 	msleep(50);
Index: Kernel/arch/arm/mach-omap2/clock44xx_data.c
===================================================================
--- Kernel.orig/arch/arm/mach-omap2/clock44xx_data.c	2011-02-16 20:25:44.000000000 +0530
+++ Kernel/arch/arm/mach-omap2/clock44xx_data.c	2011-02-16 20:26:20.000000000 +0530
@@ -121,6 +121,7 @@
 
 static const struct clksel_rate div_1_0_rates[] = {
 	{ .div = 1, .val = 0, .flags = RATE_IN_4430 },
+	{ .div = 2, .val = 0, .flags = RATE_IN_4430 },
 	{ .div = 0 },
 };
 
@@ -2912,6 +2913,8 @@
 	.recalc		= &omap2_clksel_recalc,
 	.enable_reg	= OMAP4_SCRM_AUXCLK3,
 	.enable_bit	= OMAP4_ENABLE_SHIFT,
+	.round_rate	= &omap2_clksel_round_rate,
+	.set_rate	= &omap2_clksel_set_rate,
 };
 
 static struct clk auxclk4_ck = {
