Index: Kernel/drivers/media/video/omap/omap_vout.c
===================================================================
--- Kernel.orig/drivers/media/video/omap/omap_vout.c	2011-02-08 17:55:43.000000000 +0530
+++ Kernel/drivers/media/video/omap/omap_vout.c	2011-02-15 12:04:52.050865017 +0530
@@ -1018,9 +1018,6 @@
 			outw = outh;
 			outh = temp;
 
-			temp = posx;
-			posx = posy;
-			posy = temp;
 
 			break;
 
@@ -2285,6 +2282,7 @@
 			 * to have a single function to convert from v4l2 to dss window
 			 * position so the rotation logic is only in one place..
 			 */
+			
 
 			if(rotate_90_or_270(vout))
 			{
@@ -2881,18 +2879,16 @@
 			info.enabled = 1;
 			info.paddr = addr;
 			info.p_uv_addr = uv_addr;
+			info.pos_x = vout->win.w.left;
+			info.pos_y = vout->win.w.top;
 
 			if(rotate_90_or_270(vout))
 			{
-				info.pos_x = vout->win.w.top;
-				info.pos_y = vout->win.w.left;
-                        	info.out_width = vout->win.w.height;
-                        	info.out_height = vout->win.w.width;
+               	info.out_width = vout->win.w.height;
+               	info.out_height = vout->win.w.width;
 			}
 			else
 			{
-				info.pos_x = vout->win.w.left;
-				info.pos_y = vout->win.w.top;
 				info.out_width = vout->win.w.width;
 				info.out_height = vout->win.w.height;
 			}
