diff -urN -x .git -x .gitignore -x .pc -x patches -x .mailmap Kernel-orig/arch/arm/mach-omap2/usb-musb.c Kernel/arch/arm/mach-omap2/usb-musb.c
--- Kernel-orig/arch/arm/mach-omap2/usb-musb.c	2011-02-07 09:37:28.000000000 +0530
+++ Kernel/arch/arm/mach-omap2/usb-musb.c	2011-02-15 12:46:24.870863868 +0530
@@ -63,6 +63,17 @@
 
 #ifdef CONFIG_ANDROID
 #define MAX_USB_SERIAL_NUM		17
+#ifdef CONFIG_SAMSUNG_DEVGURU_USB
+#define OMAP_VENDOR_ID				0x04E8
+#define OMAP_UMS_PRODUCT_ID			0x685D
+#define OMAP_ADB_PRODUCT_ID			0x685D
+#define OMAP_UMS_ADB_PRODUCT_ID		0x685D
+#define OMAP_RNDIS_PRODUCT_ID		0x685D
+#define OMAP_RNDIS_ADB_PRODUCT_ID	0x685D
+#define OMAP_ACM_PRODUCT_ID			0x685D
+#define OMAP_ACM_ADB_PRODUCT_ID		0x685D
+#define OMAP_ACM_UMS_ADB_PRODUCT_ID	0x685D
+#else
 #define OMAP_VENDOR_ID			0x0451
 #define OMAP_UMS_PRODUCT_ID		0xD100
 #define OMAP_ADB_PRODUCT_ID		0xD101
@@ -72,6 +83,7 @@
 #define OMAP_ACM_PRODUCT_ID		0xD105
 #define OMAP_ACM_ADB_PRODUCT_ID		0xD106
 #define OMAP_ACM_UMS_ADB_PRODUCT_ID	0xD107
+#endif
 
 static char device_serial[MAX_USB_SERIAL_NUM];
 
diff -urN -x .git -x .gitignore -x .pc -x patches -x .mailmap Kernel-orig/drivers/usb/gadget/Kconfig Kernel/drivers/usb/gadget/Kconfig
--- Kernel-orig/drivers/usb/gadget/Kconfig	2011-01-24 11:55:54.000000000 +0530
+++ Kernel/drivers/usb/gadget/Kconfig	2011-02-15 12:46:24.870863868 +0530
@@ -43,6 +43,11 @@
 
 if USB_GADGET
 
+config SAMSUNG_DEVGURU_USB
+        boolean "Enable Devguru USB Support"
+        help
+          Ths will enable Devguru USB support for CDC device.
+
 config USB_GADGET_DEBUG
 	boolean "Debugging messages (DEVELOPMENT)"
 	depends on DEBUG_KERNEL
