Index: Kernel/arch/arm/mach-omap2/board-omap4tab10-1.c
===================================================================
--- Kernel.orig/arch/arm/mach-omap2/board-omap4tab10-1.c	2011-02-16 17:58:58.672115274 +0530
+++ Kernel/arch/arm/mach-omap2/board-omap4tab10-1.c	2011-02-16 17:59:59.989606397 +0530
@@ -187,6 +187,7 @@
 
 static struct wake_lock uart_lock;
 #define TWL6030_RTC_GPIO 6
+#define BLUETOOTH_UART UART2
 
 #if (CONFIG_SAMSUNG_OMAP4_TAB_REV <= 2)
 static unsigned backlight_gpio;
@@ -339,21 +340,6 @@
 	.rep                    = 0,
 };
 
-static struct twl6040_vib_platform_data omap4tab_vib_data = {
-	.max_timeout = 15000,
-	.active_low = 0,
-	.initial_vibrate = 0,
-};
-
-static struct platform_device omap4tab_vib = {
-	.name           = VIB_NAME,
-	.id             = -1,
-	.dev            = {
-		.platform_data  = &omap4tab_vib_data,
-	},
-};
-
-
 static struct platform_device omap4tab_keypad_led = {
 	.name	=	"keypad_led",
 	.id	=	-1,
@@ -1062,7 +1048,6 @@
 static struct platform_device *omap4tab_devices[] __initdata = {
 	&wl128x_device,
 	&omap4tab_hdmi_audio_device,
-	&omap4tab_vib,
 	&omap4tab_keypad_led,
 	&samsung_omap4_pwr_device,
 	&samsung_omap4_gyro_lp530al_device,
@@ -1361,20 +1346,30 @@
 	.tblsize			= ARRAY_SIZE(omap4tab_batt_table),
 };
 
+static void omap4_audio_conf(void)
+{
+    /* twl6040 naudint */
+    omap_mux_init_signal("sys_nirq2.sys_nirq2", \
+			        OMAP_PIN_INPUT_PULLUP);
+}
+
 static struct twl4030_codec_audio_data twl6040_audio = {
-	.audio_mclk	= 38400000,
-	.audpwron_gpio	= 127,
-	.naudint_irq	= OMAP44XX_IRQ_SYS_2N,
+	/* Add audio only data */
 };
 
 static struct twl4030_codec_vibra_data twl6040_vibra = {
-	.audio_mclk	= 38400000,
+	 .max_timeout    = 15000,
+	 .initial_vibrate = 0,
+
 };
 
 static struct twl4030_codec_data twl6040_codec = {
 	.audio_mclk	= 38400000,
 	.audio = &twl6040_audio,
 	.vibra = &twl6040_vibra,
+    .audpwron_gpio  = 127,
+    .naudint_irq    = OMAP44XX_IRQ_SYS_2N,
+    .irq_base   = TWL6040_CODEC_IRQ_BASE,
 };
 
 static struct twl4030_platform_data omap4tab_twldata = {
@@ -1971,10 +1966,19 @@
 	iounmap(ctrl_core_pad0_i2c4_sda);
 }
 
-void plat_hold_wakelock(void)
+void plat_hold_wakelock(void *up, int flag)
 {
-        wake_lock_timeout(&uart_lock, 2*HZ);
-        return;
+    struct uart_omap_port *up2 = (struct uart_omap_port *)up;
+    /*Specific wakelock for bluetooth usecases*/
+    if ((up2->pdev->id == BLUETOOTH_UART)
+		        && ((flag == WAKELK_TX) || (flag == WAKELK_RX)))
+	        wake_lock_timeout(&uart_lock, 2*HZ);
+
+    /*Specific wakelock for console usecases*/
+    if ((up2->pdev->id != BLUETOOTH_UART)
+		        && ((flag == WAKELK_IRQ) || (flag == WAKELK_RESUME)))
+	        wake_lock_timeout(&uart_lock, 5*HZ);
+    return;
 }
 
 static struct omap_uart_port_info omap_serial_platform_data[] = {
@@ -1986,6 +1990,12 @@
 		.idle_timeout   = DEFAULT_IDLE_TIMEOUT,
 		.flags          = 1,
 		.plat_hold_wakelock = NULL,
+        .rts_padconf    = 0,
+        .rts_override   = 0,
+        .padconf    = OMAP4_CTRL_MODULE_PAD_SDMMC1_CMD_OFFSET,
+        .padconf_wake_ev = 0,
+        .wk_mask    = 0,
+
 	},
 	{
 		.use_dma        = 0,
@@ -1995,6 +2005,17 @@
 		.idle_timeout   = DEFAULT_IDLE_TIMEOUT,
 		.flags          = 1,
 		.plat_hold_wakelock = plat_hold_wakelock,
+        .rts_padconf    = OMAP4_CTRL_MODULE_PAD_UART2_RTS_OFFSET,
+        .rts_override   = 0,
+        .padconf    = OMAP4_CTRL_MODULE_PAD_UART2_RX_OFFSET,
+        .padconf_wake_ev =
+			            OMAP4_CTRL_MODULE_PAD_CORE_PADCONF_WAKEUPEVENT_3,
+        .wk_mask    =
+		            OMAP4_UART2_TX_DUPLICATEWAKEUPEVENT_MASK |
+		            OMAP4_UART2_RX_DUPLICATEWAKEUPEVENT_MASK |
+		            OMAP4_UART2_RTS_DUPLICATEWAKEUPEVENT_MASK |
+		            OMAP4_UART2_CTS_DUPLICATEWAKEUPEVENT_MASK,
+
 	},
 	{
 		.use_dma        = 0,
@@ -2004,6 +2025,17 @@
 		.idle_timeout   = DEFAULT_IDLE_TIMEOUT,
 		.flags          = 1,
 		.plat_hold_wakelock = plat_hold_wakelock,
+        .rts_padconf    = 0,
+        .rts_override   = 0,
+        .padconf    = OMAP4_CTRL_MODULE_PAD_UART3_RX_IRRX_OFFSET,
+        .padconf_wake_ev =
+			            OMAP4_CTRL_MODULE_PAD_CORE_PADCONF_WAKEUPEVENT_4,
+        .wk_mask    =
+		            OMAP4_UART3_TX_IRTX_DUPLICATEWAKEUPEVENT_MASK |
+		            OMAP4_UART3_RX_IRRX_DUPLICATEWAKEUPEVENT_MASK |
+		            OMAP4_UART3_RTS_SD_DUPLICATEWAKEUPEVENT_MASK |
+		            OMAP4_UART3_CTS_RCTX_DUPLICATEWAKEUPEVENT_MASK,
+
 	},
 	{
 		.use_dma        = 0,
@@ -2013,6 +2045,15 @@
 		.idle_timeout   = DEFAULT_IDLE_TIMEOUT,
 		.flags          = 1,
 		.plat_hold_wakelock = NULL,
+        .rts_padconf    = 0,
+        .rts_override   = 0,
+        .padconf    = OMAP4_CTRL_MODULE_PAD_UART4_RX_OFFSET,
+        .padconf_wake_ev =
+			            OMAP4_CTRL_MODULE_PAD_CORE_PADCONF_WAKEUPEVENT_4,
+        .wk_mask    =
+		            OMAP4_UART4_TX_DUPLICATEWAKEUPEVENT_MASK |
+		            OMAP4_UART4_RX_DUPLICATEWAKEUPEVENT_MASK,
+
 	},
 	{
 		.flags          = 0
@@ -2095,6 +2136,7 @@
 	omap4_lpddr2_config();
 	omap_init_emif_timings();
 	enable_rtc_gpio();
+	omap4_audio_conf();
 	omap4tab_gpiowk_setup();
 	omap4tab_touch_init();
 	omap4tab_vib_motor_init();
Index: Kernel/arch/arm/configs/android_omap4tab_rev05_defconfig
===================================================================
--- Kernel.orig/arch/arm/configs/android_omap4tab_rev05_defconfig	2011-02-16 17:58:58.672115274 +0530
+++ Kernel/arch/arm/configs/android_omap4tab_rev05_defconfig	2011-02-16 19:57:15.830863528 +0530
@@ -1,10 +1,9 @@
 #
 # Automatically generated make config: don't edit
 # Linux kernel version: 2.6.35.7
-# Wed Feb  2 20:11:44 2011
+# Wed Feb 16 19:57:03 2011
 #
 CONFIG_ARM=y
-CONFIG_HAVE_PWM=y
 CONFIG_SYS_SUPPORTS_APM_EMULATION=y
 CONFIG_GENERIC_GPIO=y
 CONFIG_GENERIC_TIME=y
@@ -51,6 +50,7 @@
 CONFIG_SYSVIPC_SYSCTL=y
 # CONFIG_POSIX_MQUEUE is not set
 CONFIG_BSD_PROCESS_ACCT=y
+# CONFIG_CONTEXT_SWITCH_COUNTER is not set
 # CONFIG_BSD_PROCESS_ACCT_V3 is not set
 # CONFIG_TASKSTATS is not set
 # CONFIG_AUDIT is not set
@@ -279,6 +279,7 @@
 # CONFIG_OMAP_PM_NONE is not set
 # CONFIG_OMAP_PM_NOOP is not set
 CONFIG_OMAP_PM=y
+CONFIG_OMAP_HDMI_AUDIO_WA=y
 
 #
 # OMAP Board Type
@@ -1023,10 +1024,6 @@
 CONFIG_SERIAL_CORE_CONSOLE=y
 CONFIG_SERIAL_OMAP=y
 CONFIG_SERIAL_OMAP_CONSOLE=y
-# CONFIG_SERIAL_OMAP_UART1_DMA is not set
-# CONFIG_SERIAL_OMAP_UART2_DMA is not set
-# CONFIG_SERIAL_OMAP_UART3_DMA is not set
-# CONFIG_SERIAL_OMAP_UART4_DMA is not set
 # CONFIG_SERIAL_TIMBERDALE is not set
 # CONFIG_SERIAL_ALTERA_JTAGUART is not set
 # CONFIG_SERIAL_ALTERA_UART is not set
@@ -1262,7 +1259,7 @@
 # CONFIG_TWL4030_POWER is not set
 # CONFIG_TWL4030_CODEC is not set
 # CONFIG_MFD_TC35892 is not set
-CONFIG_TWL6030_PWM=y
+# CONFIG_TWL6030_PWM is not set
 CONFIG_TWL6030_GPADC=y
 CONFIG_TWL6040_CODEC=y
 # CONFIG_MFD_TMIO is not set
@@ -1659,6 +1656,7 @@
 # CONFIG_USB_TEST is not set
 # CONFIG_USB_ISIGHTFW is not set
 CONFIG_USB_GADGET=y
+# CONFIG_SAMSUNG_DEVGURU_USB is not set
 # CONFIG_USB_GADGET_DEBUG is not set
 # CONFIG_USB_GADGET_DEBUG_FILES is not set
 # CONFIG_USB_GADGET_DEBUG_FS is not set
@@ -1919,17 +1917,16 @@
 CONFIG_ST_FM=m
 CONFIG_ST_GPS=m
 # CONFIG_ADIS16255 is not set
-CONFIG_OMAP_HSI=y
-CONFIG_OMAP_HSI_DEVICE=y
+# CONFIG_OMAP_HSI is not set
+# CONFIG_OMAP_HSI_DEVICE is not set
 # CONFIG_OMAP_SSI_DEVICE is not set
-CONFIG_OMAP_HSI_CHAR=m
-CONFIG_OMAP_HSI_PROTOCOL=y
 CONFIG_Sys_Link=y
 CONFIG_SYSLINK_PROC=y
 CONFIG_SYSLINK_PROC4430=y
 CONFIG_DUCATI_BASEIMAGE_PHYS_ADDR=0x9CF00000
 CONFIG_SYSLINK_DUCATI_PM=y
 CONFIG_SYSLINK_IPU_SELF_HIBERNATION=y
+# CONFIG_DUCATI_WATCH_DOG is not set
 # CONFIG_SYSLINK_IPU_PM_TRACES is not set
 CONFIG_OMAP_DEVICE_HANDLER=y
 CONFIG_MPU_SYSLINK_PLATFORM=y
@@ -1940,16 +1937,8 @@
 # CONFIG_DMM_DMA_API is not set
 CONFIG_HAPTIC=y
 CONFIG_HAPTIC_CLASS=y
-CONFIG_HAPTIC_SAMSUNG_PWM=y
-CONFIG_SAMSUNG_PHONE_SVNET=y
-CONFIG_PHONE_MODEMCTL=m
-# CONFIG_PHONE_ONEDRAM is not set
-# CONFIG_PHONE_IPC_SPI is not set
-CONFIG_PHONE_IPC_HSI=y
-CONFIG_PHONE_SVNET=m
-# CONFIG_PHONE_LOOPBACK_TEST is not set
-# CONFIG_PHONE_STORAGE is not set
-# CONFIG_SVNET_WHITELIST is not set
+# CONFIG_HAPTIC_SAMSUNG_PWM is not set
+# CONFIG_SAMSUNG_PHONE_SVNET is not set
 
 #
 # File systems
