Index: Kernel/drivers/input/keyboard/cypress/cypress-touchkey.c
===================================================================
--- Kernel.orig/drivers/input/keyboard/cypress/cypress-touchkey.c	2011-02-16 13:17:38.000000000 +0530
+++ Kernel/drivers/input/keyboard/cypress/cypress-touchkey.c	2011-02-17 11:42:38.450863825 +0530
@@ -196,7 +196,6 @@
 					 touchkey_keycode[2], 0);
 			retry = 10;
 			while (retry--) {
-				gpio_direction_output(_3_GPIO_TOUCH_EN, 0);
 				gpio_direction_output(_3_GPIO_TOUCH_LED_EN, 0);
 				mdelay(300);
 				init_hw();
@@ -210,7 +209,6 @@
 			//touchkey die , do not enable touchkey
 			//enable_irq(IRQ_TOUCH_INT);
 			touchkey_enable = -1;
-			gpio_direction_output(_3_GPIO_TOUCH_EN, 0);
 			gpio_direction_output(_3_GPIO_TOUCH_LED_EN, 0);
 #ifdef I2C_BIT_BANG
 			gpio_direction_output(_3_TOUCH_SDA_28V, 0);
@@ -303,7 +301,6 @@
 	disable_irq(IRQ_TOUCH_INT);
 	disable_irq_wake(IRQ_TOUCH_INT);
 	disable_irq(home_key_irq);
-	gpio_direction_output(_3_GPIO_TOUCH_EN, 0);
 	gpio_direction_output(_3_GPIO_TOUCH_LED_EN, 0);
 #ifdef I2C_BIT_BANG
 	gpio_direction_output(_3_TOUCH_SDA_28V, 0);
@@ -607,7 +604,6 @@
 	dbg("Enter %s\n", __FUNCTION__);
 	if (*buf == '0') {
 		disable_irq(IRQ_TOUCH_INT);
-		gpio_direction_output(_3_GPIO_TOUCH_EN, 0);
 		gpio_direction_output(_3_GPIO_TOUCH_LED_EN, 0);
 		touchkey_enable = -2;
 	} else if (*buf == '1') {
Index: Kernel/drivers/input/keyboard/cypress/issp_driver_routines.c
===================================================================
--- Kernel.orig/drivers/input/keyboard/cypress/issp_driver_routines.c	2011-02-16 13:17:38.000000000 +0530
+++ Kernel/drivers/input/keyboard/cypress/issp_driver_routines.c	2011-02-17 11:42:49.630863591 +0530
@@ -475,7 +475,6 @@
 // ****************************************************************************
 void RemoveTargetVDD(void)
 {
-    gpio_direction_output(_3_GPIO_TOUCH_EN, 0);
     gpio_direction_output(_3_GPIO_TOUCH_LED_EN, 0);
 }
 #endif
Index: Kernel/drivers/input/keyboard/cypress/issp_main.c
===================================================================
--- Kernel.orig/drivers/input/keyboard/cypress/issp_main.c	2011-02-16 13:17:38.000000000 +0530
+++ Kernel/drivers/input/keyboard/cypress/issp_main.c	2011-02-17 11:43:00.020863927 +0530
@@ -920,7 +920,6 @@
     // Acquire the device through reset or power cycle
     //s3c_gpio_setpull(_3_TOUCH_SCL_28V, S3C_GPIO_PULL_NONE); 
     //s3c_gpio_setpull(_3_TOUCH_SDA_28V, S3C_GPIO_PULL_NONE); 
-    gpio_direction_output(_3_GPIO_TOUCH_EN, 0);
     gpio_direction_output(_3_GPIO_TOUCH_LED_EN, 0);
     msleep(1);
 #ifdef RESET_MODE
