Index: Kernel/arch/arm/mach-omap2/hsmmc.c
===================================================================
--- Kernel.orig/arch/arm/mach-omap2/hsmmc.c	2011-02-17 18:41:36.501997421 +0530
+++ Kernel/arch/arm/mach-omap2/hsmmc.c	2011-02-15 19:00:39.641997000 +0530
@@ -236,7 +236,7 @@
 		.low_speed	= 0,
 		.wide_bus	= 1,
 		.high_power	= 0,
-		.high_speed	= 0,
+		.high_speed	= 1,
 		.disable_cd	= 1,
 	},
 	.funcs  = wifi_func_array,
@@ -357,18 +357,16 @@
 		else
 			mmc->slots[0].features |= HSMMC_HAS_PBIAS;
 
-		if (cpu_is_omap44xx()) {
-			if (omap_rev() > OMAP4430_REV_ES1_0)
-				mmc->slots[0].features |= HSMMC_HAS_UPDATED_RESET;
+		if (cpu_is_omap44xx() && (omap_rev() > OMAP4430_REV_ES1_0))
+			mmc->slots[0].features |= HSMMC_HAS_UPDATED_RESET;
 
+		if (cpu_is_omap44xx())
 			mmc->slots[0].features |= HSMMC_DVFS_24MHZ_CONST;
 
-			if (c->mmc >= 3 && c->mmc <= 5) {
-				mmc->slots[0].features |= HSMMC_HAS_48MHZ_MASTER_CLK;
-				mmc->get_context_loss_count =
-						hsmmc_get_context_loss;
-			}
-		}
+		if (cpu_is_omap44xx() &&
+			(c->mmc == 3 || c->mmc == 4 || c->mmc == 5))
+			mmc->slots[0].features |= HSMMC_HAS_48MHZ_MASTER_CLK;
+
 		switch (c->mmc) {
 		case 1:
 			if (mmc->slots[0].features & HSMMC_HAS_PBIAS) {
@@ -412,6 +410,7 @@
 			break;
 		case 4:
 		case 5:
+			mmc->get_context_loss_count = hsmmc_get_context_loss;
 			/* TODO Update required */
 			mmc->slots[0].before_set_reg = NULL;
 			mmc->slots[0].after_set_reg = NULL;
Index: Kernel/drivers/mmc/host/omap_hsmmc.c
===================================================================
--- Kernel.orig/drivers/mmc/host/omap_hsmmc.c	2011-02-17 18:46:58.201997757 +0530
+++ Kernel/drivers/mmc/host/omap_hsmmc.c	2011-02-14 19:32:17.253246000 +0530
@@ -2320,10 +2320,6 @@
 	if (mmc_slot(host).features & HSMMC_HAS_48MHZ_MASTER_CLK)
 		host->master_clock = OMAP_MMC_MASTER_CLOCK / 2;
 
-	host->master_clock = OMAP_MMC_MASTER_CLOCK;
-	if (mmc_slot(host).features & HSMMC_HAS_48MHZ_MASTER_CLK)
-		host->master_clock = OMAP_MMC_MASTER_CLOCK / 2;
-
 #ifdef CONFIG_TIWLAN_SDIO
 	if (pdev->id == CONFIG_TIWLAN_MMC_CONTROLLER-1) {
 		if (pdata->slots[0].embedded_sdio != NULL) {
