Index: Kernel/drivers/sensors/mpu3050/compass/yas529-kernel.c
===================================================================
--- Kernel.orig/drivers/sensors/mpu3050/compass/yas529-kernel.c	2011-02-17 20:39:52.629736343 +0530
+++ Kernel/drivers/sensors/mpu3050/compass/yas529-kernel.c	2011-02-17 20:40:02.122237834 +0530
@@ -374,6 +374,7 @@
 	short SX, SY1, SY2, SY, SZ;
 	short row1fixed, row2fixed, row3fixed;
 
+	printk("\n\nFunc : %s Entry +++++++++++++++++\n\n",__func__);
 	/* Config register - Measurements results - "110 00 000" */
 	dummyData[0] = YAS_REG_CONFR | 0x00;
 	result =
@@ -393,6 +394,7 @@
 	    yas529_sensor_i2c_read(mlsl_handle, pdata->address,
 				   dummyRegister, 6,
 				   (unsigned char *) &rawData);
+	printk(KERN_ERR "\n result : %d ; add4:%d\n\n",result,pdata->address);
 	ERROR_CHECK(result);
 
 	stat = rawData[0] & 0x80;
@@ -423,6 +425,7 @@
 		row3fixed =
 		    (short) ((a7 * SX + a8 * SY + a9 * SZ) >> 7) * 41;
 
+		printk(KERN_ERR" \n\n\n!!!!!!!! Func:%d => row1fixed=%d, row2fixed=%d, row3fixed=%d\n\n ",__func__,row1fixed,row2fixed,row3fixed);
 		data[0] = row1fixed >> 8;
 		data[1] = row1fixed & 0xFF;
 		data[2] = row2fixed >> 8;
@@ -454,6 +457,7 @@
 
 struct ext_slave_descr *yas529_get_slave_descr(void)
 {
+	printk("\n\nFunc : %s Entry ------------------------\n\n",__func__);
 	return &yas529_descr;
 }
 EXPORT_SYMBOL(yas529_get_slave_descr);
Index: Kernel/drivers/sensors/mpu3050/mpuirq.c
===================================================================
--- Kernel.orig/drivers/sensors/mpu3050/mpuirq.c	2011-02-17 20:39:52.639727417 +0530
+++ Kernel/drivers/sensors/mpu3050/mpuirq.c	2011-02-17 20:40:02.122237834 +0530
@@ -191,7 +191,7 @@
 
 	mpuirq_data.interruptcount++;
 
-	printk("\n\n############# mpuirq_handler called #############\n\n");
+//	printk("\n\n############# mpuirq_handler called #############\n\n");
 
 	/* wake up (unblock) for reading data from userspace */
 	/* and ignore first interrupt generated in module init */
Index: Kernel/arch/arm/mach-omap2/board-omap4tab10-1.c
===================================================================
--- Kernel.orig/arch/arm/mach-omap2/board-omap4tab10-1.c	2011-02-17 20:39:52.639727417 +0530
+++ Kernel/arch/arm/mach-omap2/board-omap4tab10-1.c	2011-02-17 20:40:02.122237834 +0530
@@ -273,8 +273,8 @@
 
 static struct mpu3050_platform_data mpu_data = {
 	.int_config  = 0x12,
-	.orientation = {  -1,  0,  0, 
-			   0,  1,  0, 
+	.orientation = {   0, -1,  0, 
+			  -1,  0,  0, 
 			   0,  0, -1 },
 	/* accel */
 	.accel = {
@@ -282,13 +282,13 @@
 	.adapt_num   = 4,
 	.bus         = EXT_SLAVE_BUS_SECONDARY,
 	.address     = 0x18,
-	.orientation = {  -1,  0,  0, 
-			   0,  1,  0, 
-			   0,  0, -1 },
+	.orientation = {   1,  0,  0, 
+			   0, -1,  0, 
+			   0,  0,  1 },
 	 },
 	/* compass */
 	.compass = {
-	.get_slave_descr = get_compass_slave_descr,
+	.get_slave_descr = yas529_get_slave_descr,
 	.adapt_num   = 4,
 	.bus         = EXT_SLAVE_BUS_PRIMARY,
 	.address     = 0x2E,
