Index: Kernel_922640/arch/arm/mach-omap2/omap4-iopad.c
===================================================================
--- Kernel_922640.orig/arch/arm/mach-omap2/omap4-iopad.c	2011-02-22 20:33:39.000000000 +0530
+++ Kernel_922640/arch/arm/mach-omap2/omap4-iopad.c	2011-02-22 20:36:45.341987138 +0530
@@ -95,7 +95,7 @@
 	//{ "gpmc_a20",	"EARPATH_SEL",		0x068, (),	() },
 	{ "gpmc_a21",	"GYRO_INT",		0x06A, (IEN | M3),	(OFF_IN | OFF_EN) },
 	{ "gpmc_a22",	"TOUCH_nINT",		0x06C, (IEN | M3),	(OFF_IN | OFF_EN) },
-	{ "gpmc_a23",	"UART_SEL",		0x06E, (PTU | EN | M3),	(OFF_IN_PU) },
+	{ "gpmc_a23",	"UART_SEL",		0x06E, (PTU | EN | M3),	(OFF_DIS) },
 	{ "gpmc_a24",	"MICBIAS_EN",		0x070, (M3),		(OFF_IN_PD) },
 	{ "gpmc_a25",	"MICBIAS_EN2",		0x072, (M3),		(OFF_IN_PD) },
 	//{ "gpmc_ncs0",	"RESET_REQ_N",		0x074, (),	() },
@@ -117,10 +117,10 @@
 	{ "gpmc_ncs6",	"BT_EN",		0x094, (M3),		(OFF_IN_PD) },
 	{ "gpmc_ncs7",	"WLAN_EN",		0x096, (M3),		(OFF_IN_PD) },
 
-	//{ "hdmi_hpd",	"AP_HDMI_HPD",		0x098, (),		() },
-	//{ "hdmi_cec",	"AP_HDMI_CEC",		0x09A, (),		() },
-	//{ "hdmi_ddc_scl", "AP_HDMI_I2C_SCL",	0x09C, (),		() },
-	//{ "hdmi_ddc_sda", "AP_HDMI_I2C_SDA",	0x09E, (),		() },
+	{ "hdmi_hpd",	"AP_HDMI_HPD",		0x098, (IEN | PTD | EN | M0),	(OFF_DIS) },
+	{ "hdmi_cec",	"AP_HDMI_CEC",		0x09A, (IEN | PTU | EN | M0),	(OFF_DIS) },
+	{ "hdmi_ddc_scl", "AP_HDMI_I2C_SCL",	0x09C, (IEN | M0),		(OFF_IN_PU) },
+	{ "hdmi_ddc_sda", "AP_HDMI_I2C_SDA",	0x09E, (IEN | M0),		(OFF_IN_PU) },
 
 	//{ "csi21_dx0",	"8M_CLK_P",		0x0A0, (),		() },
 	//{ "csi21_dy0",	"8M_CLK_N",		0x0A2, (),		() },
@@ -128,19 +128,19 @@
 	//{ "csi21_dy1",	"8M_DN_1",		0x0A6, (),		() },
 	//{ "csi21_dx2",	"8M_DP_2",		0x0A8, (),		() },
 	//{ "csi21_dy2",	"8M_DN_2",		0x0AA, (),		() },
-	//{ "csi21_dx3",	"-nc-",		0x0AC, (PTD | EN | M7),		(OFF_DIS) },
-	//{ "csi21_dy3",	"-nc-",		0x0AE, (PTD | EN | M7),		(OFF_DIS) },
-	//{ "csi21_dx4",	"-nc-",		0x0B0, (PTD | EN | M7),		(OFF_DIS) },
-	//{ "csi21_dy4",	"-nc-",		0x0B2, (PTD | EN | M7),		(OFF_DIS) },
+	{ "csi21_dx3",	"-nc-",		0x0AC, (PTD | EN | M7),		(OFF_DIS) },
+	{ "csi21_dy3",	"-nc-",		0x0AE, (PTD | EN | M7),		(OFF_DIS) },
+	{ "csi21_dx4",	"-nc-",		0x0B0, (PTD | EN | M7),		(OFF_DIS) },
+	{ "csi21_dy4",	"-nc-",		0x0B2, (PTD | EN | M7),		(OFF_DIS) },
 
 	//{ "csi22_dx0",	"1.3M_CLK_P",		0x0B4, (),		() },
 	//{ "csi22_dy0",	"1.3M_CLK_N",		0x0B6, (),		() },
 	//{ "csi22_dx1",	"1.3M_DATA_P",		0x0B8, (),		() },
 	//{ "csi22_dy1",	"1.3M_DATA_N",		0x0BA, (),		() },
 
-	//{ "cam_shutter",	"-nc-",		0x0BC, (PTD | EN | M7),		(OFF_DIS) },
-	//{ "cam_strobe",		"-nc-",		0x0BE, (PTD | EN | M7),		(OFF_DIS) },
-	//{ "cam_globalreset",	"-nc-",		0x0C0, (PTD | EN | M7),		(OFF_DIS) },
+	{ "cam_shutter",	"-nc-",		0x0BC, (PTD | EN | M7),		(OFF_DIS) },
+	{ "cam_strobe",		"-nc-",		0x0BE, (PTD | EN | M7),		(OFF_DIS) },
+	{ "cam_globalreset",	"-nc-",		0x0C0, (PTD | EN | M7),		(OFF_DIS) },
 
 	//{ "usbb1_ulpitll_clk",	"MIPI_HSI_TX_WAKE",	0x0C2, (),		() },
 	//{ "usbb1_ulpitll_stp",	"MIPI_HSI_TX_DATA",	0x0C4, (),		() },
@@ -151,34 +151,34 @@
 	//{ "usbb1_ulpitll_dat2",	"MIPI_HSI_RX_FLG",	0x0CE, (),		() },
 	//{ "usbb1_ulpitll_dat3",	"MIPI_HSI_RX_RDY",	0x0D0, (),		() },
 	//{ "usbb1_ulpitll_dat4",	"BOOT_MODE",		0x0D2, (),		() },
-	//{ "usbb1_ulpitll_dat5",	"-nc-",		0x0D4, (PTD | EN | M7),		(OFF_DIS) },
+	{ "usbb1_ulpitll_dat5",	"-nc-",		0x0D4, (PTD | EN | M7),		(OFF_DIS) },
 	//{ "usbb1_ulpitll_dat6",	"EAR_SEND_END",		0x0D6, (),		() },
 	//{ "usbb1_ulpitll_dat7",	"JIG_ON_18",		0x0D8, (),		() },
 
-	//{ "usbb1_hsic_data",	"-nc-",		0x0DA, (PTD | EN | M7),		(OFF_DIS) },
-	//{ "usbb1_hsic_strobe",	"-nc-",		0x0DC, (PTD | EN | M7),		(OFF_DIS) },
+	{ "usbb1_hsic_data",	"-nc-",		0x0DA, (M0),		(OFF_DIS) },
+	{ "usbb1_hsic_strobe",	"-nc-",		0x0DC, (M0),		(OFF_DIS) },
 	{ "usbc1_icusb_dp",	"-nc-",		0x0DE, (PTD | EN | M7),		(OFF_DIS) },
 	{ "usbc1_icusb_dm",	"-nc-",		0x0E0, (PTD | EN | M7),		(OFF_DIS) },
 
-	//{ "sdmmc1_clk",  "T_FLASH_CLK",		0x0E2, (),		() },
+	{ "sdmmc1_clk",  "T_FLASH_CLK",		0x0E2, (IEN | PTD | EN | M0),	(OFF_DIS) },
 	{ "sdmmc1_cmd",  "T_FLASH_CLK",		0x0E4, (IEN | PTU | EN | M0),	(OFF_IN_PU) },
 	{ "sdmmc1_dat0", "T_FLASH_D(0)",	0x0E6, (IEN | PTU | EN | M0),	(OFF_IN_PU) },
 	{ "sdmmc1_dat1", "T_FLASH_D(1)",	0x0E8, (IEN | PTU | EN | M0),	(OFF_IN_PU) },
 	{ "sdmmc1_dat2", "T_FLASH_D(2)",	0x0EA, (IEN | PTU | EN | M0),	(OFF_IN_PU) },
 	{ "sdmmc1_dat3", "T_FLASH_D(3)",	0x0EC, (IEN | PTU | EN | M0),	(OFF_IN_PU) },
-	//{ "sdmmc1_dat4", "-nc-",		0x0EE, (PTD | EN | M7),		(OFF_DIS) },
-	//{ "sdmmc1_dat5", "-nc-",		0x0F0, (PTD | EN | M7),		(OFF_DIS) },
-	//{ "sdmmc1_dat6", "-nc-",		0x0F2, (PTD | EN | M7),		(OFF_DIS) },
-	//{ "sdmmc1_dat7", "-nc-",		0x0F4, (PTD | EN | M7),		(OFF_DIS) },
+	{ "sdmmc1_dat4", "-nc-",		0x0EE, (PTD | EN | M7),		(OFF_DIS) },
+	{ "sdmmc1_dat5", "-nc-",		0x0F0, (PTD | EN | M7),		(OFF_DIS) },
+	{ "sdmmc1_dat6", "-nc-",		0x0F2, (PTD | EN | M7),		(OFF_DIS) },
+	{ "sdmmc1_dat7", "-nc-",		0x0F4, (PTD | EN | M7),		(OFF_DIS) },
 
 	//{ "abe_mcbsp2_clkx",	"REC_PCM_CLK",		0x0F6, (),		() },
 	//{ "abe_mcbsp2_dr",	"REC_PCM_IN",		0x0F8, (),		() },
 	//{ "abe_mcbsp2_dx",	"REC_PCM_OUT",		0x0FA, (),		() },
 	//{ "abe_mcbsp2_fsx",	"REC_PCM_SYNC",		0x0FC, (),		() },
-	//{ "abe_mcbsp1_clkx",	"BT_PCM_CLK",		0x0FE, (),		() },
-	//{ "abe_mcbsp1_dr",	"BT_PCM_OUT",		0x100, (),		() },
-	//{ "abe_mcbsp1_dx",	"BT_PCM_DIN",		0x102, (),		() },
-	//{ "abe_mcbsp1_fsx",	"BT_PCM_SYNC",		0x104, (),		() },
+	{ "abe_mcbsp1_clkx",	"BT_PCM_CLK",		0x0FE, (IEN | M0),	(OFF_IN_PD) },
+	{ "abe_mcbsp1_dr",	"BT_PCM_OUT",		0x100, (IEN | M0),	(OFF_IN_PD) },
+	{ "abe_mcbsp1_dx",	"BT_PCM_DIN",		0x102, (M0),		(OFF_DIS) },
+	{ "abe_mcbsp1_fsx",	"BT_PCM_SYNC",		0x104, (IEN | M0),	(OFF_IN_PD) },
 	//{ "abe_pdm_ul_data",	"PDM_UL_DATA",		0x106, (),		() },
 	//{ "abe_pdm_dl_data",	"PDM_DL_DATA",		0x108, (),		() },
 	//{ "abe_pdm_frame",	"PDM_FRAME",		0x10A, (),		() },
@@ -198,8 +198,8 @@
 
 	//{ "i2c1_scl",	"PHEONIX_I2C_SCL",	0x122, (),		() },
 	//{ "i2c1_sda",	"PHEONIX_I2C_SDA",	0x124, (),		() },
-	//{ "i2c2_scl",	"CAM_I2C_SCL",		0x126, (),		() },
-	//{ "i2c2_sda",	"CAM_I2C_SDA",		0x128, (),		() },
+	{ "i2c2_scl",	"CAM_I2C_SCL",		0x126, (IEN | M0),	(OFF_IN_PD) },
+	{ "i2c2_sda",	"CAM_I2C_SDA",		0x128, (IEN | M0),	(OFF_IN_PD) },
 	/* NOTE: Fuel gauge expects I2C SCL and SDA lines to be low to enter sleep mode. This save ~70uA */
 	{ "i2c3_scl",	"AP_I2C_SCL",		0x12A, (IEN | PTU | EN | M0),	(OFF_IN_PD) },
 	{ "i2c3_sda",	"AP_I2C_SDA",		0x12C, (IEN | PTU | EN | M0),	(OFF_IN_PD) },
@@ -210,14 +210,14 @@
 	//{ "mcspi1_somi", "NFC_FIRMWARE",	0x134, (),		() },
 	{ "mcspi1_simo", "NFC_EN",		0x136, (M3),		(OFF_IN_PD) },
 	{ "mcspi1_cs0",  "-nc-",		0x138, (IEN | PTD | EN | M3),		(OFF_IN_PD) },
-	//{ "mcspi1_cs1",  "8M_ISP_INT",		0x13A, (),		() },
-	//{ "mcspi1_cs2",  "8M_nRST",		0x13C, (),		() },
-	//{ "mcspi1_cs3",  "1.3M_nRST",		0x13E, (),		() },
-
-	//{ "uart3_cts_rctx",	"-nc-",		0x140, (PTD | EN | M7),		(OFF_DIS) },
-	//{ "uart3_rts_sd",	"-nc-",		0x142, (PTD | EN | M7),		(OFF_DIS) },
-	//{ "uart3_rx_irrx",	"AP_FLM_RXD",	0x144, (),		() },
-	//{ "uart3_tx_irtx",	"AP_FLM_TXD",	0x146, (),		() },
+	{ "mcspi1_cs1",  "8M_ISP_INT",		0x13A, (IEN | EN | M3),		(OFF_IN_PD) },
+	{ "mcspi1_cs2",  "8M_nRST",		0x13C, (M3),		(OFF_OUT_PD) },
+	{ "mcspi1_cs3",  "1.3M_nRST",		0x13E, (M3),		(OFF_OUT_PD) },
+
+	{ "uart3_cts_rctx",	"-nc-",		0x140, (PTD | EN | M3),		(OFF_IN_PD) },
+	{ "uart3_rts_sd",	"-nc-",		0x142, (PTD | EN | M3),		(OFF_IN_PD) },
+	{ "uart3_rx_irrx",	"AP_FLM_RXD",	0x144, (IEN | M0),		(OFF_DIS) },
+	{ "uart3_tx_irtx",	"AP_FLM_TXD",	0x146, (M0),			(OFF_DIS) },
 
 	{ "sdmmc5_clk",  "WLAN_SD_CLK",		0x148, (IEN | M0),		(OFF_IN_PD) },
 	{ "sdmmc5_cmd",  "WLAN_SD_CMD",		0x14A, (IEN | PTU | EN | M0),	(OFF_DIS) },
@@ -226,58 +226,58 @@
 	{ "sdmmc5_dat2", "WLAN_SD_D(2)",	0x150, (IEN | PTU | EN | M0),	(OFF_DIS) },
 	{ "sdmmc5_dat3", "WLAN_SD_D(3)",	0x152, (IEN | PTU | EN | M0),	(OFF_DIS) },
 
-	//{ "mcspi4_clk",  "LCD_SCLK",		0x154, (),		() },
-	//{ "mcspi4_simo", "LCD_SDI",		0x156, (),		() },
-	{ "mcspi4_somi", "-nc-",		0x158, (IEN | PTD | EN | M3),		(OFF_IN_PD) },
-	//{ "mcspi4_cs0",  "LCD_nCS",		0x15A, (),		() },
-
-	//{ "uart4_rx",	"AP_RXD",		0x15C, (),		() },
-	//{ "uart4_tx",	"AP_TXD",		0x15E, (),		() },
-
-	//{ "usbb2_ulpitll_clk",	"MSENSE_IRQ",	0x160, (),		() },
-	//{ "usbb2_ulpitll_stp",	"LCD_D(23)",	0x162, (),		() },
-	//{ "usbb2_ulpitll_dir",	"LCD_D(22)",	0x164, (),		() },
-	//{ "usbb2_ulpitll_nxt",	"LCD_D(21)",	0x166, (),		() },
-	//{ "usbb2_ulpitll_dat0",	"LCD_D(20)",	0x168, (),		() },
-	//{ "usbb2_ulpitll_dat1",	"LCD_D(19)",	0x16A, (),		() },
-	//{ "usbb2_ulpitll_dat2",	"LCD_D(18)",	0x16C, (),		() },
-	//{ "usbb2_ulpitll_dat3",	"LCD_D(15)",	0x16E, (),		() },
-	//{ "usbb2_ulpitll_dat4",	"LCD_D(14)",	0x170, (),		() },
-	//{ "usbb2_ulpitll_dat5",	"LCD_D(13)",	0x172, (),		() },
-	//{ "usbb2_ulpitll_dat6",	"LCD_D(12)",	0x174, (),		() },
-	//{ "usbb2_ulpitll_dat7",	"LCD_D(11)",	0x176, (),		() },
+	{ "mcspi4_clk",  "LCD_SCLK",		0x154, (IEN | PTD | EN | M0),	(OFF_OUT_PD) },
+	{ "mcspi4_simo", "LCD_SDI",		0x156, (IEN | PTU | EN | M0),	(OFF_OUT_PU) },
+	{ "mcspi4_somi", "-nc-",		0x158, (IEN | PTD | EN | M3),	(OFF_IN_PD) },
+	{ "mcspi4_cs0",  "LCD_nCS",		0x15A, (IEN | PTU | EN | M0),	(OFF_OUT_PU) },
+
+	{ "uart4_rx",	"AP_RXD",		0x15C, (IEN | M0),	(OFF_IN_PD) },
+	{ "uart4_tx",	"AP_TXD",		0x15E, (M0),		(OFF_OUT_PD) },
+
+	{ "usbb2_ulpitll_clk",	"MSENSE_IRQ",	0x160, (IEN | M3),	(OFF_IN_PU) },
+	{ "usbb2_ulpitll_stp",	"LCD_D(23)",	0x162, (PTD | EN | M5),	(OFF_OUT_PD) },
+	{ "usbb2_ulpitll_dir",	"LCD_D(22)",	0x164, (PTD | EN | M5),	(OFF_OUT_PD) },
+	{ "usbb2_ulpitll_nxt",	"LCD_D(21)",	0x166, (PTD | EN | M5),	(OFF_OUT_PD) },
+	{ "usbb2_ulpitll_dat0",	"LCD_D(20)",	0x168, (PTD | EN | M5),	(OFF_OUT_PD) },
+	{ "usbb2_ulpitll_dat1",	"LCD_D(19)",	0x16A, (PTD | EN | M5),	(OFF_OUT_PD) },
+	{ "usbb2_ulpitll_dat2",	"LCD_D(18)",	0x16C, (PTD | EN | M5),	(OFF_OUT_PD) },
+	{ "usbb2_ulpitll_dat3",	"LCD_D(15)",	0x16E, (PTD | EN | M5),	(OFF_OUT_PD) },
+	{ "usbb2_ulpitll_dat4",	"LCD_D(14)",	0x170, (PTD | EN | M5),	(OFF_OUT_PD) },
+	{ "usbb2_ulpitll_dat5",	"LCD_D(13)",	0x172, (PTD | EN | M5),	(OFF_OUT_PD) },
+	{ "usbb2_ulpitll_dat6",	"LCD_D(12)",	0x174, (PTD | EN | M5),	(OFF_OUT_PD) },
+	{ "usbb2_ulpitll_dat7",	"LCD_D(11)",	0x176, (PTD | EN | M5),	(OFF_OUT_PD) },
 	//{ "usbb2_hsic_data",	"2MIC_EN",	0x178, (),		() },
-	//{ "usbb2_hsic_strobe",	"-nc-",		0x17A, (PTD | EN | M7),		(OFF_DIS) },
+	{ "usbb2_hsic_strobe",	"-nc-",		0x17A, (PTD | EN | M7),		(OFF_DIS) },
 
-	{ "kdp_col3",	"HW_REV3",	0x17C, (IEN | M3),		(OFF_IN_PD) },
-	{ "kdp_col4",	"HW_REV2",	0x17E, (IEN | M3),		(OFF_IN_PU) },
-	{ "kdp_col5",	"HW_REV0",	0x180, (IEN | M3),		(OFF_IN_PU) },
-	//{ "kdp_col0",	"-nc-",		0x182, (PTD | EN | M7),		(OFF_DIS) },
-	//{ "kdp_col1",	"KBC(1)",	0x184, (),		() },
-	//{ "kdp_col2",	"KBC(2)",	0x186, (),		() },
-	{ "kpd_row3",	"HW_REV1",	0x188, (IEN | M3),		(OFF_IN_PD) },
-	//XXX: { "kpd_row4",	"HOME_KEY",	0x18A, (),		() },
-	//{ "kpd_row5",	"-nc-",		0x18C, (PTD | EN | M7),		(OFF_DIS) },
-	//{ "kpd_row0",	"-nc-",		0x18E, (PTD | EN | M7),		(OFF_DIS) },
-	//{ "kpd_row1",	"KBR(1)",	0x190, (),		() },
-	//{ "kpd_row2",	"KBR(2)",	0x192, (),		() },
+	{ "kpd_col3",	"HW_REV3",	0x17C, (IEN | PTD | EN | M3),	(OFF_DIS) },
+	{ "kpd_col4",	"HW_REV2",	0x17E, (IEN | PTU | EN | M3),	(OFF_DIS) },
+	{ "kpd_col5",	"HW_REV0",	0x180, (IEN | PTU | EN | M3),	(OFF_DIS) },
+	{ "kpd_col0",	"-nc-",		0x182, (PTD | EN | M3),		(OFF_DIS) },
+	{ "kpd_col1",	"KBC(1)",	0x184, (PTU | EN | M0),		(OFF_IN_PD) },
+	{ "kpd_col2",	"KBC(2)",	0x186, (PTU | EN | M0),		(OFF_IN_PD) },
+	{ "kpd_row3",	"HW_REV1",	0x188, (IEN | PTD | EN | M3),	(OFF_DIS) },
+	//{ "kpd_row4",	"HOME_KEY",	0x18A, (IEN | PTU | EN | M3),	(OFF_DIS) },
+	{ "kpd_row5",	"-nc-",		0x18C, (PTD | EN | M3),		(OFF_DIS) },
+	{ "kpd_row0",	"-nc-",		0x18E, (PTD | EN | M3),		(OFF_DIS) },
+	//{ "kpd_row1",	"KBR(1)",	0x190, (IEN | M0),		(OFF_DIS) },
+	//{ "kpd_row2",	"KBR(2)",	0x192, (IEN | M0),		(OFF_IN_PD) },
 
 	//{ "usba0_otg_ce", "USB_CHGEN",	0x194, (),		() },
 	//{ "usba0_otg_dp", "AP_D+",	0x196, (),		() },
 	//{ "usba0_otg_dm", "AP_D-",	0x198, (),		() },
 
 	{ "fref_clk1_out", "8M_MCLK",	0x19A, (M0),		(OFF_OUT_PD) },
-	//{ "fref_clk2_out", "1.3M_MCLK",	0x19C, (),		() },
+	{ "fref_clk2_out", "1.3M_MCLK",	0x19C, (M0),		(OFF_OUT_PD) },
 
 	//{ "sys_nirq1",	"SYS_nIRQ1",	0x19E, (),		() },
 	//{ "sys_nirq2",	"SYS_nIRQ2",	0x1A0, (),		() },
 
-	//{ "sys_boot0",	"SYS_BOOT0",	0x1A2, (),		() },
-	//{ "sys_boot1",	"SYS_BOOT1",	0x1A4, (),		() },
-	//{ "sys_boot2",	"SYS_BOOT2",	0x1A6, (),		() },
-	//{ "sys_boot3",	"SYS_BOOT3",	0x1A8, (),		() },
-	//{ "sys_boot4",	"SYS_BOOT4",	0x1AA, (),		() },
-	//{ "sys_boot5",	"SYS_BOOT5",	0x1AC, (),		() },
+	{ "sys_boot0",	"SYS_BOOT0",	0x1A2, (IEN | M0),		(OFF_DIS) },
+	{ "sys_boot1",	"SYS_BOOT1",	0x1A4, (IEN | M0),		(OFF_DIS) },
+	{ "sys_boot2",	"SYS_BOOT2",	0x1A6, (IEN | M0),		(OFF_DIS) },
+	{ "sys_boot3",	"SYS_BOOT3",	0x1A8, (IEN | PTD | EN | M0),	(OFF_DIS) },
+	{ "sys_boot4",	"SYS_BOOT4",	0x1AA, (IEN | PTD | EN | M0),	(OFF_DIS) },
+	{ "sys_boot5",	"SYS_BOOT5",	0x1AC, (IEN | PTD | EN | M0),	(OFF_DIS) },
 
 	/* NOTE: Do not set OFF mode setting for 2MIC_RST and 2MIC_POWERDOWN.
 	 * Specific sequence is required to put the 2MIC in sleep mode. It is
@@ -285,24 +285,24 @@
 	{ "dpm_emu0",	"2MIC_RST",		0x1AE, (M3),	(OFF_DIS) },
 	{ "dpm_emu1",	"2MIC_POWERDOWN",	0x1B0, (M3),	(OFF_DIS) },
 
-	//{ "dpm_emu2",	"OLED_ID",	0x1B2, (),		() },
-	//{ "dpm_emu3",	"LCD_D(10)",	0x1B4, (),		() },
-	//{ "dpm_emu4",	"LCD_D(9)",	0x1B6, (),		() },
-	//{ "dpm_emu5",	"LCD_D(16)",	0x1B8, (),		() },
-	//{ "dpm_emu6",	"LCD_D(17)",	0x1BA, (),		() },
-	//{ "dpm_emu7",	"LCD_HSYNC",	0x1BC, (),		() },
-	//{ "dpm_emu8",	"LCD_PCLK",	0x1BE, (),		() },
-	//{ "dpm_emu9",	"LCD_VSYNC",	0x1C0, (),		() },
-	//{ "dpm_emu10",	"LCD_DE",	0x1C2, (),		() },
-	//{ "dpm_emu11",	"LCD_D(8)",	0x1C4, (),		() },
-	//{ "dpm_emu12",	"LCD_D(7)",	0x1C6, (),		() },
-	//{ "dpm_emu13",	"LCD_D(6)",	0x1C8, (),		() },
-	//{ "dpm_emu14",	"LCD_D(5)",	0x1CA, (),		() },
-	//{ "dpm_emu15",	"LCD_D(4)",	0x1CC, (),		() },
-	//{ "dpm_emu16",	"LCD_D(3)",	0x1CE, (),		() },
-	//{ "dpm_emu17",	"LCD_D(2)",	0x1D0, (),		() },
-	//{ "dpm_emu18",	"LCD_D(1)",	0x1D2, (),		() },
-	//{ "dpm_emu19",	"LCD_D(0)",	0x1D4, (),		() },
+	{ "dpm_emu2",	"OLED_ID",	0x1B2, (PTD | EN | M5),		(OFF_OUT_PD) },
+	{ "dpm_emu3",	"LCD_D(10)",	0x1B4, (PTD | EN | M5),		(OFF_OUT_PD) },
+	{ "dpm_emu4",	"LCD_D(9)",	0x1B6, (PTD | EN | M5),		(OFF_OUT_PD) },
+	{ "dpm_emu5",	"LCD_D(16)",	0x1B8, (PTD | EN | M5),		(OFF_OUT_PD) },
+	{ "dpm_emu6",	"LCD_D(17)",	0x1BA, (PTD | EN | M5),		(OFF_OUT_PD) },
+	{ "dpm_emu7",	"LCD_HSYNC",	0x1BC, (PTD | EN | M5),		(OFF_OUT_PD) },
+	{ "dpm_emu8",	"LCD_PCLK",	0x1BE, (PTD | EN | M5),		(OFF_OUT_PD) },
+	{ "dpm_emu9",	"LCD_VSYNC",	0x1C0, (PTD | EN | M5),		(OFF_OUT_PD) },
+	{ "dpm_emu10",	"LCD_DE",	0x1C2, (PTD | EN | M5),		(OFF_OUT_PD) },
+	{ "dpm_emu11",	"LCD_D(8)",	0x1C4, (PTD | EN | M5),		(OFF_OUT_PD) },
+	{ "dpm_emu12",	"LCD_D(7)",	0x1C6, (PTD | EN | M5),		(OFF_OUT_PD) },
+	{ "dpm_emu13",	"LCD_D(6)",	0x1C8, (PTD | EN | M5),		(OFF_OUT_PD) },
+	{ "dpm_emu14",	"LCD_D(5)",	0x1CA, (PTD | EN | M5),		(OFF_OUT_PD) },
+	{ "dpm_emu15",	"LCD_D(4)",	0x1CC, (PTD | EN | M5),		(OFF_OUT_PD) },
+	{ "dpm_emu16",	"LCD_D(3)",	0x1CE, (PTD | EN | M5),		(OFF_OUT_PD) },
+	{ "dpm_emu17",	"LCD_D(2)",	0x1D0, (PTD | EN | M5),		(OFF_OUT_PD) },
+	{ "dpm_emu18",	"LCD_D(1)",	0x1D2, (PTD | EN | M5),		(OFF_OUT_PD) },
+	{ "dpm_emu19",	"LCD_D(0)",	0x1D4, (PTD | EN | M5),		(OFF_OUT_PD) },
 };
 
 void set_offmode_padconfig(void)
@@ -328,12 +328,17 @@
 	omap_writel(0x80000, 0x4A31E050);
 	/* sys_pwr_req */ omap_writew(0x0000, 0x4A31E064);
 
+	/* sys_pwron_reset_out */ omap_writew(0x010B,0x4A31E066);
+	/* jtag_ntrst and jtag_tck */ omap_writel(0x01080108,0x4A31E06C);
+	/* jtag_tms and jtag_rtck */ omap_writel(0x01000008,0x4A31E070);
+	/* jtag_tdi and jtag_tdo */ omap_writel(0x01180118, 0x4A31E074);
+
 	/* SYSCTRL_PADCONF_WKUP Registers */
 	/* sim_io */ omap_writew((PTD | EN | M7), 0x4A31E040);
 	/* sim_clk */ omap_writew((M3), 0x4A31E042);
 	/* sim_pwrctrl */ omap_writew((PTD | EN | M7), 0x4A31E048);
-	/* fref_clk3_req */ omap_writew((PTD | EN | M7), 0x4A31E056);
-	/* fref_clk4_req */ omap_writew((PTD | EN | M0), 0x4A31E05A);
+	/* fref_clk3_req */ omap_writew((PTD | EN | M3), 0x4A31E056);
+	/* fref_clk4_req */ omap_writew((PTD | EN | M3), 0x4A31E05A);
 	/* fref_clk4_out */ omap_writew((PTD | EN | M3), 0x4A31E05C);
 	/* AUXCLK3_SCRM, clock is disabled by software */
 	reg_val32 = omap_readl(0x4A30A31C);
