Index: Kernel/arch/arm/plat-omap/hdmi_lib.c
===================================================================
--- Kernel.orig/arch/arm/plat-omap/hdmi_lib.c	2011-02-23 16:57:11.844495132 +0530
+++ Kernel/arch/arm/plat-omap/hdmi_lib.c	2011-02-23 17:07:59.236584988 +0530
@@ -649,7 +649,7 @@
 	u8 acr_en;
 
 #ifdef CONFIG_OMAP_HDMI_AUDIO_WA
-	if (omap_chip_is(hdmi.audio_wa_chip_ids))
+	if (!omap_chip_is(hdmi.audio_wa_chip_ids))
 		acr_en = 0;
 	else
 		acr_en = 1;
@@ -1367,7 +1367,7 @@
 static void hdmi_w1_audio_enable(void)
 {
 #ifdef CONFIG_OMAP_HDMI_AUDIO_WA
-	if (omap_chip_is(hdmi.audio_wa_chip_ids)) {
+	if (!omap_chip_is(hdmi.audio_wa_chip_ids)) {
 		if (hdmi.notify_event_reg == HDMI_NOTIFY_EVENT_REG)
 			notify_send_event(SYS_M3, 0, HDMI_AUDIO_WA_EVENT,
 					hdmi.cts_interval, 0);
@@ -1381,7 +1381,7 @@
 static void hdmi_w1_audio_disable(void)
 {
 #ifdef CONFIG_OMAP_HDMI_AUDIO_WA
-	if (omap_chip_is(hdmi.audio_wa_chip_ids)) {
+	if (!omap_chip_is(hdmi.audio_wa_chip_ids)) {
 		/* Payload=0 disables workaround */
 		if (hdmi.notify_event_reg == HDMI_NOTIFY_EVENT_REG)
 			notify_send_event(SYS_M3, 0, HDMI_AUDIO_WA_EVENT, 0, 0);
