Index: Kernel/arch/arm/mach-omap2/omap4-iopad.c
===================================================================
--- Kernel.orig/arch/arm/mach-omap2/omap4-iopad.c	2011-02-23 15:56:55.182449349 +0530
+++ Kernel/arch/arm/mach-omap2/omap4-iopad.c	2011-02-23 18:16:38.292448972 +0530
@@ -142,14 +142,14 @@
 	{ "cam_strobe",		"-nc-",		0x0BE, (PTD | EN | M7),		(OFF_DIS) },
 	{ "cam_globalreset",	"-nc-",		0x0C0, (PTD | EN | M7),		(OFF_DIS) },
 
-	//{ "usbb1_ulpitll_clk",	"MIPI_HSI_TX_WAKE",	0x0C2, (),		() },
-	//{ "usbb1_ulpitll_stp",	"MIPI_HSI_TX_DATA",	0x0C4, (),		() },
-	//{ "usbb1_ulpitll_dir",	"MIPI_HSI_TX_FLG",	0x0C6, (),		() },
-	//{ "usbb1_ulpitll_nxt",	"MIPI_HSI_TX_RDY",	0x0C8, (),		() },
-	//{ "usbb1_ulpitll_dat0",	"MIPI_HSI_RX_WAKE",	0x0CA, (),		() },
-	//{ "usbb1_ulpitll_dat1",	"MIPI_HSI_RX_DATA",	0x0CC, (),		() },
-	//{ "usbb1_ulpitll_dat2",	"MIPI_HSI_RX_FLG",	0x0CE, (),		() },
-	//{ "usbb1_ulpitll_dat3",	"MIPI_HSI_RX_RDY",	0x0D0, (),		() },
+	{ "usbb1_ulpitll_clk",	"MIPI_HSI_TX_WAKE",	0x0C2, (IEN | M1),	(OFF_DIS) },
+	{ "usbb1_ulpitll_stp",	"MIPI_HSI_TX_DATA",	0x0C4, (IEN | M1),	(OFF_DIS) },
+	{ "usbb1_ulpitll_dir",	"MIPI_HSI_TX_FLG",	0x0C6, (IEN | M1),	(OFF_DIS) },
+	{ "usbb1_ulpitll_nxt",	"MIPI_HSI_TX_RDY",	0x0C8, (M1),		(OFF_DIS) },
+	{ "usbb1_ulpitll_dat0",	"MIPI_HSI_RX_WAKE",	0x0CA, (M1),		(OFF_DIS) },
+	{ "usbb1_ulpitll_dat1",	"MIPI_HSI_RX_DATA",	0x0CC, (M1),		(OFF_DIS) },
+	{ "usbb1_ulpitll_dat2",	"MIPI_HSI_RX_FLG",	0x0CE, (M1),		(OFF_DIS) },
+	{ "usbb1_ulpitll_dat3",	"MIPI_HSI_RX_RDY",	0x0D0, (IEN | M1),	(OFF_DIS) },
 	//{ "usbb1_ulpitll_dat4",	"BOOT_MODE",		0x0D2, (),		() },
 	{ "usbb1_ulpitll_dat5",	"-nc-",		0x0D4, (PTD | EN | M7),		(OFF_DIS) },
 	//{ "usbb1_ulpitll_dat6",	"EAR_SEND_END",		0x0D6, (),		() },
@@ -198,8 +198,8 @@
 
 	//{ "i2c1_scl",	"PHEONIX_I2C_SCL",	0x122, (),		() },
 	//{ "i2c1_sda",	"PHEONIX_I2C_SDA",	0x124, (),		() },
-	{ "i2c2_scl",	"CAM_I2C_SCL",		0x126, (IEN | M0),	(OFF_IN_PD) },
-	{ "i2c2_sda",	"CAM_I2C_SDA",		0x128, (IEN | M0),	(OFF_IN_PD) },
+	{ "i2c2_scl",	"CAM_I2C_SCL",		0x126, (IEN | M0),	(OFF_DIS) },
+	{ "i2c2_sda",	"CAM_I2C_SDA",		0x128, (IEN | M0),	(OFF_DIS) },
 	/* NOTE: Fuel gauge expects I2C SCL and SDA lines to be low to enter sleep mode. This save ~70uA */
 	{ "i2c3_scl",	"AP_I2C_SCL",		0x12A, (IEN | PTU | EN | M0),	(OFF_IN_PD) },
 	{ "i2c3_sda",	"AP_I2C_SDA",		0x12C, (IEN | PTU | EN | M0),	(OFF_IN_PD) },
@@ -253,8 +253,8 @@
 	{ "kpd_col4",	"HW_REV2",	0x17E, (IEN | PTU | EN | M3),	(OFF_DIS) },
 	{ "kpd_col5",	"HW_REV0",	0x180, (IEN | PTU | EN | M3),	(OFF_DIS) },
 	{ "kpd_col0",	"-nc-",		0x182, (PTD | EN | M3),		(OFF_DIS) },
-	{ "kpd_col1",	"KBC(1)",	0x184, (PTU | EN | M0),		(OFF_IN_PD) },
-	{ "kpd_col2",	"KBC(2)",	0x186, (PTU | EN | M0),		(OFF_IN_PD) },
+	{ "kpd_col1",	"KBC(1)",	0x184, (M0),		(OFF_OUT_PD) },
+	{ "kpd_col2",	"KBC(2)",	0x186, (M0),		(OFF_OUT_PD) },
 	{ "kpd_row3",	"HW_REV1",	0x188, (IEN | PTD | EN | M3),	(OFF_DIS) },
 	//{ "kpd_row4",	"HOME_KEY",	0x18A, (IEN | PTU | EN | M3),	(OFF_DIS) },
 	{ "kpd_row5",	"-nc-",		0x18C, (PTD | EN | M3),		(OFF_DIS) },
Index: Kernel/drivers/staging/omap_hsi/hsi_protocol.c
===================================================================
--- Kernel.orig/drivers/staging/omap_hsi/hsi_protocol.c	2011-02-22 10:58:55.000000000 +0530
+++ Kernel/drivers/staging/omap_hsi/hsi_protocol.c	2011-02-23 16:33:13.132448392 +0530
@@ -271,36 +271,6 @@
 }
 //EXPORT_SYMBOL(hsi_proto_write);
 
-#define OMAP4_PAD_CAM_GLOBALRESET_USBB1_ULPITLL_CLK		0x4A1000C0 // ???, MIPI_HSI_TX_WAKE
-#define OMAP4_PAD_USBB1_ULPITLL_STP_USBB1_ULPITLL_DIR		0x4A1000C4 // MIPI_HSI_TX_DATA, MIPI_HSI_TX_FLG
-#define OMAP4_PAD_USBB1_ULPITLL_NXT_USBB1_ULPITLL_DAT0		0x4A1000C8 // MIPI_HSI_TX_RDY, MIPI_HSI_RX_WAKE
-#define OMAP4_PAD_USBB1_ULPITLL_DAT1_USBB1_ULPITLL_DAT2		0x4A1000CC // MIPI_HSI_RX_DATA, MIPI_HSI_RX_FLG
-#define OMAP4_PAD_USBB1_ULPITLL_DAT3_USBB1_ULPITLL_DAT4	0x4A1000D0 // MIPI_HSI_RX_RDY, ???
-
-static void hsi_mux_setting( void )
-{
-	u32 reg_val;
-
-	reg_val = omap_readl( OMAP4_PAD_CAM_GLOBALRESET_USBB1_ULPITLL_CLK );
-	reg_val &= 0x0000FFFF;
-	reg_val |= 0x01010000;
-	omap_writel( reg_val, OMAP4_PAD_CAM_GLOBALRESET_USBB1_ULPITLL_CLK );
-
-	reg_val = 0x01010101;
-	omap_writel( reg_val, OMAP4_PAD_USBB1_ULPITLL_STP_USBB1_ULPITLL_DIR );
-
-	reg_val = 0x00010001;
-	omap_writel( reg_val, OMAP4_PAD_USBB1_ULPITLL_NXT_USBB1_ULPITLL_DAT0 );
-
-	reg_val = 0x00010001;
-	omap_writel( reg_val, OMAP4_PAD_USBB1_ULPITLL_DAT1_USBB1_ULPITLL_DAT2 );
-
-	reg_val = omap_readl( OMAP4_PAD_USBB1_ULPITLL_DAT3_USBB1_ULPITLL_DAT4 );
-	reg_val &= 0xFFFF0000;
-	reg_val |= 0x00000101;
-	omap_writel( reg_val, OMAP4_PAD_USBB1_ULPITLL_DAT3_USBB1_ULPITLL_DAT4 );
-}
-
 static int __init hsi_protocol_init(void)
 {
 	int i,ret = 0;
@@ -317,7 +287,6 @@
                 INIT_LIST_HEAD(&hsi_protocol_data[i].tx_queue);
         }
 
-	hsi_mux_setting();
 	printk( "hsi_protocol_init : hsi_mux_setting Done.\n" );
 
 	ret = if_hsi_init();
