Index: Kernel/arch/arm/mach-omap2/board-omap4tab10-1.c
===================================================================
--- Kernel.orig/arch/arm/mach-omap2/board-omap4tab10-1.c	2011-02-23 21:39:07.040988925 +0530
+++ Kernel/arch/arm/mach-omap2/board-omap4tab10-1.c	2011-02-23 21:42:11.419741525 +0530
@@ -395,7 +395,9 @@
 
 void keyboard_mux_init(void)
 {
-#if (CONFIG_SAMSUNG_OMAP4_TAB_REV == 4 || CONFIG_SAMSUNG_OMAP4_TAB_REV == 5)
+#if (CONFIG_SAMSUNG_OMAP4_TAB_REV == 4 || \
+	CONFIG_SAMSUNG_OMAP4_TAB_REV == 5 || \
+	CONFIG_SAMSUNG_OMAP4_TAB_REV == 6)
         omap_mux_init_signal("kpd_col1.kpd_col1",
                                 OMAP_WAKEUP_EN | OMAP_MUX_MODE0);
         omap_mux_init_signal("kpd_row1.kpd_row1",
@@ -1622,12 +1624,14 @@
 };
 #endif
 static struct i2c_board_info __initdata omap4tab_i2c_3_boardinfo[] = {
-#if (CONFIG_SAMSUNG_OMAP4_TAB_REV == 4 || CONFIG_SAMSUNG_OMAP4_TAB_REV == 5)
+#if (CONFIG_SAMSUNG_OMAP4_TAB_REV == 4 || \
+	CONFIG_SAMSUNG_OMAP4_TAB_REV == 5 || \
+	CONFIG_SAMSUNG_OMAP4_TAB_REV == 6)
 	{
 		I2C_BOARD_INFO("cypress_touchkey", 0x20),
 	},
 #endif
-#if (CONFIG_SAMSUNG_OMAP4_TAB_REV == 5)
+#if (CONFIG_SAMSUNG_OMAP4_TAB_REV == 5 || CONFIG_SAMSUNG_OMAP4_TAB_REV ==6)
 	{
 		I2C_BOARD_INFO("max17040", 0x36),
  	},
@@ -1639,7 +1643,7 @@
 		.platform_data = &mxt1386_ts_platform_data,
 	},
 #endif
-#if (CONFIG_SAMSUNG_OMAP4_TAB_REV == 5)
+#if (CONFIG_SAMSUNG_OMAP4_TAB_REV == 5 || CONFIG_SAMSUNG_OMAP4_TAB_REV ==6)
 	{
 		I2C_BOARD_INFO("A1026", 0x3E),
 	},
Index: Kernel/arch/arm/configs/android_omap4tab_rev06_defconfig
===================================================================
--- Kernel.orig/arch/arm/configs/android_omap4tab_rev06_defconfig	2011-02-23 22:38:44.210995592 +0530
+++ Kernel/arch/arm/configs/android_omap4tab_rev06_defconfig	2011-02-23 22:39:52.900995638 +0530
@@ -1,7 +1,7 @@
 #
 # Automatically generated make config: don't edit
 # Linux kernel version: 2.6.35.7
-# Wed Feb 23 19:28:37 2011
+# Wed Feb 23 22:39:18 2011
 #
 CONFIG_ARM=y
 CONFIG_SYS_SUPPORTS_APM_EMULATION=y
@@ -1560,7 +1560,7 @@
 # CONFIG_SND_SOC_ALL_CODECS is not set
 CONFIG_SND_SOC_DMIC=y
 CONFIG_SND_SOC_TWL6040=y
-CONFIG_SND_SOC_MIC_A1026=y
+# CONFIG_SND_SOC_MIC_A1026 is not set
 # CONFIG_SOUND_PRIME is not set
 # CONFIG_HID_SUPPORT is not set
 CONFIG_HID=y
