Index: Kernel/drivers/media/video/omap/Kconfig
===================================================================
--- Kernel.orig/drivers/media/video/omap/Kconfig	2011-02-24 13:16:27.292129102 +0530
+++ Kernel/drivers/media/video/omap/Kconfig	2011-02-24 13:16:54.712572636 +0530
@@ -16,4 +16,4 @@
 	default y if VIDEO_OMAP2_VOUT
 	help
 	  VRFB used in V4L2 in OMAP2,3
-source "drivers/media/video/samsung/mhl_v1/Kconfig"
+source "drivers/media/video/omap/mhl_v1/Kconfig"
