Index: Kernel/drivers/media/video/omap/mhl_v1/Common_Def.h
===================================================================
--- Kernel.orig/drivers/media/video/omap/mhl_v1/Common_Def.h	2011-02-24 16:13:33.632130778 +0530
+++ Kernel/drivers/media/video/omap/mhl_v1/Common_Def.h	2011-02-24 16:21:44.743379598 +0530
@@ -153,6 +153,35 @@
 #define		MHL_SCRATCHPAD_SIZE				16
 #define		MHL_MAX_BUFFER_SIZE				MHL_SCRATCHPAD_SIZE	// manually define highest number
 
+//GPIO Pin configuration of Sii9234
+#define GPIO_MHL_RST    60
+#define GPIO_MHL_INT    175
+#define GPIO_MHL_SEL    170
+#define GPIO_MHL_WAKE_UP        64
+
+#define GPIO_MHL_INT_AF
+#define GPIO_MHL_WAKE_UP_AF    
+
+#define GPIO_MSENSOR_MHL_SDA_28V    
+#define GPIO_MSENSOR_MHL_SDA_AF                 0x3
+#define GPIO_MSENSOR_MHL_SCL_28V
+#define GPIO_MSENSOR_MHL_SCL_AF                 0x3
+#define GPIO_MHL_SDA_18V            
+#define GPIO_MHL_SCL_18V           
+#define GPIO_AP_HDMI_SDA
+#define GPIO_AP_HDMI_SDA_AF
+#define GPIO_AP_HDMI_SCL     GPIO_MSENSOR_MHL_SCL_28V
+#define GPIO_AP_HDMI_SCL_AF     GPIO_MSENSOR_MHL_SCL_AF
+#define GPIO_AP_SDA_18V              GPIO_MHL_SDA_18V
+#define GPIO_AP_SCL_18V              GPIO_MHL_SCL_18V
+#define MHL_INT_IRQ          gpio_to_irq(GPIO_MHL_INT)
+#define MHL_WAKEUP_IRQ               gpio_to_irq(GPIO_MHL_WAKE_UP)
+
+#define GPIO_HDMI_EN    100
+#define GPIO_HDMI_HPD   63
+#define GPIO_HDMI_EN_REV07  GPIO_HDMI_EN 
+#define GPIO_LEVEL_LOW  0
+#define GPIO_LEVEL_HIGH 1
 
 
 enum
Index: Kernel/drivers/media/video/omap/mhl_v1/SiI9234_I2C_master.c
===================================================================
--- Kernel.orig/drivers/media/video/omap/mhl_v1/SiI9234_I2C_master.c	2011-02-24 16:13:09.372754166 +0530
+++ Kernel/drivers/media/video/omap/mhl_v1/SiI9234_I2C_master.c	2011-02-24 16:22:40.112128953 +0530
@@ -26,12 +26,12 @@
 #include <linux/irq.h>
 #include <linux/i2c.h>
 #include <linux/gpio.h>
-#include <plat/pm.h>
+//#include <plat/pm.h>
 #include <asm/irq.h>
 #include <linux/delay.h>
-#include <plat/gpio-cfg.h>
-#include <mach/regs-gpio.h>
-#include <mach/regs-clock.h>
+//#include <plat/gpio-cfg.h>
+//#include <mach/regs-gpio.h>
+//#include <mach/regs-clock.h>
 
 #include <linux/syscalls.h> 
 #include <linux/fcntl.h> 
Index: Kernel/drivers/media/video/omap/mhl_v1/si_apiCbus.c
===================================================================
--- Kernel.orig/drivers/media/video/omap/mhl_v1/si_apiCbus.c	2011-02-24 16:12:48.003379471 +0530
+++ Kernel/drivers/media/video/omap/mhl_v1/si_apiCbus.c	2011-02-24 16:23:04.592754225 +0530
@@ -15,12 +15,12 @@
 #include <linux/irq.h>
 #include <linux/i2c.h>
 #include <linux/gpio.h>
-#include <plat/pm.h>
+//#include <plat/pm.h>
 #include <asm/irq.h>
 #include <linux/delay.h>
-#include <plat/gpio-cfg.h>
-#include <mach/regs-gpio.h>
-#include <mach/regs-clock.h>
+//#include <plat/gpio-cfg.h>
+//#include <mach/regs-gpio.h>
+//#include <mach/regs-clock.h>
 
 #include <linux/syscalls.h> 
 #include <linux/fcntl.h> 
Index: Kernel/drivers/media/video/omap/mhl_v1/si_cpCbus.c
===================================================================
--- Kernel.orig/drivers/media/video/omap/mhl_v1/si_cpCbus.c	2011-02-24 16:13:24.193380145 +0530
+++ Kernel/drivers/media/video/omap/mhl_v1/si_cpCbus.c	2011-02-24 16:23:25.562129650 +0530
@@ -14,12 +14,12 @@
 #include <linux/irq.h>
 #include <linux/i2c.h>
 #include <linux/gpio.h>
-#include <plat/pm.h>
+//#include <plat/pm.h>
 #include <asm/irq.h>
 #include <linux/delay.h>
-#include <plat/gpio-cfg.h>
-#include <mach/regs-gpio.h>
-#include <mach/regs-clock.h>
+//#include <plat/gpio-cfg.h>
+//#include <mach/regs-gpio.h>
+//#include <mach/regs-clock.h>
 
 #include <linux/syscalls.h> 
 #include <linux/fcntl.h> 
Index: Kernel/drivers/media/video/omap/mhl_v1/sii9234.c
===================================================================
--- Kernel.orig/drivers/media/video/omap/mhl_v1/sii9234.c	2011-02-24 16:12:25.182754458 +0530
+++ Kernel/drivers/media/video/omap/mhl_v1/sii9234.c	2011-02-24 16:23:47.072720866 +0530
@@ -30,21 +30,21 @@
 #include <linux/irq.h>
 #include <linux/i2c.h>
 #include <linux/gpio.h>
-#include <plat/pm.h>
+//#include <plat/pm.h>
 #include <asm/irq.h>
 #include <linux/delay.h>
-#include <plat/gpio-cfg.h>
-#include <plat/gpio-core.h>
-#include <mach/regs-gpio.h>
-#include <mach/regs-clock.h>
+//#include <plat/gpio-cfg.h>
+//#include <plat/gpio-core.h>
+//#include <mach/regs-gpio.h>
+//#include <mach/regs-clock.h>
 #include <linux/miscdevice.h>
 #include <linux/slab.h>
 #include <linux/syscalls.h> 
 #include <linux/fcntl.h> 
 #include <asm/uaccess.h> 
 #include <linux/regulator/consumer.h>
-#include <linux/mfd/max8998.h>
-#include <linux/mfd/max8997.h>
+//#include <linux/mfd/max8998.h>
+//#include <linux/mfd/max8997.h>
 
 
 #include "sii9234_driver.h"
@@ -56,6 +56,7 @@
 #define SII_DEV_DBG(format,...)\
 	printk ("[ "SUBJECT " (%s,%d) ] " format "\n", __func__, __LINE__, ## __VA_ARGS__);
 
+
 struct work_struct sii9234_int_work;
 struct workqueue_struct *sii9234_wq = NULL;
 
@@ -145,14 +146,14 @@
 			printk(KERN_ERR "[MHL] try %d\n", i+1);
 			msleep(500);
 		}
-		s3c_gpio_cfgpin(GPIO_MHL_INT, GPIO_MHL_INT_AF);
-		s3c_gpio_setpull(GPIO_MHL_SEL, S3C_GPIO_PULL_UP);
+		//s3c_gpio_cfgpin(GPIO_MHL_INT, GPIO_MHL_INT_AF);
+		//s3c_gpio_setpull(GPIO_MHL_SEL, S3C_GPIO_PULL_UP);
 		gpio_set_value(GPIO_MHL_SEL, GPIO_LEVEL_HIGH);
 		sii9234_cfg_power(1);
 		sii9234_init();
 	} else {
 		sii9234_cfg_power(0);
-		s3c_gpio_setpull(GPIO_MHL_SEL, S3C_GPIO_PULL_NONE);
+		//s3c_gpio_setpull(GPIO_MHL_SEL, S3C_GPIO_PULL_NONE);
 		gpio_set_value(GPIO_MHL_SEL, GPIO_LEVEL_LOW);
 	}
 	return size;
@@ -161,12 +162,12 @@
 void MHL_On(bool on)
 {
 	if (on == 1) {
-		s3c_gpio_cfgpin(GPIO_MHL_INT, GPIO_MHL_INT_AF);
+		//s3c_gpio_cfgpin(GPIO_MHL_INT, GPIO_MHL_INT_AF);
 		sii9234_cfg_power(1);
 		sii9234_init();
 	} else {
 		sii9234_cfg_power(0);
-		s3c_gpio_setpull(GPIO_MHL_SEL, S3C_GPIO_PULL_NONE);
+		//s3c_gpio_setpull(GPIO_MHL_SEL, S3C_GPIO_PULL_NONE);
 		gpio_set_value(GPIO_MHL_SEL, GPIO_LEVEL_LOW);
 	}
 
@@ -636,15 +637,15 @@
 			reg_en = 1;
 		}
 		
-		s3c_gpio_cfgpin(GPIO_MHL_RST,S3C_GPIO_OUTPUT);	//MHL_RST
-		s3c_gpio_setpull(GPIO_MHL_RST, S3C_GPIO_PULL_NONE);
+		//s3c_gpio_cfgpin(GPIO_MHL_RST,S3C_GPIO_OUTPUT);	//MHL_RST
+		//s3c_gpio_setpull(GPIO_MHL_RST, S3C_GPIO_PULL_NONE);
 	
 
-		s3c_gpio_setpull(GPIO_AP_SCL_18V, S3C_GPIO_PULL_DOWN);
+		//s3c_gpio_setpull(GPIO_AP_SCL_18V, S3C_GPIO_PULL_DOWN);
 		gpio_set_value(GPIO_MHL_RST, GPIO_LEVEL_LOW);
 		msleep(50);
 		gpio_set_value(GPIO_MHL_RST, GPIO_LEVEL_HIGH);
-		s3c_gpio_setpull(GPIO_AP_SCL_18V, S3C_GPIO_PULL_NONE);
+		//s3c_gpio_setpull(GPIO_AP_SCL_18V, S3C_GPIO_PULL_NONE);
 	}
 	else
 	{
@@ -669,37 +670,37 @@
 
 static void sii9234_cfg_gpio()
 {
-	s3c_gpio_cfgpin(GPIO_AP_SDA_18V, S3C_GPIO_SFN(0x0));	//AP_MHL_SDA
-	s3c_gpio_setpull(GPIO_AP_SDA_18V, S3C_GPIO_PULL_NONE);
+	//s3c_gpio_cfgpin(GPIO_AP_SDA_18V, S3C_GPIO_SFN(0x0));	//AP_MHL_SDA
+	//s3c_gpio_setpull(GPIO_AP_SDA_18V, S3C_GPIO_PULL_NONE);
 
-	s3c_gpio_cfgpin(GPIO_AP_SCL_18V,S3C_GPIO_SFN(0x1));	//AP_MHL_SCL
-	s3c_gpio_setpull(GPIO_AP_SCL_18V, S3C_GPIO_PULL_NONE);
+	//s3c_gpio_cfgpin(GPIO_AP_SCL_18V,S3C_GPIO_SFN(0x1));	//AP_MHL_SCL
+	//s3c_gpio_setpull(GPIO_AP_SCL_18V, S3C_GPIO_PULL_NONE);
 
-	s3c_gpio_cfgpin(GPIO_MHL_WAKE_UP,S3C_GPIO_INPUT);//GPH1(6) XEINT 14
+	//s3c_gpio_cfgpin(GPIO_MHL_WAKE_UP,S3C_GPIO_INPUT);//GPH1(6) XEINT 14
 	set_irq_type(MHL_WAKEUP_IRQ, IRQ_TYPE_EDGE_RISING);
-	s3c_gpio_setpull(GPIO_MHL_WAKE_UP, S3C_GPIO_PULL_DOWN);
+	//s3c_gpio_setpull(GPIO_MHL_WAKE_UP, S3C_GPIO_PULL_DOWN);
 
-	s3c_gpio_setpull(GPIO_MHL_INT, S3C_GPIO_PULL_DOWN);
+	//s3c_gpio_setpull(GPIO_MHL_INT, S3C_GPIO_PULL_DOWN);
 	set_irq_type(MHL_INT_IRQ, IRQ_TYPE_EDGE_RISING);
 //	s3c_gpio_cfgpin(GPIO_MHL_INT, GPIO_MHL_INT_AF); 
 
 	if(system_rev < 7) {
-		s3c_gpio_cfgpin(GPIO_HDMI_EN,S3C_GPIO_OUTPUT);	//HDMI_EN
+		//s3c_gpio_cfgpin(GPIO_HDMI_EN,S3C_GPIO_OUTPUT);	//HDMI_EN
 		gpio_set_value(GPIO_HDMI_EN,GPIO_LEVEL_LOW);
-		s3c_gpio_setpull(GPIO_HDMI_EN, S3C_GPIO_PULL_NONE);
+		//s3c_gpio_setpull(GPIO_HDMI_EN, S3C_GPIO_PULL_NONE);
 	}
 	else {
-		s3c_gpio_cfgpin(GPIO_HDMI_EN_REV07,S3C_GPIO_OUTPUT);	//HDMI_EN
+		//s3c_gpio_cfgpin(GPIO_HDMI_EN_REV07,S3C_GPIO_OUTPUT);	//HDMI_EN
 		gpio_set_value(GPIO_HDMI_EN_REV07,GPIO_LEVEL_LOW);
-		s3c_gpio_setpull(GPIO_HDMI_EN_REV07, S3C_GPIO_PULL_NONE);
+		//s3c_gpio_setpull(GPIO_HDMI_EN_REV07, S3C_GPIO_PULL_NONE);
 	}
 
-	s3c_gpio_cfgpin(GPIO_MHL_RST,S3C_GPIO_OUTPUT);
-	s3c_gpio_setpull(GPIO_MHL_RST, S3C_GPIO_PULL_NONE);
+	//s3c_gpio_cfgpin(GPIO_MHL_RST,S3C_GPIO_OUTPUT);
+	//s3c_gpio_setpull(GPIO_MHL_RST, S3C_GPIO_PULL_NONE);
 	gpio_set_value(GPIO_MHL_RST, GPIO_LEVEL_LOW);
 
-	s3c_gpio_cfgpin(GPIO_MHL_SEL,S3C_GPIO_OUTPUT);
-	s3c_gpio_setpull(GPIO_MHL_SEL, S3C_GPIO_PULL_NONE);
+	//s3c_gpio_cfgpin(GPIO_MHL_SEL,S3C_GPIO_OUTPUT);
+	//s3c_gpio_setpull(GPIO_MHL_SEL, S3C_GPIO_PULL_NONE);
 	gpio_set_value(GPIO_MHL_SEL, GPIO_LEVEL_LOW);
 
 }
Index: Kernel/drivers/media/video/omap/mhl_v1/sii9234_driver.c
===================================================================
--- Kernel.orig/drivers/media/video/omap/mhl_v1/sii9234_driver.c	2011-02-24 16:12:07.953379603 +0530
+++ Kernel/drivers/media/video/omap/mhl_v1/sii9234_driver.c	2011-02-24 16:24:08.792129441 +0530
@@ -33,17 +33,17 @@
 #include <linux/irq.h>
 #include <linux/i2c.h>
 #include <linux/gpio.h>
-#include <plat/pm.h>
+//#include <plat/pm.h>
 #include <asm/irq.h>
 #include <linux/delay.h>
-#include <plat/gpio-cfg.h>
-#include <mach/regs-gpio.h>
-#include <mach/regs-clock.h>
+//#include <plat/gpio-cfg.h>
+//#include <mach/regs-gpio.h>
+//#include <mach/regs-clock.h>
 
 #include <linux/syscalls.h> 
 #include <linux/fcntl.h> 
 #include <asm/uaccess.h> 
-#include <plat/gpio-core.h>
+//#include <plat/gpio-core.h>
 
 
 #include "SiI9234_Reg.h"
@@ -319,7 +319,7 @@
 
 void	ProcessRgnd( void );
 /* jinho96.kim 2011.01.06 */
-extern void rcp_cbus_uevent(u8);		//Subhransu
+//extern void rcp_cbus_uevent(u8);		//Subhransu
 
 
 
@@ -366,7 +366,7 @@
 {
 	printk(KERN_ERR ">>TxHW_Reset()\n");
 
-	s3c_gpio_setpull(GPIO_MHL_RST, S3C_GPIO_PULL_NONE);
+	//s3c_gpio_setpull(GPIO_MHL_RST, S3C_GPIO_PULL_NONE);
 	gpio_set_value(GPIO_MHL_RST, GPIO_LEVEL_LOW);
 	//PinTxHwReset = LOW;
 	DelayMS(TX_HW_RESET_PERIOD);
@@ -770,7 +770,7 @@
 #endif
 					sii9234_cfg_power(0);
 					//s3c_gpio_setpull(GPIO_MHL_SEL, S3C_GPIO_PULL_DOWN);
-					s3c_gpio_cfgpin(GPIO_MHL_INT, S3C_GPIO_INPUT); 
+					//s3c_gpio_cfgpin(GPIO_MHL_INT, S3C_GPIO_INPUT); 
 					gpio_set_value(GPIO_MHL_SEL, GPIO_LEVEL_LOW);
 					printk(KERN_ERR "MHL_SEL to 0\n");
          			}
@@ -798,7 +798,7 @@
 				mhl_cable_status =MHL_INIT_POWER_OFF;
 				sii9234_cfg_power(0);
 				//s3c_gpio_setpull(GPIO_MHL_SEL, S3C_GPIO_PULL_DOWN);
-				s3c_gpio_cfgpin(GPIO_MHL_INT, S3C_GPIO_INPUT); 
+				//s3c_gpio_cfgpin(GPIO_MHL_INT, S3C_GPIO_INPUT); 
 				gpio_set_value(GPIO_MHL_SEL, GPIO_LEVEL_LOW);
 				printk(KERN_ERR "MHL_SEL to 0\n");
 			}
@@ -2006,7 +2006,7 @@
 				*eventParameter = mhlTxConfig.mscMsgData; // key code
 #if 1	// xmoondash :: compile error
 /* jinho96.kim 2011.01.06 */
-				rcp_cbus_uevent(*eventParameter);	//MHL v1 //NAGSM_Android_SEL_Kernel_Aakash_20101126
+				//rcp_cbus_uevent(*eventParameter);	//MHL v1 //NAGSM_Android_SEL_Kernel_Aakash_20101126
 #endif
 
 				printk(KERN_ERR "Key Code:%x \n",(int)mhlTxConfig.mscMsgData);
