Index: Kernel/arch/arm/mach-omap2/omap4-iopad.c
===================================================================
--- Kernel.orig/arch/arm/mach-omap2/omap4-iopad.c	2011-02-24 19:55:53.000000000 +0530
+++ Kernel/arch/arm/mach-omap2/omap4-iopad.c	2011-02-24 20:03:56.073363618 +0530
@@ -198,8 +198,8 @@
 
 	//{ "i2c1_scl",	"PHEONIX_I2C_SCL",	0x122, (),		() },
 	//{ "i2c1_sda",	"PHEONIX_I2C_SDA",	0x124, (),		() },
-	{ "i2c2_scl",	"CAM_I2C_SCL",		0x126, (IEN | M0),	(OFF_DIS) },
-	{ "i2c2_sda",	"CAM_I2C_SDA",		0x128, (IEN | M0),	(OFF_DIS) },
+	{ "i2c2_scl",	"CAM_I2C_SCL",		0x126, (IEN | PTU | EN | M0),	(OFF_DIS) },
+	{ "i2c2_sda",	"CAM_I2C_SDA",		0x128, (IEN | PTU | EN | M0),	(OFF_DIS) },
 	/* NOTE: Fuel gauge expects I2C SCL and SDA lines to be low to enter sleep mode. This save ~70uA */
 	{ "i2c3_scl",	"AP_I2C_SCL",		0x12A, (IEN | PTU | EN | M0),	(OFF_IN_PD) },
 	{ "i2c3_sda",	"AP_I2C_SDA",		0x12C, (IEN | PTU | EN | M0),	(OFF_IN_PD) },
