Index: Kernel_933422/arch/arm/mach-omap2/omap4-iopad.c
===================================================================
--- Kernel_933422.orig/arch/arm/mach-omap2/omap4-iopad.c	2011-02-28 21:09:04.000000000 +0530
+++ Kernel_933422/arch/arm/mach-omap2/omap4-iopad.c	2011-03-01 18:25:17.862456416 +0530
@@ -119,8 +119,8 @@
 
 	{ "hdmi_hpd",	"AP_HDMI_HPD",		0x098, (IEN | PTD | EN | M0),	(OFF_DIS) },
 	{ "hdmi_cec",	"AP_HDMI_CEC",		0x09A, (IEN | PTU | EN | M0),	(OFF_DIS) },
-	{ "hdmi_ddc_scl", "AP_HDMI_I2C_SCL",	0x09C, (IEN | M0),		(OFF_IN_PU) },
-	{ "hdmi_ddc_sda", "AP_HDMI_I2C_SDA",	0x09E, (IEN | M0),		(OFF_IN_PU) },
+	{ "hdmi_ddc_scl", "AP_HDMI_I2C_SCL",	0x09C, (IEN | PTU | EN | M0),	(OFF_DIS) },
+	{ "hdmi_ddc_sda", "AP_HDMI_I2C_SDA",	0x09E, (IEN | PTU | EN | M0),	(OFF_DIS) },
 
 	//{ "csi21_dx0",	"8M_CLK_P",		0x0A0, (),		() },
 	//{ "csi21_dy0",	"8M_CLK_N",		0x0A2, (),		() },
@@ -318,10 +318,12 @@
 		omap_writew((pmux->off_mode | pmux->func_mode),
 				(OMAP4_CTRL_MODULE_PAD_CORE_MUX_PBASE + pmux->offset));
 
+		/*
 		printk("%s, %s, 0x%08X, 0x%04X\n",
 				pmux->omap_pad_name, pmux->signal_name,
 				(pmux->offset + OMAP4_CTRL_MODULE_PAD_CORE_MUX_PBASE),
 				(pmux->func_mode | pmux->off_mode));
+		*/
 	}
 
 	/* FREF_CLK_REQ */
Index: Kernel_933422/drivers/video/omap2/dss/hdmi.c
===================================================================
--- Kernel_933422.orig/drivers/video/omap2/dss/hdmi.c	2011-02-25 20:33:23.000000000 +0530
+++ Kernel_933422/drivers/video/omap2/dss/hdmi.c	2011-03-01 18:17:06.682456397 +0530
@@ -1569,13 +1569,6 @@
 		goto err;
 	}
 
-	/* PAD0_HDMI_HPD_PAD1_HDMI_CEC */
-	omap_writel(0x01100110, 0x4A100098);
-	/* PAD0_HDMI_DDC_SCL_PAD1_HDMI_DDC_SDA */
-	omap_writel(0x01100110 , 0x4A10009C);
-	/* CONTROL_HDMI_TX_PHY */
-	omap_writel(0x10000000, 0x4A100610);
-
 	if (dssdev->platform_enable)
 		dssdev->platform_enable(dssdev);
 
