Index: Kernel/drivers/mmc/host/omap_hsmmc.c
===================================================================
--- Kernel.orig/drivers/mmc/host/omap_hsmmc.c	2011-03-04 20:11:35.000000000 +0530
+++ Kernel/drivers/mmc/host/omap_hsmmc.c	2011-03-04 20:16:57.472444122 +0530
@@ -2702,6 +2702,47 @@
 	twl_i2c_write_u8(TWL_MODULE_PM_MASTER, 0x01, 0x90-0x001F); //VCXIO_CFG_GRP
 	twl_i2c_write_u8(TWL_MODULE_PM_MASTER, 0x01, 0x91-0x001F); //VCXIO_CFG_TRANS
 
+	twl_i2c_write_u8(TWL_MODULE_USB, 0x21, 0xEA); // CHARGERUSB_CTRL3
+	twl_i2c_write_u8(TWL_MODULE_USB, 0xFF, 0x05); // USB_VBUS_CTRL_CLR
+	twl_i2c_write_u8(TWL_MODULE_USB, 0xFF, 0x07); // USB_ID_CTRL_CLR
+	twl_i2c_write_u8(TWL_MODULE_USB, 0x00, 0x2E); // GPADC_CTRL
+	twl_i2c_write_u8(TWL_MODULE_USB, 0x51, 0x90); // TOGGLE1
+
+	twl_i2c_write_u8(TWL_MODULE_PM_MASTER, 0x20, 0x9A - 0x001F); // VMMC_CFG_STATE
+	twl_i2c_write_u8(TWL_MODULE_PM_MASTER, 0x20, 0xA6 - 0x001F); // VUSIM_CFG_STATE
+	twl_i2c_write_u8(TWL_MODULE_PM_MASTER, 0x00, 0x9E - 0x001F); // VPP_CFG_STATE
+	//twl_i2c_write_u8(TWL_MODULE_PM_MASTER, 0x7F, 0xF4 - 0x001F); // CFG_LDO_PD1
+	twl_i2c_write_u8(TWL_MODULE_PM_MASTER, 0x00, 0xE4 - 0x001F); // MISC1
+	twl_i2c_write_u8(TWL_MODULE_PM_MASTER, 0x00, 0xE5 - 0x001F); // MISC2
+	twl_i2c_write_u8(TWL_MODULE_PM_MASTER, 0x62, 0xE6 - 0x001F); // BBSPOR_CFG
+	twl_i2c_write_u8(TWL_MODULE_PM_MASTER, 0x65, 0xF1 - 0x001F); // CFG_INPUT_PUPD2
+	twl_i2c_write_u8(TWL_MODULE_PM_MASTER, 0x00, 0xF3 - 0x001F); // CFG_INPUT_PUPD4
+	twl_i2c_write_u8(TWL_MODULE_PM_MASTER, 0x00, 0xF5 - 0x001F); // CFG_LDO_PD2
+
+	twl_i2c_write_u8(TWL_MODULE_PM_MASTER, 0x01, 0x80 - 0x001F); // VANA_CFG_GRP
+	twl_i2c_write_u8(TWL_MODULE_PM_MASTER, 0x01, 0x81 - 0x001F); // VANA_CFG_TRANS
+	twl_i2c_write_u8(TWL_MODULE_PM_MASTER, 0x20, 0x82 - 0x001F); // VANA_CFG_STATE
+
+	twl_i2c_write_u8(TWL_MODULE_PM_MASTER, 0x01, 0xCE - 0x001F); // TMP_CFG_GRP
+	twl_i2c_write_u8(TWL_MODULE_PM_MASTER, 0x01, 0xCF - 0x001F); // TMP_CFG_TRANS
+	twl_i2c_write_u8(TWL_MODULE_PM_MASTER, 0x20, 0xD0 - 0x001F); // TMP_CFG_STATE
+
+	//twl_i2c_write_u8(TWL_MODULE_PM_MASTER, 0x01, 0x8C - 0x001F); //VAUX3_CFG_GRP
+	//twl_i2c_write_u8(TWL_MODULE_PM_MASTER, 0x01, 0x8D - 0x001F); //VAUX3_CFG_TRANS
+	//twl_i2c_write_u8(TWL_MODULE_PM_MASTER, 0x20, 0x8E - 0x001F); //VAUX3_CFG_STATE
+
+	twl_i2c_write_u8(TWL_MODULE_PM_MASTER, 0x01, 0xC2 - 0x001F); // VRTC_CFG_GRP
+	twl_i2c_write_u8(TWL_MODULE_PM_MASTER, 0x01, 0xC3 - 0x001F); // VRTC_CFG_TRANS
+	twl_i2c_write_u8(TWL_MODULE_PM_MASTER, 0x20, 0xC4 - 0x001F); // VRTC_CFG_STATE
+	twl_i2c_write_u8(TWL_MODULE_PM_MASTER, 0x40, 0xC4 - 0x001F); // VRTC_CFG_STATE
+	twl_i2c_write_u8(TWL_MODULE_PM_MASTER, 0x80, 0xC4 - 0x001F); // VRTC_CFG_STATE
+
+	twl_i2c_write_u8(TWL_MODULE_PM_MASTER, 0x01, 0x88-0x001F); // VAUX2_CFG_GRP
+	twl_i2c_write_u8(TWL_MODULE_PM_MASTER, 0x01, 0x89-0x001F); // VAUX2_CFG_TRANS
+	twl_i2c_write_u8(TWL_MODULE_PM_MASTER, 0x20, 0x8A-0x001F); // VAUX2_CFG_STATE
+
+	twl_i2c_write_u8(TWL_MODULE_PM_MASTER, 0x06, 0x25 - 0x001F); // PHEONIX_DEV_ON
+
 	return ret;
 }
 
