Index: Kernel/arch/arm/mach-omap2/board-omap4tab10-1.c
===================================================================
--- Kernel.orig/arch/arm/mach-omap2/board-omap4tab10-1.c	2011-03-01 20:49:30.000000000 +0900
+++ Kernel/arch/arm/mach-omap2/board-omap4tab10-1.c	2011-03-02 16:54:30.829127639 +0900
@@ -90,7 +90,7 @@
 #endif // CONFIG_SAMSUNG_PHONE_SVNET
 
 #define _OMAP_MUX_SETTING
-#if (CONFIG_SAMSUNG_OMAP4_TAB_REV == 5)
+#if (CONFIG_SAMSUNG_OMAP4_TAB_REV >= 5) && !defined(CONFIG_VIDEO_M5MO)
 #define DISABLE_I2C2_HW_SPINLOCK_FOR_REV5
 #else
 #undef DISABLE_I2C2_HW_SPINLOCK_FOR_REV5
Index: Kernel/arch/arm/mach-omap2/omap4-iopad.c
===================================================================
--- Kernel.orig/arch/arm/mach-omap2/omap4-iopad.c	2011-03-01 20:49:30.000000000 +0900
+++ Kernel/arch/arm/mach-omap2/omap4-iopad.c	2011-03-02 18:07:32.811628026 +0900
@@ -89,8 +89,8 @@
 	//{ "gpmc_ad14",	"OLED_DET",	0x05C, (IEN | M3),	(OFF_IN | OFF_EN) },
 	{ "gpmc_ad15",	"TA_CURRENT_SEL",	0x05E, (IEN | M3),	(OFF_IN | OFF_EN) },
 	{ "gpmc_a16",	"LCD_EN",		0x060, (M3),		(OFF_IN_PD) },
-	{ "gpmc_a17",	"XN_RST_OUT",		0x062, (M3),		(OFF_IN_PU) },
-	{ "gpmc_a18",	"MOTOR_EN",		0x064, (M3),		(OFF_IN_PD) },
+	{ "gpmc_a17",	"XN_RST_OUT",		0x062, (IEN | EN | M3),		(OFF_IN_PU) },
+	{ "gpmc_a18",	"MOTOR_EN",		0x064, (M3),	 	(OFF_IN_PD) },
 	{ "gpmc_a19",	"CAM_PMIC_EN",		0x066, (M3),		(OFF_IN_PD) },
 	//{ "gpmc_a20",	"EARPATH_SEL",		0x068, (),	() },
 	{ "gpmc_a21",	"GYRO_INT",		0x06A, (IEN | M3),	(OFF_IN | OFF_EN) },
@@ -119,8 +119,8 @@
