Index: Kernel/drivers/power/twl6030_bci_battery.c
===================================================================
--- Kernel.orig/drivers/power/twl6030_bci_battery.c	2011-03-06 23:13:02.000000000 +0900
+++ Kernel/drivers/power/twl6030_bci_battery.c	2011-03-06 23:21:53.419962001 +0900
@@ -259,9 +259,11 @@
 };
 static struct blocking_notifier_head notifier_list;
 extern u32 wakeup_timer_seconds;
+#ifdef CONFIG_MACH_OMAP4_TAB_10_1
 #if ((CONFIG_SAMSUNG_OMAP4_TAB_REV == 5) || (CONFIG_SAMSUNG_OMAP4_TAB_REV == 6))
 	extern int read_max17040_vcell(void);
 #endif
+#endif
 
 static void twl6030_config_min_vbus_reg(struct twl6030_bci_device_info *di,
 						unsigned int value)
@@ -882,8 +884,10 @@
 		twl6030_stop_charger(di);
 		/*voltage setteling time*/
 		msleep(200);
-        #if ((CONFIG_SAMSUNG_OMAP4_TAB_REV == 5) || (CONFIG_SAMSUNG_OMAP4_TAB_REV == 6))
+		#ifdef CONFIG_MACH_OMAP4_TAB_10_1
+        	#if ((CONFIG_SAMSUNG_OMAP4_TAB_REV == 5) || (CONFIG_SAMSUNG_OMAP4_TAB_REV == 6))
 			di->voltage_uV = read_max17040_vcell();
+		#endif
 		#else
 			di->voltage_uV = twl6030_get_gpadc_conversion(7);
 		#endif
@@ -952,9 +956,10 @@
 		dev_dbg(di->dev, "gpadc conversion failed: %d\n", ret);
 		return;
 	}
-
-    #if ((CONFIG_SAMSUNG_OMAP4_TAB_REV == 5) || (CONFIG_SAMSUNG_OMAP4_TAB_REV == 6))
+	#ifdef CONFIG_MACH_OMAP4_TAB_10_1
+    	#if ((CONFIG_SAMSUNG_OMAP4_TAB_REV == 5) || (CONFIG_SAMSUNG_OMAP4_TAB_REV == 6))
 		di->voltage_uV = read_max17040_vcell();
+	#endif
 	#else
  		if (req.rbuf[7] > 0)
  			di->voltage_uV = req.rbuf[7];
@@ -1097,9 +1102,10 @@
 		val->intval = di->charge_status;
 		break;
 	case POWER_SUPPLY_PROP_VOLTAGE_NOW:
-
-        #if ((CONFIG_SAMSUNG_OMAP4_TAB_REV == 5) || (CONFIG_SAMSUNG_OMAP4_TAB_REV == 6)) 
+		#ifdef CONFIG_MACH_OMAP4_TAB_10_1
+        	#if ((CONFIG_SAMSUNG_OMAP4_TAB_REV == 5) || (CONFIG_SAMSUNG_OMAP4_TAB_REV == 6)) 
 			di->voltage_uV = read_max17040_vcell();
+		#endif
 		#else
 			di->voltage_uV = twl6030_get_gpadc_conversion(7);
 		#endif
@@ -1790,9 +1796,10 @@
 	di->charger_incurrentmA = di->max_charger_currentmA;
 	di->charger_outcurrentmA = di->max_charger_currentmA;
 	di->watchdog_duration = 32;
-
-    #if ((CONFIG_SAMSUNG_OMAP4_TAB_REV == 5) || (CONFIG_SAMSUNG_OMAP4_TAB_REV == 6))
+	#ifdef CONFIG_MACH_OMAP4_TAB_10_1
+    	#if ((CONFIG_SAMSUNG_OMAP4_TAB_REV == 5) || (CONFIG_SAMSUNG_OMAP4_TAB_REV == 6))
 		di->voltage_uV = read_max17040_vcell();
+	#endif
 	#else
 		di->voltage_uV = twl6030_get_gpadc_conversion(7);
 	#endif
Index: Kernel/arch/arm/mach-omap2/board-4430univ-sensors.c
===================================================================
--- Kernel.orig/arch/arm/mach-omap2/board-4430univ-sensors.c	2011-03-05 19:17:00.000000000 +0900
+++ Kernel/arch/arm/mach-omap2/board-4430univ-sensors.c	2011-03-06 23:23:29.869962002 +0900
@@ -68,10 +68,12 @@
 
 void __init omap4430univ_sensors_init(void)
 {
+#ifdef CONFIG_MACH_OMAP4_TAB_10_1
 #if (CONFIG_SAMSUNG_OMAP4_TAB_REV <=2)
 	omap4430univ_akm8973_init();
 	omap4430univ_isl29023_init();
 #endif
+#endif
 	omap4430univ_kxsd9_init();
 }
 
