Index: Kernel/arch/arm/mach-omap2/board-omap4tab10-1.c
===================================================================
--- Kernel.orig/arch/arm/mach-omap2/board-omap4tab10-1.c	2011-03-09 10:59:37.000000000 +0900
+++ Kernel/arch/arm/mach-omap2/board-omap4tab10-1.c	2011-03-09 14:08:26.707878115 +0900
@@ -234,6 +234,7 @@
 static void omap4_charger_cb(u8 attached);
 static void omap4_jig_cb(u8 attached);
 static void omap4_fsa9480_reset_cb(void);
+static void omap4_mhl_cb(u8 attached);
 
 static struct wake_lock uart_lock;
 #define TWL6030_RTC_GPIO 6
@@ -372,6 +373,7 @@
        .charger_cb     = omap4_charger_cb,
        .jig_cb         = omap4_jig_cb,
        .reset_cb       = omap4_fsa9480_reset_cb,
+	   .mhl_cb         = omap4_mhl_cb,
 };
 
 
@@ -529,6 +531,30 @@
        printk("\nBoard file [FSA9480]: Reset Callback \n");
 }
 
+extern void MHL_On(bool on);
+
+static bool fsa9480_is_mhl_attached(void)
+{
+        int val;
+
+        gpio_request(OMAP4430_GPIO_MHL_SEL, "MHL_SEL");
+        val = gpio_get_value(OMAP4430_GPIO_MHL_SEL);
+        gpio_free(OMAP4430_GPIO_MHL_SEL);
+
+        return !!val;
+}
+
+static void omap4_mhl_cb(u8 attached)
+{
+    /*TODO */
+       printk("\nBoard file [FSA9480]: MHL Callback\n");
+       gpio_request(OMAP4430_GPIO_MHL_SEL, "MHL_SEL");
+       gpio_direction_output(OMAP4430_GPIO_MHL_SEL,1);
+       gpio_free(OMAP4430_GPIO_MHL_SEL);
+       MHL_On(attached);
+}
+
+
 
 static struct matrix_keymap_data omap4tab_keymap_data = {
 	.keymap			= omap4tab_keymap,
Index: Kernel/arch/arm/mach-omap2/omap4-iopad.c
===================================================================
--- Kernel.orig/arch/arm/mach-omap2/omap4-iopad.c	2011-03-08 13:53:20.000000000 +0900
+++ Kernel/arch/arm/mach-omap2/omap4-iopad.c	2011-03-09 14:05:06.677878035 +0900
@@ -304,9 +304,12 @@
 	{ "dpm_emu18",	"LCD_D(1)",	0x1D2, (PTD | EN | M5),		(OFF_OUT_PD) },
 	{ "dpm_emu19",	"LCD_D(0)",	0x1D4, (PTD | EN | M5),		(OFF_OUT_PD) },
 	/* NOTE: configuring the pins for I2C-using GPIOs Operations required for MHL*/
+	 #if ( (CONFIG_SAMSUNG_OMAP4_TAB_REV == 6) \
+     || (CONFIG_SAMSUNG_OMAP4_TAB_REV == 7))
+
   	{ "usbc1_icusb_dm",     "MHL_SCL_1.8V", 0x0E0,  (PTD | EN | M3),        (OFF_OUT_PD) },
     { "usbc1_icusb_dp",     "MHL_SDA_1.8V", 0x0E4,  (PTD | EN | M3),        (OFF_OUT_PD) },
-
+	#endif
 };
 
 void set_offmode_padconfig(void)
Index: Kernel/drivers/usb/otg/fsa9480.c
===================================================================
--- Kernel.orig/drivers/usb/otg/fsa9480.c	2011-03-07 19:09:55.000000000 +0900
+++ Kernel/drivers/usb/otg/fsa9480.c	2011-03-09 14:04:33.637878113 +0900
@@ -101,6 +101,9 @@
 #define FSA9480_DEV_T2_JIG_MASK                (DEV_JIG_USB_OFF | DEV_JIG_USB_ON | \
                                        DEV_JIG_UART_OFF | DEV_JIG_UART_ON)
 
+#define DEV_MHL                 (DEV_AV)
+#define FSA9480_DEV_T2_MHL_MASK         (DEV_MHL)
+
 struct fsa9480_usbsw {
        struct i2c_client               *client;
        struct fsa9480_platform_data    *pdata;
@@ -339,6 +342,11 @@
                        if (pdata->jig_cb)
                                pdata->jig_cb(FSA9480_ATTACHED);
                }
+			  if(val2 & FSA9480_DEV_T2_MHL_MASK) {
+                    if(pdata->mhl_cb) {
+                        pdata->mhl_cb(FSA9480_ATTACHED);
+                    }
+                }
        } else if (intr & (1 << 1)) {
                if (usbsw->dev1 & FSA9480_DEV_T1_USB_MASK ||
                        usbsw->dev2 & FSA9480_DEV_T2_USB_MASK) {
@@ -362,6 +370,11 @@
                        if (pdata->jig_cb)
                                pdata->jig_cb(FSA9480_DETACHED);
                }
+			  if (usbsw->dev2 & FSA9480_DEV_T2_MHL_MASK) {
+
+                       if (pdata->mhl_cb)
+                               pdata->mhl_cb(FSA9480_DETACHED);
+              }
        }
 
        usbsw->dev1 = val1;
Index: Kernel/include/linux/fsa9480.h
===================================================================
--- Kernel.orig/include/linux/fsa9480.h	2011-03-03 11:18:35.000000000 +0900
+++ Kernel/include/linux/fsa9480.h	2011-03-09 14:00:06.907877605 +0900
@@ -32,6 +32,7 @@
        void (*charger_cb) (u8 attached);
        void (*jig_cb) (u8 attached);
        void (*reset_cb) (void);
+	   void (*mhl_cb)(u8 attached);
 #define FSA9480_ATTACHED (1)
 #define FSA9480_DETACHED (0)
 };
