Index: Kernel/arch/arm/mach-omap2/board-omap4tab10-1.c
===================================================================
--- Kernel.orig/arch/arm/mach-omap2/board-omap4tab10-1.c	2011-03-09 22:59:21.000000000 +0900
+++ Kernel/arch/arm/mach-omap2/board-omap4tab10-1.c	2011-03-09 23:12:55.807878053 +0900
@@ -205,18 +205,14 @@
 #endif
 
 /*MHL Related GPIOs */
-
-#if ((CONFIG_SAMSUNG_OMAP4_TAB_REV == 6) \
-         || (CONFIG_SAMSUNG_OMAP4_TAB_REV == 7))
+#ifdef CONFIG_VIDEO_MHL_V1
 #define OMAP4430_GPIO_MHL_RST           60
 #define OMAP4430_GPIO_MHL_SEL          170
 #define OMAP4430_GPIO_MHL_INT          175
 #define OMAP4430_GPIO_MHL_WAKEUP    64
 #define OMAP4430_GPIO_MHL_SCL_18V   99
 #define OMAP4430_GPIO_MHL_SDA_18V   98
-
 #endif
-
 /*For FM*/
 #if ((CONFIG_SAMSUNG_OMAP4_TAB_REV == 6) \
          || (CONFIG_SAMSUNG_OMAP4_TAB_REV == 7))
@@ -254,7 +250,10 @@
 static void omap4_charger_cb(u8 attached);
 static void omap4_jig_cb(u8 attached);
 static void omap4_fsa9480_reset_cb(void);
+
+#ifdef CONFIG_VIDEO_MHL_V1
 static void omap4_mhl_cb(u8 attached);
+#endif
 
 static struct wake_lock uart_lock;
 #define TWL6030_RTC_GPIO 6
@@ -393,7 +392,9 @@
        .charger_cb     = omap4_charger_cb,
        .jig_cb         = omap4_jig_cb,
        .reset_cb       = omap4_fsa9480_reset_cb,
+	#ifdef CONFIG_VIDEO_MHL_V1
 	   .mhl_cb         = omap4_mhl_cb,
+	#endif
 };
 
 
@@ -550,7 +551,7 @@
        /* TODO: need to implement this once uUSB charging available in HW */
        printk("\nBoard file [FSA9480]: Reset Callback \n");
 }
-
+#ifdef CONFIG_VIDEO_MHL_V1
 extern void MHL_On(bool on);
 
 static bool fsa9480_is_mhl_attached(void)
@@ -575,6 +576,7 @@
 }
 
 
+#endif
 
 static struct matrix_keymap_data omap4tab_keymap_data = {
 	.keymap			= omap4tab_keymap,
