Index: Kernel/arch/arm/mach-omap2/opp44xx_data.c
===================================================================
--- Kernel.orig/arch/arm/mach-omap2/opp44xx_data.c	2011-03-02 10:32:05.542496811 +0530
+++ Kernel/arch/arm/mach-omap2/opp44xx_data.c	2011-03-09 22:28:50.822664927 +0530
@@ -93,10 +93,10 @@
 #if 0
 	OMAP_OPP_DEF("hsi", true, 96000000, 930000),
 #else
-	OMAP_OPP_DEF("hsi", true, 192000000, 1100000),
+	OMAP_OPP_DEF("hsi", true, 96000000, 930000),
 #endif
 	/* HSI OPP2 - OPP100 */
-	OMAP_OPP_DEF("hsi", true, 192000000, 1100000),
+	OMAP_OPP_DEF("hsi", true, 96000000, 1100000),
 };
 
 static struct omap_opp_def __initdata omap44xx_opp_def_list[] = {
@@ -157,10 +157,10 @@
 #if 0
 	OMAP_OPP_DEF("hsi", true, 96000000, 930000),
 #else
-	OMAP_OPP_DEF("hsi", true, 192000000, 1100000),
+	OMAP_OPP_DEF("hsi", true, 96000000, 930000),
 #endif
 	/* HSI OPP2 - OPP100 */
-	OMAP_OPP_DEF("hsi", true, 192000000, 1100000),
+	OMAP_OPP_DEF("hsi", true, 96000000, 1100000),
 };
 
 #define	L3_OPP50_RATE			100000000
Index: Kernel/arch/arm/mach-omap2/omap4-iopad.c
===================================================================
--- Kernel.orig/arch/arm/mach-omap2/omap4-iopad.c	2011-03-09 22:06:19.772039825 +0530
+++ Kernel/arch/arm/mach-omap2/omap4-iopad.c	2011-03-09 22:40:08.422040307 +0530
@@ -18,6 +18,7 @@
 #include <plat/io.h>
 #include "mux44xx.h"
 
+#define WKUP_EN		(1 << 14)
 #define OFF_PD		(1 << 12)
 #define OFF_PU		(3 << 12)
 #define OFF_OUT_PTD	(0 << 11)
@@ -345,6 +346,7 @@
 	/* SYSCTRL_PADCONF_WKUP Registers */
 	/* sim_io */ omap_writew((PTD | EN | M7), 0x4A31E040);
 	/* sim_clk */ omap_writew((M3), 0x4A31E042);
+	/* sim_cd */ omap_writew((WKUP_EN | IEN | PTU | EN | M3), 0x4A31E046);
 	/* sim_pwrctrl */ omap_writew((PTD | EN | M7), 0x4A31E048);
 	/* fref_clk3_req */ omap_writew((PTD | EN | M3), 0x4A31E056);
 	/* fref_clk4_req */ omap_writew((PTD | EN | M3), 0x4A31E05A);
