Index: Kernel/arch/arm/mach-omap2/board-omap4tab10-1.c
===================================================================
--- Kernel.orig/arch/arm/mach-omap2/board-omap4tab10-1.c	2011-03-10 22:09:33.000000000 +0530
+++ Kernel/arch/arm/mach-omap2/board-omap4tab10-1.c	2011-03-10 22:11:26.632040304 +0530
@@ -2699,7 +2699,9 @@
 		omap_mux_set_gpio(gpio_val, OMAP4TAB_GP2A_PS_VOUT);
 	}
 
-#if (CONFIG_SAMSUNG_OMAP4_TAB_REV == 5)
+#if ((CONFIG_SAMSUNG_OMAP4_TAB_REV == 5) \
+		|| (CONFIG_SAMSUNG_OMAP4_TAB_REV == 6) \
+		|| (CONFIG_SAMSUNG_OMAP4_TAB_REV == 7) )
 	/*
 	 * Enable IO daisy for sys_nirq1/2 & HSI CAWAKE line, to be able to
 	 * wakeup from interrupts from PMIC/Audio IC.
@@ -2780,7 +2782,9 @@
 	.vdd0_on = 1350000,        /* 1.35v */
 	.vdd0_onlp = 1350000,      /* 1.35v */
 	.vdd0_ret = 837500,       /* 0.8375v */
-#if (CONFIG_SAMSUNG_OMAP4_TAB_REV == 5)
+#if ((CONFIG_SAMSUNG_OMAP4_TAB_REV == 5) \
+		|| (CONFIG_SAMSUNG_OMAP4_TAB_REV == 6) \
+		|| (CONFIG_SAMSUNG_OMAP4_TAB_REV == 7) )
 	.vdd0_off = 0,          /* 0 v */
 #else
 	.vdd0_off = 600000,       /* 0.6v */
@@ -2788,7 +2792,9 @@
 	.vdd1_on = 1100000,        /* 1.1v */
 	.vdd1_onlp = 1100000,      /* 1.1v */
 	.vdd1_ret = 837500,       /* 0.8375v */
-#if (CONFIG_SAMSUNG_OMAP4_TAB_REV == 5)
+#if ((CONFIG_SAMSUNG_OMAP4_TAB_REV == 5) \
+		|| (CONFIG_SAMSUNG_OMAP4_TAB_REV == 6) \
+		|| (CONFIG_SAMSUNG_OMAP4_TAB_REV == 7) )
 	.vdd1_off = 0,		/* 0 v */
 #else
 	.vdd1_off = 600000,       /* 0.6v */
@@ -2796,7 +2802,9 @@
 	.vdd2_on = 1100000,        /* 1.1v */
 	.vdd2_onlp = 1100000,      /* 1.1v */
 	.vdd2_ret = 837500,       /* .8375v */
-#if (CONFIG_SAMSUNG_OMAP4_TAB_REV == 5)
+#if ((CONFIG_SAMSUNG_OMAP4_TAB_REV == 5) \
+		|| (CONFIG_SAMSUNG_OMAP4_TAB_REV == 6) \
+		|| (CONFIG_SAMSUNG_OMAP4_TAB_REV == 7) )
 	.vdd2_off = 0,		/* 0 v */
 #else
 	.vdd2_off = 600000,       /* 0.6v */
