Index: Kernel/drivers/input/touchscreen/atmel_touchscreen.c
===================================================================
--- Kernel.orig/drivers/input/touchscreen/atmel_touchscreen.c	2011-03-11 18:01:19.912471122 +0530
+++ Kernel/drivers/input/touchscreen/atmel_touchscreen.c	2011-03-11 18:01:51.332471706 +0530
@@ -1218,7 +1218,6 @@
 static int touchscreen_suspend(struct platform_device *pdev, pm_message_t state) {
 	dbg("touchscreen_suspend : touch power off\n");
 	disable_irq_nosync(tsp.irq);
-	disable_irq_wake(tsp.irq);
 	atmel_suspend();
 	initialize_multi_touch();
 	return 0;
@@ -1239,7 +1238,6 @@
 #endif
 	atmel_resume();
 	initialize_multi_touch();
-	enable_irq_wake(tsp.irq);
 	enable_irq(tsp.irq);
 	dbg("touchscreen_resume end\n");
 	return 0;
