Index: kernel/arch/arm/mach-omap2/board-t1.c
===================================================================
--- kernel.orig/arch/arm/mach-omap2/board-t1.c	2011-03-13 12:17:59.488452080 +0900
+++ kernel/arch/arm/mach-omap2/board-t1.c	2011-03-13 15:17:13.567224866 +0900
@@ -1599,7 +1599,7 @@
 err0:
 	return;
 }
-
+#if 0
 static void t1_omap4_touch_init(void)
 {
 	printk("[SHANKAR] %s[%d] Entry \n", __func__, __LINE__);
@@ -1610,6 +1610,16 @@
 	printk("[SHANKAR] %s[%d] exit \n", __func__, __LINE__);
 	t1_omap_touch_init_hw(&t1_omap_mxt224E_platform_data);
 }
+#endif
+static void t1_omap4_touch_init(void)
+{
+        printk("[SHANKAR] %s[%d] Entry \n", __func__, __LINE__);
+        gpio_request(OMAP_GPIO_TSP_EN, "TOUCH ENable GPIO");
+        gpio_direction_output(OMAP_GPIO_TSP_EN, 1);
+        msleep(250);
+        printk("[SHANKAR] %s[%d] exit \n", __func__, __LINE__);
+
+}
 
 static void t1_omap4_vib_motor_init(void)
 {
Index: kernel/drivers/input/touchscreen/atmel_touch.h
===================================================================
--- kernel.orig/drivers/input/touchscreen/atmel_touch.h	2011-03-13 12:18:14.538435983 +0900
+++ kernel/drivers/input/touchscreen/atmel_touch.h	2011-03-13 15:17:13.587223983 +0900
@@ -1,3 +1,4 @@
+#ifdef CONFIG_MACH_OMAP4_TAB_10_1
 #if (CONFIG_SAMSUNG_OMAP4_TAB_REV == 4)
 #define OMAP_GPIO_TSP_EN		101
 #elif ((CONFIG_SAMSUNG_OMAP4_TAB_REV == 5 ) \
@@ -5,6 +6,11 @@
 	 || (CONFIG_SAMSUNG_OMAP4_TAB_REV == 7))
 #define OMAP_GPIO_TSP_EN                54
 #endif
+#endif
+
+#if defined(CONFIG_MACH_OMAP4_SAMSUNG)
+#define OMAP_GPIO_TSP_EN                54
+#endif
 typedef enum
 {
 	DEFAULT_MODEL,
Index: kernel/drivers/input/touchscreen/atmel_touchscreen.c
===================================================================
--- kernel.orig/drivers/input/touchscreen/atmel_touchscreen.c	2011-03-13 12:18:14.548436563 +0900
+++ kernel/drivers/input/touchscreen/atmel_touchscreen.c	2011-03-13 15:17:13.587223983 +0900
@@ -59,6 +59,7 @@
 #define TOUCH_KEYSET_T31			31u
 #define TOUCH_XSLIDERSET_T32			32u
 
+#ifdef CONFIG_MACH_OMAP4_TAB_10_1
 #if (CONFIG_SAMSUNG_OMAP4_TAB_REV == 4)
 #define  OMAP_GPIO_TSP_INT		43
 #elif ((CONFIG_SAMSUNG_OMAP4_TAB_REV == 5) \
@@ -66,7 +67,13 @@
 	 || (CONFIG_SAMSUNG_OMAP4_TAB_REV == 7))
 #define  OMAP_GPIO_TSP_INT              46
 #endif
+#endif
+
+#if defined(CONFIG_MACH_OMAP4_SAMSUNG)
+#undef OMAP_GPIO_TSP_INT
 
+#define  OMAP_GPIO_TSP_INT              46
+#endif
 #include <linux/errno.h>
 #include <linux/kernel.h>
 #include <linux/module.h>
Index: kernel/arch/arm/configs/android_t1_omap4430_r00_eng_defconfig
===================================================================
--- kernel.orig/arch/arm/configs/android_t1_omap4430_r00_eng_defconfig	2011-03-13 16:01:37.000000000 +0900
+++ kernel/arch/arm/configs/android_t1_omap4430_r00_eng_defconfig	2011-03-13 16:04:30.388148861 +0900
@@ -1,7 +1,7 @@
 #
 # Automatically generated make config: don't edit
 # Linux kernel version: 2.6.35.7
-# Sun Mar 13 11:20:17 2011
+# Sun Mar 13 16:04:03 2011
 #
 CONFIG_ARM=y
 CONFIG_SYS_SUPPORTS_APM_EMULATION=y
@@ -998,9 +998,9 @@
 # CONFIG_TOUCHSCREEN_TSC2007 is not set
 # CONFIG_TOUCHSCREEN_W90X900 is not set
 # CONFIG_TOUCHSCREEN_SYNTM12XX is not set
-# CONFIG_TOUCHSCREEN_AT42QT602240 is not set
+CONFIG_TOUCHSCREEN_AT42QT602240=y
 # CONFIG_TOUCHSCREEN_TPS6507X is not set
-CONFIG_TOUCHSCREEN_ATMEL_MXT224E=y
+# CONFIG_TOUCHSCREEN_ATMEL_MXT224E is not set
 CONFIG_INPUT_MISC=y
 # CONFIG_INPUT_AD714X is not set
 # CONFIG_INPUT_BH1780 is not set
@@ -1177,10 +1177,9 @@
 # CONFIG_TEST_POWER is not set
 # CONFIG_BATTERY_DS2760 is not set
 # CONFIG_BATTERY_DS2782 is not set
-CONFIG_TWL6030_BCI_BATTERY=y
-# CONFIG_CHARGER_BQ2415x is not set
+# CONFIG_TWL6030_BCI_BATTERY is not set
 # CONFIG_BATTERY_BQ27x00 is not set
-CONFIG_BATTERY_MAX17040=y
+# CONFIG_BATTERY_MAX17040 is not set
 CONFIG_HWMON=y
 # CONFIG_HWMON_VID is not set
 # CONFIG_HWMON_DEBUG_CHIP is not set
