Index: kernel/arch/arm/mach-omap2/board-t1.c
===================================================================
--- kernel.orig/arch/arm/mach-omap2/board-t1.c	2011-03-15 10:57:07.268435372 +0900
+++ kernel/arch/arm/mach-omap2/board-t1.c	2011-03-15 11:36:30.697227741 +0900
@@ -1655,7 +1655,7 @@
  * Same devices installed on EMIF1 and EMIF2
  */
 static __initdata struct emif_device_details emif_devices = {
-	.cs0_device = &elpida_2G_S4,
+	.cs0_device = &elpida_4G_S4,
 	.cs1_device = NULL
 };
 
Index: kernel/arch/arm/mach-omap2/include/mach/lpddr2-elpida.h
===================================================================
--- kernel.orig/arch/arm/mach-omap2/include/mach/lpddr2-elpida.h	2011-03-13 12:17:59.598436321 +0900
+++ kernel/arch/arm/mach-omap2/include/mach/lpddr2-elpida.h	2011-03-15 11:36:30.717227733 +0900
@@ -19,5 +19,6 @@
 extern const struct lpddr2_timings timings_elpida_400_mhz;
 extern const struct lpddr2_min_tck min_tck_elpida;
 extern struct lpddr2_device_info elpida_2G_S4;
+extern struct lpddr2_device_info elpida_4G_S4;
 
 #endif
Index: kernel/arch/arm/mach-omap2/lpddr2_elpida_data.c
===================================================================
--- kernel.orig/arch/arm/mach-omap2/lpddr2_elpida_data.c	2011-03-13 12:17:59.598436321 +0900
+++ kernel/arch/arm/mach-omap2/lpddr2_elpida_data.c	2011-03-15 11:36:30.717227733 +0900
@@ -108,3 +108,14 @@
 	.density	= LPDDR2_DENSITY_2Gb,
 	.io_width	= LPDDR2_IO_WIDTH_32
 };
+struct lpddr2_device_info elpida_4G_S4 = {
+	.device_timings = {
+		&timings_elpida_200_mhz,
+		&timings_elpida_333_mhz,
+		&timings_elpida_400_mhz
+	},
+	.min_tck	= &min_tck_elpida,
+	.type		= LPDDR2_TYPE_S4,
+	.density	= LPDDR2_DENSITY_4Gb,
+	.io_width	= LPDDR2_IO_WIDTH_32
+};
