Index: kernel/drivers/media/video/dmm/tmm_pat.c
===================================================================
--- kernel.orig/drivers/media/video/dmm/tmm_pat.c	2011-04-08 22:30:28.000000000 +0900
+++ kernel/drivers/media/video/dmm/tmm_pat.c	2011-04-08 22:30:45.217145035 +0900
@@ -255,6 +255,7 @@
 				} else {
 					__free_page(
 						((struct mem *)f->mem[i])->pg);
+					kfree(f->mem[i]);
 					count--;
 				}
 			}
