Index: kernel/sound/soc/omap/omap-mcbsp.c
===================================================================
--- kernel.orig/sound/soc/omap/omap-mcbsp.c	2011-04-18 15:44:24.000000000 +0900
+++ kernel/sound/soc/omap/omap-mcbsp.c	2011-04-18 19:14:56.098428147 +0900
@@ -447,9 +447,6 @@
 	 * phase type	: dual 
 	 * Transmit word length  : 16bit
 	 * frame length : 32bit 
-	 * 
-	 * REVIST:
-	 * Currently, voice is clear with 32 bit word length.
 	 */
 #if 0
         /* Fix for voice clarity */
@@ -614,17 +611,8 @@
 		 * FS active low. TX data driven on falling edge of bit clock
 		 * and RX data sampled on rising edge of bit clock.
 		 */
-#if 0
 		regs->pcr0	|= FSXP | FSRP |
 				   CLKXP | CLKRP;
-#else
-		/*
-		 * [T1] - voicecall : As per CP recommendation,
-		 * TX data driven on rising edge of bit clock
-		 * and RX data sampled on falling edge of bit clock.
-		 */
-		regs->pcr0	|= FSXP | FSRP;
-#endif
 		break;
 	case SND_SOC_DAIFMT_NB_IF:
 		regs->pcr0	|= CLKXP | CLKRP;
