Index: kernel/sound/soc/omap/sdp4430.c
===================================================================
--- kernel.orig/sound/soc/omap/sdp4430.c	2011-04-22 22:30:21.118389034 +0900
+++ kernel/sound/soc/omap/sdp4430.c	2011-04-23 12:17:13.467148004 +0900
@@ -189,23 +189,22 @@
 	int ret = 0;
 
 	if (be_id == OMAP_ABE_DAI_BT_VX) {
+		/* 
+		 * BT chip and modem uses I2S protocal during voicecall. 
+		 * So McBSP1 also should use I2S. 
+		 */
 		 ret = snd_soc_dai_set_fmt(cpu_dai,
-                                   SND_SOC_DAIFMT_DSP_B |
-                                   SND_SOC_DAIFMT_NB_IF |
+                                   /* SND_SOC_DAIFMT_DSP_B | */
+                                   SND_SOC_DAIFMT_I2S |
+                                   /* SND_SOC_DAIFMT_NB_IF | */
+                                   SND_SOC_DAIFMT_NB_NF |
                                    SND_SOC_DAIFMT_CBM_CFM);
 		}
 	else{
 		ret = snd_soc_dai_set_fmt(cpu_dai,
-		/* 
-		 * BT chip and modem uses I2S protocal during voicecall. 
-		 * So McBSP1 also should use I2S. 
-		 * 
-		 * Once BT and modem changes to PCM, then McBSP1 
-		 * should also use PCM.
-		 */
-                                 /* SND_SOC_DAIFMT_DSP_B | */
-				  SND_SOC_DAIFMT_I2S |
-                                  SND_SOC_DAIFMT_NB_IF |
+                                  /* SND_SOC_DAIFMT_DSP_B | */
+                                  SND_SOC_DAIFMT_I2S |
+                                  SND_SOC_DAIFMT_NB_NF |
                                   SND_SOC_DAIFMT_CBM_CFM);
 	}
 
@@ -288,25 +287,28 @@
 	unsigned int be_id = rtd->dai_link->be_id;
 	unsigned int threshold;
 
-        be_id = rtd->dai_link->be_id;
-
 	switch (be_id) {
 	case OMAP_ABE_DAI_MM_FM:
 		channels->min = 2;
 		threshold = 2;
 		break;
 	case OMAP_ABE_DAI_BT_VX:
-		channels->min = 1;
 		/*	 
 		 * [T1] - voicecall: As per CP & 2mic recommendation,
-	 	 * phase type	: dual 
+	 	 * phase type	: dual mono
 	 	 * Transmit word length  : 16bit
 	 	 * frame length : 32bit 
 	 	 * 
 	 	 * REVIST:
-	 	 * Currently, voice is clear with 32 bit word length.
+	 	 * After BT pcm config finalization.
 	 	 */
+#if 0
+		channels->min = 1;
 		threshold = 1;
+#else
+		channels->min = 2;
+		threshold = 2;
+#endif
 		break;
 	deafult:
 		threshold = 1;
Index: kernel/sound/soc/omap/omap-mcbsp.c
===================================================================
--- kernel.orig/sound/soc/omap/omap-mcbsp.c	2011-04-22 22:30:21.088356660 +0900
+++ kernel/sound/soc/omap/omap-mcbsp.c	2011-04-23 12:18:16.927147814 +0900
@@ -464,9 +464,12 @@
 	regs->rcr1	|= RFRLEN1(wpf - 1);
 	regs->xcr1	|= XFRLEN1(wpf - 1);
 #else
+	if (/*channels == 2 && */(format == SND_SOC_DAIFMT_I2S ||
+			      format == SND_SOC_DAIFMT_LEFT_J)) {
 		/* Use dual-phase frames */
 		regs->rcr2	|= RPHASE;
 		regs->xcr2	|= XPHASE;
+	}
 #endif
 
 	switch (params_format(params)) {
Index: kernel/sound/soc/omap/omap-abe.c
===================================================================
--- kernel.orig/sound/soc/omap/omap-abe.c	2011-04-22 22:30:21.000000000 +0900
+++ kernel/sound/soc/omap/omap-abe.c	2011-04-23 12:11:50.747147915 +0900
@@ -1099,14 +1099,16 @@
 				if (stream == SNDRV_PCM_STREAM_PLAYBACK) {
 					/* BT_DL connection to McBSP 1 ports */
 					format.f = 8000;
-					format.samp_format = MONO_RSHIFTED_16;
+					//format.samp_format = MONO_RSHIFTED_16;
+					format.samp_format = STEREO_RSHIFTED_16;
 					abe_connect_serial_port(BT_VX_DL_PORT,
 								&format, MCBSP1_TX);
 					abe_dai_enable_data_transfer(BT_VX_DL_PORT);
 				} else {
 					/* BT_UL connection to McBSP 1 ports */
 					format.f = 8000;
-					format.samp_format = MONO_RSHIFTED_16;
+					//format.samp_format = MONO_RSHIFTED_16;
+					format.samp_format = STEREO_RSHIFTED_16;
 					abe_connect_serial_port(BT_VX_UL_PORT,
 								&format, MCBSP1_RX);
 					abe_dai_enable_data_transfer(BT_VX_UL_PORT);
