Index: kernel/arch/arm/mach-omap2/ipu_drv.c
===================================================================
--- kernel.orig/arch/arm/mach-omap2/ipu_drv.c	2011-04-20 11:47:09.000000000 +0900
+++ kernel/arch/arm/mach-omap2/ipu_drv.c	2011-04-22 10:16:57.000000000 +0900
@@ -229,7 +229,9 @@
 				/* sysm3 is handling hibernation of ducati
 				 * currently
 				 */
-				ipu_pm_save_ctx(SYS_M3);
+				retval = ipu_pm_save_ctx(SYS_M3);
+				if (retval)
+					goto error;
 				need_resume_notifications = 1;
 			}
 		}
Index: kernel/drivers/dsp/syslink/ipu_pm/ipu_pm.c
===================================================================
--- kernel.orig/drivers/dsp/syslink/ipu_pm/ipu_pm.c	2011-04-20 11:47:40.000000000 +0900
+++ kernel/drivers/dsp/syslink/ipu_pm/ipu_pm.c	2011-04-22 10:20:43.000000000 +0900
@@ -2474,13 +2474,6 @@
 		return 0;
 	}
 
-#ifdef CONFIG_OMAP_PM
-	retval = omap_pm_set_max_sdma_lat(&pm_qos_handle_2,
-						IPU_PM_NO_MPU_LAT_CONSTRAINT);
-	if (retval)
-		pr_info("Unable to remove cstr on IPU\n");
-#endif
-
 	/* Because of the current scheme, we need to check
 	 * if APPM3 is enable and we need to shut it down too
 	 * Sysm3 is the only want sending the hibernate message
@@ -2548,6 +2541,12 @@
 			pr_err("Not able to save iommu");
 	} else
 		goto error;
+#ifdef CONFIG_OMAP_PM
+		retval = omap_pm_set_max_sdma_lat(&pm_qos_handle_2,
+							IPU_PM_NO_MPU_LAT_CONSTRAINT);
+		if (retval)
+			pr_info("Unable to remove cstr on IPU\n");
+#endif
 exit:
 	mutex_unlock(ipu_pm_state.gate_handle);
 	return 0;
