Index: kernel/arch/arm/mach-omap2/board-t1.c
===================================================================
--- kernel.orig/arch/arm/mach-omap2/board-t1.c	2011-04-30 16:53:23.000000000 +0900
+++ kernel/arch/arm/mach-omap2/board-t1.c	2011-04-30 18:58:42.000000000 +0900
@@ -587,10 +587,19 @@
 static bool fsa9480_is_mhl_attached(void)
 {
 	int val;
-
-	gpio_request(OMAP_GPIO_MHL_SEL, "MHL_SEL");
-	val = gpio_get_value(OMAP_GPIO_MHL_SEL);
-	gpio_free(OMAP_GPIO_MHL_SEL);
+	int ver = 0;
+        ver=get_hw_rev_gpio();
+	if(ver<6){
+		gpio_request(OMAP_GPIO_MHL_SEL, "MHL_SEL");
+		val = gpio_get_value(OMAP_GPIO_MHL_SEL);
+		gpio_free(OMAP_GPIO_MHL_SEL);
+	}
+	else
+	{
+		gpio_request(OMAP_GPIO_ACC_EN,"MHL_SEL");
+		val = gpio_get_value(OMAP_GPIO_ACC_EN);
+		 gpio_free(OMAP_GPIO_ACC_EN);
+	}
 
 	return !!val;
 }
@@ -599,6 +608,9 @@
 
 static void omap4_mhl_cb(u8 attached)
 {
+
+	int ver = 0;
+        ver=get_hw_rev_gpio();
 	int ret=0; 
 	    /*TODO */
        printk("\nBoard file [FSA9480]: MHL Callback == %d\n", attached);
@@ -621,12 +633,16 @@
 			 printk("\nJUST return\n");
 			 return;
 		}
-		else
+		else{
 		   MHL_On(1);
+		}
 #endif  
 	 } else {
 	 	if (!mhl_audio_path_info){
-	 		gpio_set_value(OMAP_GPIO_MHL_SEL,0);
+			if(ver<6)
+	 			gpio_set_value(OMAP_GPIO_MHL_SEL,0);
+			else if(ver >= 6)
+				gpio_set_value(OMAP_GPIO_ACC_EN,0);
 	 		MHL_On(0);
 	 	}
 		else{ //audio dock path
@@ -2470,7 +2486,6 @@
 }
 
 
-
 /* WLAN */
 
 /* OMAP4 tablet 10-1 specific WLAN GPIOs */ 
Index: kernel/drivers/media/video/omap/mhl_v1/Common_Def.h
===================================================================
--- kernel.orig/drivers/media/video/omap/mhl_v1/Common_Def.h	2011-04-30 16:53:47.000000000 +0900
+++ kernel/drivers/media/video/omap/mhl_v1/Common_Def.h	2011-04-30 18:58:42.000000000 +0900
@@ -157,6 +157,9 @@
 #define GPIO_MHL_RST    OMAP_GPIO_MHL_RST
 #define GPIO_MHL_INT    OMAP_GPIO_MHL_INT
 #define GPIO_MHL_SEL    OMAP_GPIO_MHL_SEL
+//For Rev0.4 and onwards, MHL_SEL and ACC_EN pins swapped
+#define GPIO_ACC_EN 	OMAP_GPIO_ACC_EN
+
 #define GPIO_MHL_WAKE_UP        OMAP_GPIO_MHL_WAKEUP
 
 #define GPIO_MHL_INT_AF
@@ -183,7 +186,7 @@
 #define GPIO_LEVEL_LOW  0
 #define GPIO_LEVEL_HIGH 1
 
-
+u8 hw_get_rev_gpio();
 enum
 {
     MHL_MSC_MSG_RCP             = 0x10,     // RCP sub-command
Index: kernel/drivers/media/video/omap/mhl_v1/sii9234.c
===================================================================
--- kernel.orig/drivers/media/video/omap/mhl_v1/sii9234.c	2011-04-30 16:53:47.000000000 +0900
+++ kernel/drivers/media/video/omap/mhl_v1/sii9234.c	2011-04-30 18:58:42.000000000 +0900
@@ -121,6 +121,17 @@
 struct device *mhl_switch;
 EXPORT_SYMBOL(mhl_switch);
 
+u8 hw_get_rev_gpio()
+{
+        u8 hw_rev = 0xFF;
+
+        hw_rev = gpio_get_value(OMAP_GPIO_HW_REV3) << 3;
+        hw_rev |= gpio_get_value(OMAP_GPIO_HW_REV2) << 2;
+        hw_rev |= gpio_get_value(OMAP_GPIO_HW_REV1) << 1;
+        hw_rev |= gpio_get_value(OMAP_GPIO_HW_REV0);
+return hw_rev;
+}
+
 static ssize_t check_MHL_command(struct device *dev, struct device_attribute *attr, char *buf)
 {
 	int count;
@@ -149,7 +160,13 @@
 		}
 		sii9234_cfg_power(1);
 		sii9234_init();
-		gpio_set_value(GPIO_MHL_SEL, GPIO_LEVEL_HIGH);
+		int ver =0;
+    		ver = hw_get_rev_gpio();
+    		if(ver>=6)
+			gpio_set_value(GPIO_ACC_EN, GPIO_LEVEL_HIGH);
+       		else
+			gpio_set_value(GPIO_MHL_SEL, GPIO_LEVEL_HIGH);
+		
 	} else {
 		sii9234_cfg_power(0);
 		gpio_set_value(GPIO_MHL_SEL, GPIO_LEVEL_LOW);
@@ -159,13 +176,22 @@
 
 void MHL_On(bool on)
 {
+	int ver =0;
+	ver = hw_get_rev_gpio();
+	printk("Hardware revision:%d\n",ver);
 	if (on == 1) {
 		sii9234_cfg_power(1);
 		sii9234_init();
+	if(ver>=6)
+		gpio_direction_output(GPIO_ACC_EN, GPIO_LEVEL_HIGH);
+	else	
 		gpio_direction_output(GPIO_MHL_SEL, GPIO_LEVEL_HIGH); 
 	} else {
 		sii9234_cfg_power(0);
-		gpio_direction_output(GPIO_MHL_SEL, GPIO_LEVEL_LOW); 
+		if(ver>=6)
+			gpio_direction_output(GPIO_ACC_EN, GPIO_LEVEL_HIGH);
+		else
+			gpio_direction_output(GPIO_MHL_SEL, GPIO_LEVEL_LOW); 
 	}
 
 }
@@ -359,10 +385,16 @@
 irqreturn_t mhl_int_irq_handler(int irq, void *dev_id)
 {
 	printk(KERN_ERR "mhl_int_irq_handler() is called\n");
-	printk("%s MHL_SEL=%d\n",__func__,gpio_get_value(GPIO_MHL_SEL));
-	if (gpio_get_value(GPIO_MHL_SEL))	
-		mhl_int_irq_handler_sched();
-
+	int ver =0;
+	ver = hw_get_rev_gpio();
+	if(ver>=6){
+		if (gpio_get_value(GPIO_ACC_EN))
+                	mhl_int_irq_handler_sched();
+	}
+	else{
+		if (gpio_get_value(GPIO_MHL_SEL))	
+			mhl_int_irq_handler_sched();
+	}
 	return IRQ_HANDLED;
 }
 
@@ -641,11 +673,19 @@
          printk("%s error in making GPIO_HDMI_EN Low\n",__func__);
     gpio_request(GPIO_MHL_RST,NULL);
     if(gpio_direction_output(GPIO_MHL_RST,GPIO_LEVEL_LOW))
-        printk("%s error in making GPIO_HDMI_EN Low\n",__func__);
-    gpio_request(GPIO_MHL_SEL,NULL);
-    if(gpio_direction_output(GPIO_MHL_SEL,GPIO_LEVEL_LOW))
-        printk("%s error in making GPIO_HDMI_EN Low\n",__func__);
-
+        printk("%s error in making GPIO_MHL_RST Low\n",__func__);
+    int ver =0;
+    ver = hw_get_rev_gpio();
+    if(ver>=6){
+	gpio_request(GPIO_ACC_EN,NULL);
+    	if(gpio_direction_output(GPIO_ACC_EN,GPIO_LEVEL_LOW))
+        printk("%s error in making GPIO_ACC_EN Low\n",__func__);
+	}
+    else{
+    	 gpio_request(GPIO_MHL_SEL,NULL);
+    	 if(gpio_direction_output(GPIO_MHL_SEL,GPIO_LEVEL_LOW))
+        	printk("%s error in making GPIO_MHL_SEL Low\n",__func__);
+	}
 }
 
 static int mhl_open(struct inode *ip, struct file *fp)
Index: kernel/drivers/media/video/omap/mhl_v1/sii9234_driver.c
===================================================================
--- kernel.orig/drivers/media/video/omap/mhl_v1/sii9234_driver.c	2011-04-30 16:53:47.000000000 +0900
+++ kernel/drivers/media/video/omap/mhl_v1/sii9234_driver.c	2011-04-30 19:05:44.000000000 +0900
@@ -756,6 +756,8 @@
 ////////////////////////////////////////////////////////////////////
 void	Int1RsenIsr( void )
 {
+	u8  ver =0;
+	ver=hw_get_rev_gpio();
 	byte		reg71 = I2C_ReadByte(SA_TX_Page0_Primary, 0x71);
 	byte		rsen  = I2C_ReadByte(SA_TX_Page0_Primary, 0x09) & BIT_2;
 	// Look at RSEN interrupt
@@ -803,7 +805,10 @@
 #endif
 					dsHpdStatus = 0;
 					sii9234_cfg_power(0);
-					gpio_direction_output(GPIO_MHL_SEL,GPIO_LEVEL_LOW);
+					if(ver >=6)
+						gpio_direction_output(GPIO_ACC_EN,GPIO_LEVEL_LOW);
+					else
+						gpio_direction_output(GPIO_MHL_SEL,GPIO_LEVEL_LOW);
          			}
         		 	else
         			{
@@ -830,7 +835,10 @@
 				sii9234_cfg_power(0);
 				//s3c_gpio_setpull(GPIO_MHL_SEL, S3C_GPIO_PULL_DOWN);
 				//s3c_gpio_cfgpin(GPIO_MHL_INT, S3C_GPIO_INPUT); 
-				gpio_direction_output(GPIO_MHL_SEL, GPIO_LEVEL_LOW);
+				if(ver>=6)
+					gpio_direction_output(GPIO_ACC_EN, GPIO_LEVEL_LOW);
+				else
+					gpio_direction_output(GPIO_MHL_SEL, GPIO_LEVEL_LOW);
 				printk(KERN_ERR "MHL_SEL to 0\n");
 			}
 
