Index: kernel/sound/soc/omap/omap-abe-dsp.c
===================================================================
--- kernel.orig/sound/soc/omap/omap-abe-dsp.c	2011-04-30 20:19:44.267257644 +0900
+++ kernel/sound/soc/omap/omap-abe-dsp.c	2011-04-30 20:19:55.949619270 +0900
@@ -263,7 +263,11 @@
 	if (abe->first_irq) {
 		abe->first_irq = 0;
 	} else {
-		if (abe->psubs)
+		/* 
+		 * Do not call ALSA function when substream may not have a buffer and
+		 * also when runtime information is NULL.
+		 */
+		if (abe->psubs && (abe->psubs->hw_no_buffer) && (abe->psubs->runtime)) 
 			snd_pcm_period_elapsed(abe->psubs);
 	}
 }
