Index: kernel/sound/soc/omap/sdp4430.c
===================================================================
--- kernel.orig/sound/soc/omap/sdp4430.c	2011-05-04 18:09:22.867185790 +0900
+++ kernel/sound/soc/omap/sdp4430.c	2011-05-04 20:15:41.380854634 +0900
@@ -259,6 +259,11 @@
                                   SND_SOC_DAIFMT_CBM_CFM);
 	}
 
+	if (rtd->current_fe == ABE_FRONTEND_DAI_MODEM) {
+		/* set Modem McBSP configuration  */
+		ret = sdp4430_modem_mcbsp_configure(substream, params, 1);
+	}
+
 	if (ret < 0) {
 		printk(KERN_ERR "can't set cpu DAI configuration\n");
 		return ret;
