Index: kernel/drivers/video/omap2/dss/hdmi.c
===================================================================
--- kernel.orig/drivers/video/omap2/dss/hdmi.c	2011-05-20 13:46:41.000000000 +0900
+++ kernel/drivers/video/omap2/dss/hdmi.c	2011-05-20 13:52:59.000000000 +0900
@@ -1433,18 +1433,6 @@
 			goto done;
 
 		
-		/*
-		 * We only clear timings when hpd notification is emminent.
-		 * By contract we guarantee an hpd 1 event after each time we
-		 * clear the timings (if HDMI is plugged in), in case a previous
-		 * hpd 1 event is being processed after the clearing.  In this
-		 * case the listener would read all 0-s, but can simply assume
-		 * HDMI is off as it will get another hpd 1 event if HDMI is
-		 * plugged in (after EDID is successfully read).
-		 */
-		memset(&dssdev->panel.timings, 0,
-						sizeof(dssdev->panel.timings));
-
 		/* turn audio power off */
 		notify = !audio_on; /* notification is sent if audio is on */
 		hdmi_audio_power_off();
@@ -1688,9 +1676,6 @@
 {
 	edid_set = false;
 
-	/* clear timings */
-	memset(&dssdev->panel.timings, 0, sizeof(dssdev->panel.timings));
-
 	HDMI_W1_StopVideoFrame(HDMI_WP);
 	hdmi_set_irqs(1);
 
