Index: kernel/arch/arm/mach-omap2/dpll-44xx.c
===================================================================
--- kernel.orig/arch/arm/mach-omap2/dpll-44xx.c	2011-05-21 00:24:21.748347866 +0900
+++ kernel/arch/arm/mach-omap2/dpll-44xx.c	2011-05-21 00:38:21.049628440 +0900
@@ -38,11 +38,6 @@
 #define LP_98M_RATE		98304000
 #define LP_DELAY		1000000
 
-#ifdef CONFIG_OMAP_PM
-#include <plat/omap-pm.h>
-static struct pm_qos_request_list *pm_qos_dpll_handle;
-#endif
-
 
 static struct delayed_work	lpmode_work;
 struct clk *dpll_core_clk=NULL;
@@ -597,17 +592,6 @@
 	struct omap_opp *opp;
 	struct voltagedomain *vdd_mpu, *vdd_iva, *vdd_core;
 
-	pr_info("[%s]\n", __func__);
-
-#ifdef CONFIG_OMAP_PM
-	ret = omap_pm_set_max_mpu_wakeup_lat(&pm_qos_dpll_handle, 7);
-	if (ret) {
-		pr_err("%s %d Error setting MPU Constrain\n", __func__, __LINE__);
-		return ret;
-	}
-
-#endif
-
 	dpll_abe_ck = clk_get(NULL, "dpll_abe_ck");
 	dpll_mpu_ck = clk_get(NULL, "dpll_mpu_ck");
 	div_mpu_hs_clk = clk_get(NULL, "div_mpu_hs_clk");
@@ -949,14 +933,6 @@
 		cancel_delayed_work_sync(&lpmode_work);
 
 	if (!omap4_lpmode) {
-#ifdef CONFIG_OMAP_PM
-		ret = omap_pm_set_max_mpu_wakeup_lat(&pm_qos_dpll_handle, -1);
-		if (ret) {
-			pr_err("%s %d Error setting MPU Constrain\n", __func__, __LINE__);
-			return ret;
-		}
-
-#endif
 		return 0;
 	}
 
@@ -1064,13 +1040,5 @@
 
 out:
 
-#ifdef CONFIG_OMAP_PM
-	ret = omap_pm_set_max_mpu_wakeup_lat(&pm_qos_dpll_handle, -1);
-	if (ret) {
-		pr_err("%s %d Error setting MPU Constrain\n", __func__, __LINE__);
-		return ret;
-	}
-
-#endif
 	return ret;
 }
