Index: kernel/arch/arm/plat-omap/devices.c
===================================================================
--- kernel.orig/arch/arm/plat-omap/devices.c	2011-05-21 01:29:26.000000000 +0900
+++ kernel/arch/arm/plat-omap/devices.c	2011-05-21 19:56:11.278389177 +0900
@@ -395,8 +395,10 @@
 	}
 
 	pdata->get_context_loss_count = omap_pm_get_dev_context_loss_count;
+#if 0
 	pdata->enter_dpll_cascade = omap4_dpll_low_power_cascade_check_entry;
 	pdata->exit_dpll_cascade = omap4_dpll_low_power_cascade_exit;
+#endif
 
 	od = omap_device_build("omap-aess-audio", -1, oh, pdata,
 				sizeof(struct omap4_abe_dsp_pdata),
Index: kernel/arch/arm/mach-omap2/pm44xx.c
===================================================================
--- kernel.orig/arch/arm/mach-omap2/pm44xx.c	2011-05-21 19:53:09.000000000 +0900
+++ kernel/arch/arm/mach-omap2/pm44xx.c	2011-05-21 20:04:44.609639196 +0900
@@ -486,7 +486,7 @@
 		omap_writel(0x2, 0x4A009550);
 		omap_writel(0xD, 0x48020054);
 		/* Modem HSI wakeup */
-		if (omap_hsi_is_io_wakeup_from_hsi() && !omap4_lpmode)
+		if (omap_hsi_is_io_wakeup_from_hsi())
 			omap_hsi_wakeup();
 
 		/* usbhs remote wakeup */
