Index: kernel/arch/arm/plat-omap/hdmi_lib.c
===================================================================
--- kernel.orig/arch/arm/plat-omap/hdmi_lib.c	2011-05-14 18:21:21.000000000 +0900
+++ kernel/arch/arm/plat-omap/hdmi_lib.c	2011-05-24 23:10:22.000000000 +0900
@@ -1870,6 +1870,11 @@
 	if (val & HDMI_WP_IRQSTATUS_PHYCONNECT) {
 		DBG("connect, ");
 		*r |= HDMI_CONNECT;
+		u32 temp = omap_readl(0x4804602C);
+                printk("WP_IRQ_ENABLE_SET read as %x\n",temp);
+                temp =(temp | (1<<4));
+                omap_writel(temp,0x4804602C);
+                printk("Enable OCP_timeout Interrupt..WP_IRQ_ENABLE_SET read as %x\n",temp);
 	}
 
 	if ((val & HDMI_WP_IRQSTATUS_CORE) && (core_state & 0x1)) {
@@ -1906,10 +1911,17 @@
 		dirty = 0;
 		first_hpd = 0;
 		*r |= HDMI_DISCONNECT;
+#if 0
+		u32 temp = omap_readl(0x48046030);
+                printk("WP_IRQ_ENABLE_CLR read as %x\n",temp);
+                temp =(temp | (1<<4));
+                omap_writel(temp,0x48046030);
+                printk("Cleared OCP_timeout if any..WP_IRQ_ENABLE_CLR read as %x\n",temp);
+#endif
 	}
 
 	/*
-	   Handle ocp timeout in correct way. SCP clock will reset the divder
+	   Handle obcp timeout in correct way. SCP clock will reset the divder
 	   to 7 automatically. We need to speed up the clock to avoid performance
 	   degradation.
 	*/
Index: kernel/drivers/media/video/omap/mhl_v1/sii9234_driver.c
===================================================================
--- kernel.orig/drivers/media/video/omap/mhl_v1/sii9234_driver.c	2011-05-14 18:21:43.000000000 +0900
+++ kernel/drivers/media/video/omap/mhl_v1/sii9234_driver.c	2011-05-24 22:39:48.000000000 +0900
@@ -841,8 +841,12 @@
 
 			           TX_DEBUG_PRINT((KERN_ERR "Really RSEN Low\n"));
 					//mhl_cable_status =MHL_INIT_POWER_OFF;
+					u32 temp = omap_readl(0x48046030);
+                			printk("WP_IRQ_ENABLE_CLR read as %x\n",temp);
+                			temp =(temp | (1<<4));
+                			omap_writel(temp,0x48046030);
+                			printk("Cleared OCP_timeout if any..WP_IRQ_ENABLE_CLR read as %x\n",temp);
           
-
 					mhl_cable_status = MHL_TV_OFF_CABLE_CONNECT;
 #if defined(CONFIG_S5PC110_DEMPSEY_BOARD)
 	FSA9480_MhlSwitchSel(0);
Index: kernel/drivers/media/video/omap/mhl_v1/sii9234.c
===================================================================
--- kernel.orig/drivers/media/video/omap/mhl_v1/sii9234.c	2011-05-24 15:57:00.000000000 +0900
+++ kernel/drivers/media/video/omap/mhl_v1/sii9234.c	2011-05-25 12:07:53.000000000 +0900
@@ -180,12 +180,12 @@
 	ver = hw_get_rev_gpio();
 	printk("Hardware revision:%d\n",ver);
 	if (on == 1) {
+		if(ver>=6)
+			gpio_direction_output(GPIO_ACC_EN, GPIO_LEVEL_HIGH);
+		else	
+			gpio_direction_output(GPIO_MHL_SEL, GPIO_LEVEL_HIGH); 
 		sii9234_cfg_power(1);
 		sii9234_init();
-	if(ver>=6)
-		gpio_direction_output(GPIO_ACC_EN, GPIO_LEVEL_HIGH);
-	else	
-		gpio_direction_output(GPIO_MHL_SEL, GPIO_LEVEL_HIGH); 
 	} else {
 		sii9234_cfg_power(0);
 		if(ver>=6)
@@ -626,11 +626,6 @@
                 printk("[MHL_TEST]error in making HDMI_EN High\n");
            
 	    printk("[MHL_TEST]HDMI_EN:%d\n",gpio_get_value(GPIO_HDMI_EN));
-	 // Pulling(internal) up the HPD line
-               printk("Before (Powering up  MHL) : HPD-padconfig Read value:%x\n",omap_readl(0x4A100098));
-               omap_writel(0x000F0118,0x4A100098);
-               printk("After (Powering up MHL) : HPD-padconfigRead value:%x\n",omap_readl(0x4A100098));
-
 			
  #if 0
 		 if(gpio_direction_output(GPIO_MHL_RST, GPIO_LEVEL_LOW))
@@ -640,10 +635,16 @@
 
  #endif
 	
-	   msleep(200);
+	   msleep(20);
            if(gpio_direction_output(GPIO_MHL_RST, GPIO_LEVEL_HIGH))
                 printk("[MHL_TEST]error in making GPIO_MHL_RST Low\n");
            printk("[MHL_TEST]MHL_RST:%d",gpio_get_value(GPIO_MHL_RST));
+	#if 1 
+	// Pulling(internal) up the HPD line
+               printk("Before (Powering up  MHL) : HPD-padconfig Read value:%x\n",omap_readl(0x4A100098));
+               omap_writel(0x000F0118,0x4A100098);
+               printk("After (Powering up MHL) : HPD-padconfigRead value:%x\n",omap_readl(0x4A100098));
+	#endif
 	}
 	else
 	{
@@ -651,11 +652,12 @@
 		printk("sii9234_cfg_power off GPIO_HDMI_EN low\n");
 	  	gpio_direction_output(GPIO_HDMI_EN,GPIO_LEVEL_LOW);
           	gpio_direction_output(GPIO_MHL_RST, GPIO_LEVEL_LOW); 
-               
+          #if 1     
 		// Pulling(internal) Down the HPD line
                printk("Before (Powering down MHL) : HPD-padconfig Read value:%x\n",omap_readl(0x4A100098));
                omap_writel(0x000F0108,0x4A100098);
                printk("After (Powering down MHL) : HPD-padconfig Read value:%x\n",omap_readl(0x4A100098));
+	  #endif
 	}
 	printk(KERN_ERR "[MHL]%s : %d \n",__func__,on);
 
