Index: kernel/arch/arm/mach-omap2/voltage.c
===================================================================
--- kernel.orig/arch/arm/mach-omap2/voltage.c	2011-06-04 20:36:14.288349272 +0900
+++ kernel/arch/arm/mach-omap2/voltage.c	2011-06-04 21:22:10.668390309 +0900
@@ -396,7 +396,6 @@
 	{.volt_nominal = 1313000, .sr_errminlimit = 0xFA, .vp_errgain = 0x23, .abb_type = NOMINAL_OPP},
 	{.volt_nominal = 1374000, .sr_errminlimit = 0xFA, .vp_errgain = 0x27, .abb_type = FAST_OPP},
 	{.volt_nominal = 1375000, .sr_errminlimit = 0xFA, .vp_errgain = 0x27, .abb_type = FAST_OPP},
-
 };
 
 static struct omap_volt_data omap44xx_vdd_iva_volt_data[] = {
@@ -439,8 +438,8 @@
 	{.main_vdd_volt = 1025000, .dep_vdd_volt = 1025000},
 	{.main_vdd_volt = 1200000, .dep_vdd_volt = 1200000},
 	{.main_vdd_volt = 1313000, .dep_vdd_volt = 1200000},
-	{.main_vdd_volt = 1374000, .dep_vdd_volt = 120000},
-	{.main_vdd_volt = 1375000, .dep_vdd_volt = 120000},
+	{.main_vdd_volt = 1374000, .dep_vdd_volt = 1200000},
+	{.main_vdd_volt = 1375000, .dep_vdd_volt = 1200000},
 	{.main_vdd_volt = 0, .dep_vdd_volt = 0},
 };
 
@@ -2511,6 +2510,7 @@
 			OMAP4_CTRL_MODULE_PAD_CORE_CONTROL_EFUSE_1);
 		}
 	}
+
 #ifdef CONFIG_PM_DEBUG
 	voltage_dir = debugfs_create_dir("voltage", pm_dbg_main_dir);
 #endif
