Index: kernel/arch/arm/mach-omap2/dpll-44xx.c
===================================================================
--- kernel.orig/arch/arm/mach-omap2/dpll-44xx.c	2011-06-07 15:22:17.000000000 +0900
+++ kernel/arch/arm/mach-omap2/dpll-44xx.c	2011-06-07 15:23:45.000000000 +0900
@@ -69,6 +69,7 @@
 	unsigned long dpll_core_ck_rate;
 	u32 dpll_core_m2_div;
 	struct clk *per_hsd_byp_clk_mux_ck_parent;
+	u32 mpu_static_dependency_value;
 } state;
 
 struct dpll_cascading_blocker {
@@ -599,6 +600,18 @@
 }
 EXPORT_SYMBOL(dpll_cascading_blocker_release);
 
+/** Functions to save & restore MPU static dependency register values .
+ */
+static void save_mpu_static_dependency_value(void)
+{
+	state.mpu_static_dependency_value = __raw_readl(OMAP4430_CM_MPU_STATICDEP);
+}
+
+static void restore_mpu_static_dependency_value(void)
+{
+	__raw_writel(state.mpu_static_dependency_value, OMAP4430_CM_MPU_STATICDEP);
+}
+
 /**
  * omap4_dpll_low_power_cascade - configure system for low power DPLL cascade
  *
@@ -839,6 +852,9 @@
 
 	recalculate_root_clocks();
 
+	/* Save MPU static dependency value before clearing it. */
+	save_mpu_static_dependency_value();
+
 	/*explicitly clearing static dependency for MP3 PB*/
 	__raw_writel(0x00000020, OMAP4430_CM_MPU_STATICDEP);
 
@@ -1034,7 +1050,7 @@
 	recalculate_root_clocks();
 
 		/*put back static dependency default vlaues during exit*/
-		__raw_writel(0x0000B0E0, OMAP4430_CM_MPU_STATICDEP);
+	restore_mpu_static_dependency_value();
 
 out:
 
