Index: kernel/drivers/cpufreq/cpufreq_hotplug.c
===================================================================
--- kernel.orig/drivers/cpufreq/cpufreq_hotplug.c	2011-06-09 10:55:42.000000000 +0530
+++ kernel/drivers/cpufreq/cpufreq_hotplug.c	2011-06-09 19:22:31.272498383 +0530
@@ -56,6 +56,8 @@
 		unsigned int event);
 static int hotplug_boost(struct cpufreq_policy *policy);
 
+static struct device *mpu_dev;
+
 
 #ifndef CONFIG_CPU_FREQ_DEFAULT_GOV_HOTPLUG
 static
@@ -551,8 +553,6 @@
 
 			queue_work_on(this_dbs_info->cpu, khotplug_wq,
 							&this_dbs_info->cpu_up_work);
-
-			dpll_cascading_blocker_hold(mpu_dev);
 			goto out;
 		}
 	}
@@ -580,9 +580,6 @@
 				queue_work_on(this_dbs_info->cpu, khotplug_wq,
 								&this_dbs_info->cpu_down_work);
 
-				dpll_cascading_blocker_release(mpu_dev);
-
-
 			}
 			goto out;
 		}
@@ -614,11 +611,15 @@
 static void do_cpu_up(struct work_struct *work)
 {
 	cpu_up(1);
+	if (num_online_cpus() > 1)
+		dpll_cascading_blocker_hold(mpu_dev);
 }
 
 static void do_cpu_down(struct work_struct *work)
 {
 	cpu_down(1);
+	if (num_online_cpus() < 2)
+		dpll_cascading_blocker_release(mpu_dev);
 }
 
 static void do_dbs_timer(struct work_struct *work)
@@ -626,7 +627,7 @@
 	struct cpu_dbs_info_s *dbs_info =
 		container_of(work, struct cpu_dbs_info_s, work.work);
 	unsigned int cpu = dbs_info->cpu;
-	struct device *mpu_dev = omap2_get_mpuss_device();
+	mpu_dev = omap2_get_mpuss_device();
 
 	int delay = 0;
 	mutex_lock(&dbs_info->timer_mutex);
@@ -640,9 +641,6 @@
 		
 		if (num_online_cpus() < 2)
 			queue_work_on(cpu, khotplug_wq, &dbs_info->cpu_up_work);
-			
-		if (num_online_cpus() > 1)
-			dpll_cascading_blocker_hold(mpu_dev);
   	}
 	queue_delayed_work_on(cpu, khotplug_wq, &dbs_info->work, delay);
 	mutex_unlock(&dbs_info->timer_mutex);
Index: kernel/drivers/usb/otg/fsa9480.c
===================================================================
--- kernel.orig/drivers/usb/otg/fsa9480.c	2011-06-09 18:36:58.000000000 +0530
+++ kernel/drivers/usb/otg/fsa9480.c	2011-06-09 19:30:16.692498035 +0530
@@ -743,7 +743,7 @@
 			 * for LowPower mode When USB/MHL/TA module removals.
 			 */
 			if (dpll_cascading_blocker_release(&client->dev) < 0)
-				dev_warn(&client->dev, "Error releasing DPLL cascading constraint\n");
+				dev_info(&client->dev, "DPLL cascading constraint for this device does not exist\n");
 	}
 
        usbsw->dev1 = val1;
