--- \\10.254.123.46\sanghyub.lee\kiran\android\kernel\drivers\dsp\syslink\ipu_pm\ipu_pm.h	2011-06-13 19:12:15.000000000 +-0900
+++ \\10.254.123.46\sanghyub.lee\kiran\ipu_pm.h	2011-06-13 18:13:09.000000000 +-0900
@@ -101,13 +101,13 @@
 #define TESLA 0
 
 #define PM_CSTR_PERF_MASK	0x00000001
 #define PM_CSTR_LAT_MASK	0x00000002
 #define PM_CSTR_BW_MASK		0x00000004
 
-#define IPU_PM_MM_MPU_LAT_CONSTRAINT	400
+#define IPU_PM_MM_MPU_LAT_CONSTRAINT	10
 #define IPU_PM_NO_MPU_LAT_CONSTRAINT	-1
 #define NO_FREQ_CONSTRAINT		0
 #define NO_LAT_CONSTRAINT		-1
 #define NO_BW_CONSTRAINT		-1
 
 #define RCB_SIZE 8
