Index: kernel/sound/soc/omap/omap-mcpdm.c
===================================================================
--- kernel.orig/sound/soc/omap/omap-mcpdm.c	2011-06-15 00:59:18.432467704 +0530
+++ kernel/sound/soc/omap/omap-mcpdm.c	2011-06-15 01:06:26.563091969 +0530
@@ -703,9 +703,7 @@
 	mutex_lock(&mcpdm->mutex);
 
 	/* make sure we stop any pre-existing shutdown */
-	if (substream->stream == SNDRV_PCM_STREAM_PLAYBACK) {
-		cancel_delayed_work(&mcpdm->delayed_abe_work);
-	}
+	cancel_delayed_work(&mcpdm->delayed_abe_work);
 
 	if (!dai->active && mcpdm->free) {
 		ret = omap_mcpdm_request(mcpdm);
@@ -741,27 +739,9 @@
 		mcpdm->ul_active--;
 
 	if (!dai->active) {
-		if (!mcpdm->ul_active && substream->stream == SNDRV_PCM_STREAM_CAPTURE) {
-			if (!mcpdm->dl_active) {
-				abe_disable_data_transfer(PDM_DL_PORT);
-				abe_disable_data_transfer(PDM_UL_PORT);
-				udelay(250);
-				omap_mcpdm_stop_all(mcpdm);
-				omap_mcpdm_playback_close(mcpdm, mcpdm->downlink);
-				omap_mcpdm_capture_close(mcpdm, mcpdm->uplink);
-				abe_dsp_mcpdm_shutdown();
-				abe_dsp_pm_put();
-			}
-			if (!mcpdm->free && !mcpdm->dn_channels &&
-			    !mcpdm->dl_active)
-				omap_mcpdm_free(mcpdm);
-
-		}
-		if (!mcpdm->dl_active && substream->stream == SNDRV_PCM_STREAM_PLAYBACK) {
-			wake_lock_timeout(&mcpdm->wake_lock, 2 * HZ);
-			schedule_delayed_work(&mcpdm->delayed_abe_work,
-					      msecs_to_jiffies(1000)); /* TODO: pdata ? */
-		}
+		wake_lock_timeout(&mcpdm->wake_lock, 2 * HZ);
+		schedule_delayed_work(&mcpdm->delayed_abe_work,
+				      msecs_to_jiffies(1000)); /* TODO: pdata ? */
 	}
 
 	mutex_unlock(&mcpdm->mutex);
@@ -785,10 +765,11 @@
 		abe_dsp_mcpdm_shutdown();
 		abe_dsp_pm_put();
 	}
-	mutex_unlock(&mcpdm->mutex);
 
 	if (!mcpdm->free && !mcpdm->ul_active && !mcpdm->dl_active)
 		omap_mcpdm_free(mcpdm);
+	
+	mutex_unlock(&mcpdm->mutex);
 
 }
 
