Index: kernel/arch/arm/plat-omap/include/plat/hdmi_lib.h
===================================================================
--- kernel.orig/arch/arm/plat-omap/include/plat/hdmi_lib.h	2011-04-13 10:01:54.000000000 +0530
+++ kernel/arch/arm/plat-omap/include/plat/hdmi_lib.h	2011-06-14 10:12:41.632049694 +0530
@@ -564,5 +564,7 @@
 int hdmi_lib_start_acr_wa(void);
 int hdmi_lib_stop_acr_wa(void);
 #endif
+int hdmi_audio_w1_wrapper_disable(u32);
+int hdmi_audio_w1_wrapper_enable(u32);
 #endif
 
Index: kernel/drivers/video/omap2/dss/hdmi.c
===================================================================
--- kernel.orig/drivers/video/omap2/dss/hdmi.c	2011-06-14 10:12:16.122054509 +0530
+++ kernel/drivers/video/omap2/dss/hdmi.c	2011-06-14 10:12:41.632049694 +0530
@@ -2748,3 +2748,27 @@
 		DSSINFO("[HDMI]DSS = %d, enable=%d\n", dssdev->state);
 }
 EXPORT_SYMBOL(ocp_interrupt_clear);
+int hdmi_audio_w1_wrapper_disable(u32 instanceName)
+{
+	int ret = 0;
+	struct omap_dss_device *dssdev = get_hdmi_device();
+	mutex_lock(&hdmi.lock_aux);
+	if (dssdev->state == OMAP_DSS_DISPLAY_ACTIVE){
+	ret=hdmi_w1_wrapper_disable(instanceName);
+	}
+	mutex_unlock(&hdmi.lock_aux);
+	return ret;
+}
+EXPORT_SYMBOL(hdmi_audio_w1_wrapper_disable);
+int hdmi_audio_w1_wrapper_enable(u32 instanceName)
+{
+	int ret = 0;
+	struct omap_dss_device *dssdev = get_hdmi_device();
+	mutex_lock(&hdmi.lock_aux);
+	if (dssdev->state == OMAP_DSS_DISPLAY_ACTIVE){
+	ret=hdmi_w1_wrapper_enable(instanceName);
+	}
+	mutex_unlock(&hdmi.lock_aux);
+	return ret;
+}
+EXPORT_SYMBOL(hdmi_audio_w1_wrapper_enable);
Index: kernel/sound/soc/omap/omap-hdmi.c
===================================================================
--- kernel.orig/sound/soc/omap/omap-hdmi.c	2011-04-13 10:01:54.000000000 +0530
+++ kernel/sound/soc/omap/omap-hdmi.c	2011-06-14 10:12:41.632049694 +0530
@@ -102,14 +102,14 @@
 		break;
 	case HDMI_EVENT_POWERPHYON:
 		if (substream) {
-			hdmi_w1_wrapper_enable(HDMI_WP);
+			hdmi_audio_w1_wrapper_enable(HDMI_WP);
 			hdmi_w1_start_audio_transfer(HDMI_WP);
 		}
 		break;
 	case HDMI_EVENT_POWERPHYOFF:
 		if (substream) {
 			hdmi_w1_stop_audio_transfer(HDMI_WP);
-			hdmi_w1_wrapper_disable(HDMI_WP);
+			hdmi_audio_w1_wrapper_disable(HDMI_WP);
 		}
 		break;
 	default:
@@ -134,7 +134,7 @@
 #endif
 	hdmi_set_audio_power(1);
 	hdmi_data.substream = substream;
-	err = hdmi_w1_wrapper_enable(HDMI_WP);
+	err = hdmi_audio_w1_wrapper_enable(HDMI_WP);
 	return err;
 }
 
@@ -142,7 +142,7 @@
 				    struct snd_soc_dai *dai)
 {
 	int err = 0;
-	hdmi_w1_wrapper_disable(HDMI_WP);
+	hdmi_audio_w1_wrapper_disable(HDMI_WP);
 	hdmi_data.substream = NULL;
 
 	if (hdmi_data.active)
