Index: kernel/drivers/cpufreq/cpufreq_hotplug.c
===================================================================
--- kernel.orig/drivers/cpufreq/cpufreq_hotplug.c	2011-06-18 16:07:17.000000000 +0900
+++ kernel/drivers/cpufreq/cpufreq_hotplug.c	2011-06-18 16:10:39.000000000 +0900
@@ -610,11 +610,10 @@
 
 static void do_cpu_up(struct work_struct *work)
 {
-
-	if (num_online_cpus()  < 2 )
-		dpll_cascading_blocker_hold(&hotplug_dev);
-
 	cpu_up(1);
+	
+	if (num_online_cpus()  > 1 )
+		dpll_cascading_blocker_hold(&hotplug_dev);
 }
 
 static void do_cpu_down(struct work_struct *work)
