Index: kernel/arch/arm/plat-omap/hdmi_lib.c
===================================================================
--- kernel.orig/arch/arm/plat-omap/hdmi_lib.c	2011-06-17 11:59:28.000000000 +0530
+++ kernel/arch/arm/plat-omap/hdmi_lib.c	2011-06-24 11:49:58.202046798 +0530
@@ -70,6 +70,7 @@
 #define HDMI_WP_PWR_CTRL			0x40ul
 #define HDMI_WP_DEBOUNCE			0x44ul
 #define HDMI_WP_IRQENABLE_SET			0x2Cul
+#define HDMI_WP_IRQENABLE_CLR			0x30ul
 #define HDMI_WP_VIDEO_CFG			0x50ul
 #define HDMI_WP_VIDEO_SIZE			0x60ul
 #define HDMI_WP_VIDEO_TIMING_H			0x68ul
Index: kernel/drivers/video/omap2/dss/hdmi.c
===================================================================
--- kernel.orig/drivers/video/omap2/dss/hdmi.c	2011-06-24 11:49:57.012050170 +0530
+++ kernel/drivers/video/omap2/dss/hdmi.c	2011-06-24 12:20:02.699543966 +0530
@@ -1564,6 +1564,10 @@
 
 	if ((action & HDMI_DISCONNECT) && !(r & HDMI_IN_RESET) &&
 	    (hdmi_power == HDMI_POWER_FULL)) {
+		if (hdmi_suspend) {  
+			DSSINFO ("Ignoring disconnect during suspend\n");  
+			goto done2;  
+		}  
 		/*
 		 * Wait at least 100ms after HDMI_CONNECT to decide if
 		 * cable is really disconnected
@@ -1797,6 +1801,7 @@
 
 static void hdmi_power_off(struct omap_dss_device *dssdev)
 {
+	hdmi_power = HDMI_POWER_OFF;
 	if (!in_hdmi_restart)
 	set_hdmi_hot_plug_status(dssdev, false);
 	/* ignore return value for now */
@@ -2788,6 +2793,38 @@
 	else if( dssdev->state == OMAP_DSS_DISPLAY_SUSPENDED)
 		DSSINFO("[HDMI]DSS suspend\n");
 	else
-		DSSINFO("[HDMI]DSS = %d, enable=%d\n", dssdev->state);
+		DSSINFO("[HDMI]DSS = enable=%d\n", dssdev->state);
 }
 EXPORT_SYMBOL(ocp_interrupt_clear);
+int hdmi_wrapper_enable(u32 instanceName) 
+{
+	int ret = 0;
+	if (hdmi_opt_clk_state)
+		ret = hdmi_w1_wrapper_enable(instanceName);
+	return ret;
+}
+EXPORT_SYMBOL(hdmi_wrapper_enable);
+int hdmi_wrapper_disable(u32 instanceName)
+{
+	int ret = 0;
+	if (hdmi_opt_clk_state)
+		ret = hdmi_w1_wrapper_disable(instanceName);
+	return ret;
+}
+EXPORT_SYMBOL(hdmi_wrapper_disable);
+int hdmi_start_audio_transfer(u32 instanceName)
+{
+	int ret = -EINVAL;
+	if(hdmi_opt_clk_state)
+		ret = hdmi_w1_start_audio_transfer( instanceName);
+	return ret;
+}
+EXPORT_SYMBOL(hdmi_start_audio_transfer);
+int hdmi_stop_audio_transfer(u32 instanceName)
+{
+	int ret = 0;
+	if(hdmi_opt_clk_state)
+		ret = hdmi_w1_stop_audio_transfer(instanceName);
+	return ret;
+}
+EXPORT_SYMBOL(hdmi_stop_audio_transfer);
Index: kernel/sound/soc/omap/omap-hdmi.c
===================================================================
--- kernel.orig/sound/soc/omap/omap-hdmi.c	2011-06-17 12:00:08.000000000 +0530
+++ kernel/sound/soc/omap/omap-hdmi.c	2011-06-24 11:49:58.212049047 +0530
@@ -102,14 +102,14 @@
 		break;
 	case HDMI_EVENT_POWERPHYON:
 		if (substream) {
-			hdmi_w1_wrapper_enable(HDMI_WP);
-			hdmi_w1_start_audio_transfer(HDMI_WP);
+			hdmi_wrapper_enable(HDMI_WP);
+			hdmi_start_audio_transfer(HDMI_WP);
 		}
 		break;
 	case HDMI_EVENT_POWERPHYOFF:
 		if (substream) {
-			hdmi_w1_stop_audio_transfer(HDMI_WP);
-			hdmi_w1_wrapper_disable(HDMI_WP);
+			hdmi_stop_audio_transfer(HDMI_WP);
+			hdmi_wrapper_disable(HDMI_WP);
 		}
 		break;
 	default:
@@ -142,7 +142,7 @@
 				    struct snd_soc_dai *dai)
 {
 	int err = 0;
-	hdmi_w1_wrapper_disable(HDMI_WP);
+	hdmi_wrapper_disable(HDMI_WP);
 	hdmi_data.substream = NULL;
 
 	if (hdmi_data.active)
@@ -174,13 +174,13 @@
 	case SNDRV_PCM_TRIGGER_PAUSE_RELEASE:
 		if (cmd == SNDRV_PCM_TRIGGER_RESUME)
 			hdmi_set_audio_power(1);
-		err = hdmi_w1_start_audio_transfer(HDMI_WP);
+		err = hdmi_start_audio_transfer(HDMI_WP);
 		break;
 
 	case SNDRV_PCM_TRIGGER_STOP:
 	case SNDRV_PCM_TRIGGER_SUSPEND:
 	case SNDRV_PCM_TRIGGER_PAUSE_PUSH:
-		err = hdmi_w1_stop_audio_transfer(HDMI_WP);
+		err = hdmi_stop_audio_transfer(HDMI_WP);
 		if (err)
 			return err;
 		if (cmd == SNDRV_PCM_TRIGGER_SUSPEND)
Index: kernel/arch/arm/plat-omap/include/plat/hdmi_lib.h
===================================================================
--- kernel.orig/arch/arm/plat-omap/include/plat/hdmi_lib.h	2011-06-17 12:00:01.000000000 +0530
+++ kernel/arch/arm/plat-omap/include/plat/hdmi_lib.h	2011-06-24 11:49:58.222048013 +0530
@@ -564,5 +564,9 @@
 int hdmi_lib_start_acr_wa(void);
 int hdmi_lib_stop_acr_wa(void);
 #endif
+int hdmi_wrapper_enable(u32 instanceName); 
+int hdmi_wrapper_disable(u32 instanceName);
+int hdmi_start_audio_transfer(u32 instanceName);
+int hdmi_stop_audio_transfer(u32 instanceName);
 #endif
 
