Index: kernel/drivers/video/omap2/dss/dss.c
===================================================================
--- kernel.orig/drivers/video/omap2/dss/dss.c	2011-05-16 14:22:42.000000000 +0530
+++ kernel/drivers/video/omap2/dss/dss.c	2011-06-21 10:45:44.410304557 +0530
@@ -160,8 +160,25 @@
 		if (ret)
 			dss_opt_clock_disable();
 #ifdef CONFIG_PM_RUNTIME
-		else
-			ret = pm_runtime_get_sync(&dss.pdev->dev);
+		else {
+				ret = pm_runtime_get_sync(&dss.pdev->dev);
+				
+				/* Work around added by Shankar
+				 * some time pm_runtime_get_sync fails to enable DSS clock
+				 */
+				{
+					#define CM_DSS_DSS_CLKCTRL 0x4A009120
+					int val =0;
+	           		val = omap_readl(CM_DSS_DSS_CLKCTRL);
+           			if (((val & 0x02) != 0x02) && ret == 0 ){
+	              		val = val | (1<<1);
+	               		omap_writel(val, CM_DSS_DSS_CLKCTRL);
+	               		printk("%s[%d] pm_runtime_get_sync failed set DSS clk reg, setting it forcefully clk=0x%x \n",
+			                       __func__, __LINE__, omap_readl(CM_DSS_DSS_CLKCTRL));
+					}
+				}
+
+		}
 #endif
 
 		if (!ret)
