Index: kernel/drivers/video/omap2/dss/hdmi.c
===================================================================
--- kernel.orig/drivers/video/omap2/dss/hdmi.c	2011-06-24 12:20:02.699543966 +0530
+++ kernel/drivers/video/omap2/dss/hdmi.c	2011-06-24 12:21:18.752057903 +0530
@@ -1539,7 +1539,7 @@
 	mutex_lock(&hdmi.lock_aux);
 	if (dssdev == NULL || dssdev->state != OMAP_DSS_DISPLAY_ACTIVE)
 		/* HDMI is disabled, there cannot be any HDMI irqs */
-		goto done;
+		goto done2;
 
 	DSSINFO("irqstatus=%d hdmi/vid_power=%d/%d\n",
 		r, hdmi_power, video_power);
@@ -1583,7 +1583,7 @@
 
 		if (dssdev->state != OMAP_DSS_DISPLAY_ACTIVE)
 			/* HDMI is disabled, no need to process */
-			goto done;
+			goto done2;
 
 		HDMI_W1_StopVideoFrame(HDMI_WP);
 		if (dssdev->platform_disable)
@@ -1631,10 +1631,10 @@
 
 		edid_set = false;
 		custom_set = true;
-		hdmi_reconfigure(dssdev);
 		if (dssdev->state != OMAP_DSS_DISPLAY_ACTIVE)
 			/* HDMI is disabled, no need to process */
-			goto done;
+			goto done2;
+		hdmi_reconfigure(dssdev);
 		custom_set = false;
 		DSSINFO("Physical Connect Done:\n\n");
 	}
