Index: kernel/drivers/video/omap2/dss/hdmi.c
===================================================================
--- kernel.orig/drivers/video/omap2/dss/hdmi.c	2011-07-01 17:34:06.602153503 +0530
+++ kernel/drivers/video/omap2/dss/hdmi.c	2011-07-01 17:34:12.172149058 +0530
@@ -1718,6 +1718,14 @@
 		DSSINFO("Reconfigure HDMI PHY Done- HDMI_HPD_MODIFY\n\n");
 	}
 
+	if (r & HDMI_HPD_LOW &&
+		hdmi_power == HDMI_POWER_FULL &&
+		hdmi_connected) {
+		DSSINFO("Switch AV on HDMI_HPD_LOW \n");
+		hdmi_notify_status(dssdev, false);
+	}
+
+
 done2:
 	mutex_unlock(&hdmi.lock_aux);
 	mutex_unlock(&hdmi.lock);
