Index: kernel/sound/soc/omap/omap-mcpdm.c
===================================================================
--- kernel.orig/sound/soc/omap/omap-mcpdm.c	2011-07-06 00:39:16.791998794 +0530
+++ kernel/sound/soc/omap/omap-mcpdm.c	2011-07-06 01:00:28.021998288 +0530
@@ -754,7 +754,7 @@
 			cancel_delayed_work_sync(&mcpdm->delayed_abe_work); 
 		wake_lock_timeout(&mcpdm->wake_lock, 2 * HZ);
 		schedule_delayed_work(&mcpdm->delayed_abe_work,
-				      msecs_to_jiffies(1000)); /* TODO: pdata ? */
+				      msecs_to_jiffies(550)); /* TODO: pdata ? */
 	}
 
 	mutex_unlock(&mcpdm->mutex);
@@ -767,6 +767,7 @@
 	struct omap_mcpdm *mcpdm =
 			container_of(work, struct omap_mcpdm, delayed_abe_work.work);
 
+	printk("%s:%d mcpdm_abe_dai_shutdown starts...\n",__func__,__LINE__);
 	mutex_lock(&mcpdm->mutex);
 	if (!mcpdm->dl_active && !mcpdm->ul_active) {
 		abe_disable_data_transfer(PDM_DL_PORT);
@@ -783,6 +784,7 @@
 		omap_mcpdm_free(mcpdm);
 	
 	mutex_unlock(&mcpdm->mutex);
+	printk("%s:%d mcpdm_abe_dai_shutdown ends...\n",__func__,__LINE__);
 
 }
 
Index: kernel/sound/soc/omap/sdp4430.c
===================================================================
--- kernel.orig/sound/soc/omap/sdp4430.c	2011-07-06 00:39:16.791998794 +0530
+++ kernel/sound/soc/omap/sdp4430.c	2011-07-06 01:01:13.982622424 +0530
@@ -122,7 +122,7 @@
 	mutex_lock(&sdp4430_data.mutex);
 	/* mcpdm shutdown work is delayed 1s wrt stream close */
 	schedule_delayed_work(&sdp4430_data.delayed_work,
-		msecs_to_jiffies(1500));
+		msecs_to_jiffies(1050));
 	sdp4430_data.twl6040_active--;
 	mutex_unlock(&sdp4430_data.mutex);
 }
