Index: kernel/drivers/video/omap2/dss/dispc.c
===================================================================
--- kernel.orig/drivers/video/omap2/dss/dispc.c	2011-07-14 13:16:53.390522314 +0530
+++ kernel/drivers/video/omap2/dss/dispc.c	2011-07-14 13:54:07.299278557 +0530
@@ -2439,7 +2439,7 @@
 	   _dispc_set_vid_accu2_1(plane, accuh, accu1); */
 }
 static void _dispc_set_rotation_attrs(enum omap_plane plane, u8 rotation,
-		bool mirroring, enum omap_color_mode color_mode)
+		bool mirroring, enum omap_color_mode color_mode, enum device_n_buffer_type ilace)
 {
 	BUG_ON(plane == OMAP_DSS_WB);
 
@@ -2483,7 +2483,7 @@
 	if (plane != OMAP_DSS_GFX) {
 		if (color_mode == OMAP_DSS_COLOR_NV12) {
 			/* DOUBLESTRIDE : 0 for 90-, 270-; 1 for 0- and 180- */
-			if (rotation == 1 || rotation == 3)
+			if (ilace || (rotation == 1 || rotation == 3))
 				REG_FLD_MOD(dispc_reg_att[plane], 0x0, 22, 22);
 			else
 				REG_FLD_MOD(dispc_reg_att[plane], 0x1, 22, 22);
@@ -3129,8 +3129,10 @@
 			height /= 2;
 #endif
 		pic_height /= 2;
-		if (ilace & OMAP_FLAG_IDEV)
+		if (ilace & OMAP_FLAG_IDEV){
 			out_height /= 2;
+			pos_y /=2;
+		}
 
 		DSSDBG("adjusting for ilace: height %d, pos_y %d, "
 				"out_height %d\n",
@@ -3328,7 +3330,7 @@
 
 	if (OMAP_DSS_COLOR_NV12 == color_mode) {
 		_dispc_set_plane_ba_uv0(plane, puv_addr + offset0);
-		_dispc_set_plane_ba_uv1(plane, puv_addr + offset1);
+		_dispc_set_plane_ba_uv1(plane, puv_addr + offset0);
 	}
 
 	_dispc_set_row_inc(plane, row_inc);
@@ -3354,7 +3356,8 @@
 		/* account for chroma decimation */
 		switch (color_mode) {
 		case OMAP_DSS_COLOR_NV12:
-			ch_height >>= 1; /* UV is subsampled by 2 vertically*/
+			if(!ilace)
+				ch_height >>= 1; /* UV is subsampled by 2 vertically*/
 		case OMAP_DSS_COLOR_YUV2:
 		case OMAP_DSS_COLOR_UYVY:
 			/*For YUV422 with rotation, we don't upsample chroma*/
@@ -3412,7 +3415,7 @@
 		_dispc_enable_vid_color_conv(plane, cconv);
 	}
 
-	_dispc_set_rotation_attrs(plane, rotation, mirror, color_mode);
+	_dispc_set_rotation_attrs(plane, rotation, mirror, color_mode, ilace);
 
 	if ((plane != OMAP_DSS_VIDEO1) || (cpu_is_omap44xx()))
 		_dispc_setup_global_alpha(plane, global_alpha);
