Index: kernel/sound/soc/omap/omap-mcpdm.c
===================================================================
--- kernel.orig/sound/soc/omap/omap-mcpdm.c	2011-07-25 10:11:09.000000000 +0530
+++ kernel/sound/soc/omap/omap-mcpdm.c	2011-07-25 10:40:03.188248341 +0530
@@ -771,6 +771,14 @@
 	mutex_lock(&mcpdm->mutex);
 
 	if (!mcpdm->free && !mcpdm->dl_active && !mcpdm->ul_active) {
+		/*
+		 * Problem: abe_block_copy kernel panic at T1-OMAP4 PVR/PRA/SRA stages.
+		 * Issue: AESS CLK CTRL is disabled by software.
+		 * Fix: Work around -> Added additional abe_dsp_pm_get to hold the resource which 
+		 * will ensure the AESS clock will be active before accessing by incrementing 
+		 * the device's usage counter. 
+		 */
+		abe_dsp_pm_get();
 		abe_disable_data_transfer(PDM_DL_PORT);
 		abe_disable_data_transfer(PDM_UL_PORT);
 		udelay(250);
@@ -779,6 +787,14 @@
 		omap_mcpdm_capture_close(mcpdm, mcpdm->uplink);
 		abe_dsp_mcpdm_shutdown();
 		abe_dsp_pm_put();
+		/*
+		 * Problem: abe_block_copy kernel panic at T1-OMAP4 PVR/PRA/SRA stages.
+		 * Issue: AESS CLK CTRL is disabled by software.
+		 * Fix: Work around -> Added additional abe_dsp_pm_put to release the resource which 
+		 * will ensure the AESS clock will be disabled after accessing by decrementing 
+		 * the device's usage counter. 
+		 */
+		abe_dsp_pm_put();
 		omap_mcpdm_free(mcpdm);
 	}
 
