Index: kernel/arch/arm/plat-omap/include/plat/display.h
===================================================================
--- kernel.orig/arch/arm/plat-omap/include/plat/display.h	2011-08-05 16:37:40.380640304 +0530
+++ kernel/arch/arm/plat-omap/include/plat/display.h	2011-08-05 16:38:31.719397036 +0530
@@ -685,6 +685,7 @@
 	int (*apply)(struct omap_overlay_manager *mgr);
 	int (*wait_for_go)(struct omap_overlay_manager *mgr);
 	int (*wait_for_vsync)(struct omap_overlay_manager *mgr);
+	bool (*get_transient_state)(struct omap_overlay_manager *mgr);
 
 	int (*enable)(struct omap_overlay_manager *mgr);
 	int (*disable)(struct omap_overlay_manager *mgr);
Index: kernel/drivers/video/omap2/dss/dpi.c
===================================================================
--- kernel.orig/drivers/video/omap2/dss/dpi.c	2011-08-05 16:30:23.840639598 +0530
+++ kernel/drivers/video/omap2/dss/dpi.c	2011-08-05 16:42:33.836895004 +0530
@@ -252,7 +252,12 @@
 	dssdev->state = OMAP_DSS_DISPLAY_DISABLED;
 	if (!cpu_is_omap44xx())
 		dss_clk_disable(DSS_CLK_ICK | DSS_CLK_FCK1);
-	dss_mainclk_state_disable(true);
+
+	struct omap_overlay_manager *mgr;
+	mgr = dssdev->manager;
+	if(!mgr || !mgr->get_transient_state(mgr))
+		dss_mainclk_state_disable(true);
+
 	if (cpu_is_omap34xx() && !cpu_is_omap3630())
 		regulator_disable(dpi.vdds_dsi_reg);
 err0:
@@ -282,7 +287,10 @@
 	dssdev->state = OMAP_DSS_DISPLAY_DISABLED;
 	if (!cpu_is_omap44xx())
 		dss_clk_disable(DSS_CLK_ICK | DSS_CLK_FCK1);
-	dss_mainclk_state_disable(true);
+	struct omap_overlay_manager *mgr;
+	mgr = dssdev->manager;
+	if(!mgr || !mgr->get_transient_state(mgr))
+		dss_mainclk_state_disable(true);
 
 	if (cpu_is_omap34xx() && !cpu_is_omap3630())
 		regulator_disable(dpi.vdds_dsi_reg);
Index: kernel/drivers/video/omap2/dss/dsi.c
===================================================================
--- kernel.orig/drivers/video/omap2/dss/dsi.c	2011-08-05 16:30:43.110618280 +0530
+++ kernel/drivers/video/omap2/dss/dsi.c	2011-08-05 16:42:33.836895004 +0530
@@ -4013,7 +4013,11 @@
 
 	/* cut clocks(s) */
 	dssdev->state = OMAP_DSS_DISPLAY_DISABLED;
-	dss_mainclk_state_disable(true);
+
+	struct omap_overlay_manager *mgr;
+	mgr = dssdev->manager;
+	if(!mgr || !mgr->get_transient_state(mgr))
+		dss_mainclk_state_disable(true);
 
 	p_dsi->enabled = false;
 
Index: kernel/drivers/video/omap2/dss/hdmi.c
===================================================================
--- kernel.orig/drivers/video/omap2/dss/hdmi.c	2011-08-05 16:30:37.010616002 +0530
+++ kernel/drivers/video/omap2/dss/hdmi.c	2011-08-05 16:42:33.836895004 +0530
@@ -1860,9 +1860,12 @@
 	hdmi_enable_clocks(0);
 
 	/* cut clock(s) only if disconnected */
-	if(!hdmi_connected)
-		dss_mainclk_state_disable(true);
-
+	if(!hdmi_connected) {
+		struct omap_overlay_manager *mgr;
+		mgr = dssdev->manager;
+		if(!mgr || !mgr->get_transient_state(mgr))
+			dss_mainclk_state_disable(true);
+	}
 	/* reset to default */
 }
 
Index: kernel/drivers/video/omap2/dss/manager.c
===================================================================
--- kernel.orig/drivers/video/omap2/dss/manager.c	2011-08-05 16:30:29.040639501 +0530
+++ kernel/drivers/video/omap2/dss/manager.c	2011-08-05 16:47:50.549399125 +0530
@@ -37,6 +37,7 @@
 #define MAX_DSS_MANAGERS (cpu_is_omap44xx() ? 3 : 2)
 
 static int num_managers;
+static bool mgr_transient = false;
 static struct list_head manager_list;
 
 static ssize_t manager_name_show(struct omap_overlay_manager *mgr, char *buf)
@@ -557,6 +558,11 @@
 	return omap_dispc_wait_for_irq_interruptible_timeout(irq, timeout);
 }
 
+static bool dss_mgr_get_transient_state(struct omap_overlay_manager *mgr)
+{
+       return mgr_transient;
+}
+
 static int dss_mgr_wait_for_go(struct omap_overlay_manager *mgr)
 {
 	unsigned long timeout = msecs_to_jiffies(500);
@@ -1374,7 +1380,7 @@
 			"mgr_apply, returning\n");
 		return 0;
 	}
-
+	mgr_transient = true;
 	spin_lock_irqsave(&dss_cache.lock, flags);
 
 	if (cpu_is_omap44xx())
@@ -1574,6 +1580,9 @@
 	}
 	configure_dispc();
 
+	mgr_transient = false;
+	dss_mainclk_state_disable(true);
+
 	spin_unlock_irqrestore(&dss_cache.lock, flags);
 
 	return r;
@@ -1872,6 +1881,7 @@
 		mgr->get_manager_info = &omap_dss_mgr_get_info;
 		mgr->wait_for_go = &dss_mgr_wait_for_go;
 		mgr->wait_for_vsync = &dss_mgr_wait_for_vsync;
+		mgr->get_transient_state = &dss_mgr_get_transient_state;
 
 		mgr->enable = &dss_mgr_enable;
 		mgr->disable = &dss_mgr_disable;
